
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>riboSeed &#8212; riboSeed 0.4.9 documentation</title>
    <link rel="stylesheet" href="../_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.4.9',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
  </head>
  <body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">riboSeed 0.4.9 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" accesskey="U">Module code</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for riboSeed</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python3</span>
<span class="c1">#-*- coding: utf-8 -*-</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Created on Sun Jul 24 19:33:37 2016</span>

<span class="sd">See README.md for more info and usage</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">shutil</span>
<span class="kn">import</span> <span class="nn">multiprocessing</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">traceback</span>
<span class="kn">import</span> <span class="nn">pysam</span>
<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">pkg_resources</span>

<span class="k">try</span><span class="p">:</span>  <span class="c1"># development mode</span>
    <span class="kn">from</span> <span class="nn">_version</span> <span class="k">import</span> <span class="n">__version__</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>  <span class="c1"># ie, if an installed pkg from pip or other using setup.py</span>
    <span class="n">__version__</span> <span class="o">=</span> <span class="n">pkg_resources</span><span class="o">.</span><span class="n">require</span><span class="p">(</span><span class="s2">&quot;riboSeed&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">version</span>

<span class="kn">from</span> <span class="nn">bisect</span> <span class="k">import</span> <span class="n">bisect</span>
<span class="kn">from</span> <span class="nn">itertools</span> <span class="k">import</span> <span class="n">chain</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">namedtuple</span>
<span class="kn">from</span> <span class="nn">Bio</span> <span class="k">import</span> <span class="n">SeqIO</span>
<span class="kn">from</span> <span class="nn">Bio.Seq</span> <span class="k">import</span> <span class="n">Seq</span>
<span class="kn">from</span> <span class="nn">Bio.SeqRecord</span> <span class="k">import</span> <span class="n">SeqRecord</span>
<span class="kn">from</span> <span class="nn">Bio.Alphabet</span> <span class="k">import</span> <span class="n">IUPAC</span>

<span class="k">try</span><span class="p">:</span>
    <span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
    <span class="kn">import</span> <span class="nn">matplotlib</span> <span class="k">as</span> <span class="nn">mpl</span>
    <span class="n">mpl</span><span class="o">.</span><span class="n">use</span><span class="p">(</span><span class="s1">&#39;Agg&#39;</span><span class="p">)</span>
    <span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
    <span class="kn">import</span> <span class="nn">matplotlib.patches</span> <span class="k">as</span> <span class="nn">patches</span>
    <span class="n">PLOT</span> <span class="o">=</span> <span class="kc">True</span>
<span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>  <span class="c1"># most likely an ImportError, but Im not taking chances</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">looks like you have some issue with matplotlib.  &quot;</span> <span class="o">+</span>
          <span class="s2">&quot;Classic matplotlib, amirite? Plotting is disabled</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">PLOT</span> <span class="o">=</span> <span class="kc">False</span>


<span class="c1"># need this line for unittesting</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s1">&#39;..&#39;</span><span class="p">,</span> <span class="s1">&#39;riboSeed&#39;</span><span class="p">))</span>

<span class="kn">from</span> <span class="nn">pyutilsnrw.utils3_5</span> <span class="k">import</span> <span class="n">set_up_logging</span><span class="p">,</span> \
    <span class="n">combine_contigs</span><span class="p">,</span> <span class="n">get_ave_read_len_from_fastq</span><span class="p">,</span> \
    <span class="n">get_number_mapped</span><span class="p">,</span> \
    <span class="n">keep_only_first_contig</span><span class="p">,</span> <span class="n">get_fasta_lengths</span><span class="p">,</span> \
    <span class="n">file_len</span><span class="p">,</span> <span class="n">check_version_from_cmd</span>

<span class="kn">from</span> <span class="nn">riboSnag</span> <span class="k">import</span> <span class="n">parse_clustered_loci_file</span><span class="p">,</span> <span class="n">pad_genbank_sequence</span><span class="p">,</span> \
    <span class="n">extract_coords_from_locus</span>

<span class="c1"># GLOBALS</span>
<span class="n">SAMTOOLS_MIN_VERSION</span> <span class="o">=</span> <span class="s1">&#39;1.3.1&#39;</span>
<span class="c1"># --------------------------- classes --------------------------- #</span>


<div class="viewcode-block" id="SeedGenome"><a class="viewcode-back" href="../riboSeed.html#riboSeed.SeedGenome">[docs]</a><span class="k">class</span> <span class="nc">SeedGenome</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; This object is the &quot;master&quot; object which holds slots for each</span>
<span class="sd">    cluster&#39;s mappings, the sequencing library, and keeps track of the</span>
<span class="sd">    current iteration as execution progresses.</span>

<span class="sd">    When instantiated, self.check_mands() checks that all required attributes</span>
<span class="sd">    are present, self.attach_genome_seqRecords() parses and loads the gb data,</span>
<span class="sd">    self.write_fasta_genome() writes a .fasta version of the genome for later</span>
<span class="sd">    use, and self.make_map_paths_and_dir() sets up the required directories for</span>
<span class="sd">    each iteration</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genbank_path</span><span class="p">,</span> <span class="n">final_long_reads_dir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">this_iteration</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">ref_fasta</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">next_reference_path</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">loci_clusters</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">output_root</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">initial_map_bam</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">unmapped_ngsLib</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">iter_mapping_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">reads_mapped_txt</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">unmapped_mapping_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">max_iterations</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">initial_map_sam</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">unmapped_sam</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">clustered_loci_txt</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">seq_records</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">master_ngs_ob</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">initial_map_sorted_bam</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">initial_map_prefix</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">assembled_seeds</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">seq_records_count</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>  <span class="c1"># get from commsanline in case running multiple</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">this_iteration</span> <span class="o">=</span> <span class="n">this_iteration</span>  <span class="c1"># this should always start at 0</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">max_iterations</span> <span class="o">=</span> <span class="n">max_iterations</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">iter_mapping_list</span> <span class="o">=</span> <span class="n">iter_mapping_list</span>  <span class="c1"># holds each mapping object</span>
        <span class="c1"># The main output for resulting files, all other are relative</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">output_root</span> <span class="o">=</span> <span class="n">output_root</span>
        <span class="c1"># from command line</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genbank_path</span> <span class="o">=</span> <span class="n">genbank_path</span>
        <span class="c1"># This is created from genbank</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ref_fasta</span> <span class="o">=</span> <span class="n">ref_fasta</span>  <span class="c1"># this is made dynamically</span>
        <span class="c1"># output from riboSelect</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">clustered_loci_txt</span> <span class="o">=</span> <span class="n">clustered_loci_txt</span>
        <span class="c1"># holds a list of LociCluster objects</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">loci_clusters</span> <span class="o">=</span> <span class="n">loci_clusters</span>
        <span class="c1"># this set below with attach_genome_seqRecor</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seq_records</span> <span class="o">=</span> <span class="n">seq_records</span>  <span class="c1"># this is set</span>
        <span class="c1"># this will hold prefix for initial mapping</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">initial_map_prefix</span> <span class="o">=</span> <span class="n">initial_map_prefix</span>  <span class="c1"># set this dynamically</span>
        <span class="c1"># extracting reads by position requires an indexed, sorted bam</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">initial_map_sorted_bam</span> <span class="o">=</span> <span class="n">initial_map_sorted_bam</span>  <span class="c1"># set dynamically</span>
        <span class="c1"># inial mapping result (combined s and pe)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">initial_map_bam</span> <span class="o">=</span> <span class="n">initial_map_bam</span>  <span class="c1"># set this dynamically</span>
        <span class="c1"># see above</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">initial_map_sam</span> <span class="o">=</span> <span class="n">initial_map_sam</span>  <span class="c1"># set this dynamically</span>
        <span class="c1"># holds user-provided sequencing data. Keep intact for final assembly</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">master_ngs_ob</span> <span class="o">=</span> <span class="n">master_ngs_ob</span>  <span class="c1"># for ngslib object</span>
        <span class="c1"># each round of seeding results in a ml list for remaining reads</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">unmapped_mapping_list</span> <span class="o">=</span> <span class="n">unmapped_mapping_list</span>
        <span class="c1"># holds dynamically updated list of remaining unmapped</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">unmapped_sam</span> <span class="o">=</span> <span class="n">unmapped_sam</span>
        <span class="c1"># after partitioning, the last mapping list is extracted into this ngs ob</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">unmapped_ngsLib</span> <span class="o">=</span> <span class="n">unmapped_ngsLib</span>
        <span class="c1"># path to file mapped read names are appended to</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reads_mapped_txt</span> <span class="o">=</span> <span class="n">reads_mapped_txt</span>
        <span class="c1"># destination for seeded contigs prior to final assemblies</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">final_long_reads_dir</span> <span class="o">=</span> <span class="n">final_long_reads_dir</span>
        <span class="c1"># after faux genome construction, store path here</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">next_reference_path</span> <span class="o">=</span> <span class="n">next_reference_path</span>
        <span class="c1"># where to put the combined contigs at the end:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">assembled_seeds</span> <span class="o">=</span> <span class="n">assembled_seeds</span>
        <span class="c1"># this is set dynamically as well</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seq_records_count</span> <span class="o">=</span> <span class="n">seq_records_count</span>
        <span class="c1"># a logger</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span> <span class="o">=</span> <span class="n">logger</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">check_mands</span><span class="p">()</span>
        <span class="c1"># self.attach_genome_seqrecords()  # this method comes first,</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">refreshSeqRecGenerator</span><span class="p">()</span>  <span class="c1"># this method comes first,</span>
        <span class="c1"># sets self.seq_records_count</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">countSeqRecords</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">write_fasta_genome</span><span class="p">()</span>  <span class="c1"># because this method relies on it</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">make_map_paths_and_dir</span><span class="p">()</span>

<div class="viewcode-block" id="SeedGenome.check_mands"><a class="viewcode-back" href="../riboSeed.html#riboSeed.SeedGenome.check_mands">[docs]</a>    <span class="k">def</span> <span class="nf">check_mands</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; check that all mandatory arguments are not none</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">mandatory</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">genbank_path</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_iterations</span><span class="p">,</span>
                     <span class="bp">self</span><span class="o">.</span><span class="n">output_root</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">clustered_loci_txt</span><span class="p">]</span>
        <span class="k">if</span> <span class="kc">None</span> <span class="ow">in</span> <span class="n">mandatory</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;SeedGenome must be instantiated with at least &quot;</span>
                             <span class="s2">&quot;genbank_path, max_iterations, cluster file, &quot;</span> <span class="o">+</span>
                             <span class="s2">&quot;and output_root&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="SeedGenome.make_map_paths_and_dir"><a class="viewcode-back" href="../riboSeed.html#riboSeed.SeedGenome.make_map_paths_and_dir">[docs]</a>    <span class="k">def</span> <span class="nf">make_map_paths_and_dir</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Given a output root, prepare all the needed subdirs and paths</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">iter_mapping_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_iterations</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">LociMapping</span><span class="p">(</span>
                <span class="n">name</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_mapping_iteration_</span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">i</span><span class="p">),</span>
                <span class="n">iteration</span><span class="o">=</span><span class="n">i</span><span class="p">,</span>
                <span class="n">mapping_subdir</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">output_root</span><span class="p">,</span>
                    <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_mapping_for_iteration_</span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">i</span><span class="p">)),</span>
                <span class="n">assembly_subdir_needed</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">final_long_reads_dir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">final_long_reads_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_root</span><span class="p">,</span>
                                                     <span class="s2">&quot;final_long_reads&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">final_long_reads_dir</span><span class="p">):</span>
            <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">final_long_reads_dir</span><span class="p">)</span></div>

<div class="viewcode-block" id="SeedGenome.write_fasta_genome"><a class="viewcode-back" href="../riboSeed.html#riboSeed.SeedGenome.write_fasta_genome">[docs]</a>    <span class="k">def</span> <span class="nf">write_fasta_genome</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Given a genbank file, write out as (multi)fasta</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genbank_path</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ref_fasta</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">output_root</span><span class="p">,</span>
                                      <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">+</span> <span class="s2">&quot;.fasta&quot;</span><span class="p">))</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genbank_path</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ref_fasta</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfh</span><span class="p">:</span>
                <span class="n">sequences</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">)</span>
                <span class="n">count</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">sequences</span><span class="p">,</span> <span class="n">outfh</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">count</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_records_count</span><span class="p">,</span> <span class="s2">&quot;Error parsing genbank file!&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">refreshSeqRecGenerator</span><span class="p">()</span></div>

<div class="viewcode-block" id="SeedGenome.pad_genbank"><a class="viewcode-back" href="../riboSeed.html#riboSeed.SeedGenome.pad_genbank">[docs]</a>    <span class="k">def</span> <span class="nf">pad_genbank</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pad</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span> <span class="n">circular</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; if the genome is circular (which it is, by default) adjust the</span>
<span class="sd">        cluster coordinates and rewrite the reference fasta as padded.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">circular</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">clu</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">loci_clusters</span><span class="p">:</span>
                <span class="n">clu</span><span class="o">.</span><span class="n">padding</span> <span class="o">=</span> <span class="n">pad</span>
                <span class="n">clu</span> <span class="o">=</span> <span class="n">pad_genbank_sequence</span><span class="p">(</span><span class="n">cluster</span><span class="o">=</span><span class="n">clu</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;rewriting fasta-formated genome padded &quot;</span> <span class="o">+</span>
                        <span class="s2">&quot;by </span><span class="si">%i</span><span class="s2"> bases on each end&quot;</span><span class="p">,</span> <span class="n">pad</span><span class="p">)</span>
            <span class="n">new_fasta_ref</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ref_fasta</span><span class="p">),</span>
                <span class="nb">str</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ref_fasta</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span>
                    <span class="s2">&quot;_padded.fasta&quot;</span><span class="p">))</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genbank_path</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">fh</span><span class="p">:</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">new_fasta_ref</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfh</span><span class="p">:</span>
                    <span class="n">recs</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fh</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">)</span>
                    <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
                    <span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">recs</span><span class="p">:</span>
                        <span class="n">new_rec</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span>
                            <span class="nb">id</span><span class="o">=</span><span class="n">rec</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                            <span class="n">seq</span><span class="o">=</span><span class="n">Seq</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="o">-</span><span class="n">pad</span><span class="p">:]</span> <span class="o">+</span> <span class="n">rec</span><span class="o">.</span><span class="n">seq</span> <span class="o">+</span>
                                    <span class="n">rec</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span> <span class="n">pad</span><span class="p">]),</span>
                                    <span class="n">IUPAC</span><span class="o">.</span><span class="n">IUPACAmbiguousDNA</span><span class="p">()))</span>
                        <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">new_rec</span><span class="p">,</span> <span class="n">outfh</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
                        <span class="n">count</span> <span class="o">=</span> <span class="n">count</span> <span class="o">+</span> <span class="mi">1</span>
                    <span class="k">assert</span> <span class="n">count</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_records_count</span><span class="p">,</span> \
                        <span class="s2">&quot;Error parsing genbank file!&quot;</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ref_fasta</span> <span class="o">=</span> <span class="n">new_fasta_ref</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span></div>

<div class="viewcode-block" id="SeedGenome.countSeqRecords"><a class="viewcode-back" href="../riboSeed.html#riboSeed.SeedGenome.countSeqRecords">[docs]</a>    <span class="k">def</span> <span class="nf">countSeqRecords</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Nothing fancy; just a quick way to get the number of records</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seq_records_count</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">seq_records</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">refreshSeqRecGenerator</span><span class="p">()</span></div>

<div class="viewcode-block" id="SeedGenome.refreshSeqRecGenerator"><a class="viewcode-back" href="../riboSeed.html#riboSeed.SeedGenome.refreshSeqRecGenerator">[docs]</a>    <span class="k">def</span> <span class="nf">refreshSeqRecGenerator</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;instead of using stored genbank records, this method restarts</span>
<span class="sd">        the generator so that each time self.seq_records is accessed</span>
<span class="sd">        this method can get things going again</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">seq_records</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">genbank_path</span><span class="p">,</span> <span class="s2">&quot;genbank&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="SeedGenome.purge_old_files"><a class="viewcode-back" href="../riboSeed.html#riboSeed.SeedGenome.purge_old_files">[docs]</a>    <span class="k">def</span> <span class="nf">purge_old_files</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">all_iters</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; remove bulky files from two iterations ago, or if</span>
<span class="sd">        all_iters, remove all big mapping files</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="n">logger</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Must use logging&quot;</span>
        <span class="k">if</span> <span class="n">all_iters</span><span class="p">:</span>
            <span class="n">target_iters</span> <span class="o">=</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">max_iterations</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">target_iters</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">this_iteration</span> <span class="o">-</span> <span class="mi">2</span><span class="p">]</span>
            <span class="c1"># ensure that you are deleting files from not the previous, but the</span>
            <span class="c1"># TWICE previous iterations&#39;s files, (which cant be less than 0)</span>
            <span class="k">assert</span> <span class="n">target_iters</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">this_iteration</span> <span class="o">-</span> <span class="mi">1</span> <span class="ow">and</span> \
                <span class="n">target_iters</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">,</span> \
                <span class="s2">&quot;previous mapping is required, can only purge 2nd previous&quot;</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Looking for temp files to remove. &quot;</span> <span class="o">+</span>
                     <span class="s2">&quot;To keep all temp files, use the --keep_temps flag&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">target_iters</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">pe_map_bam</span><span class="p">,</span>
                      <span class="bp">self</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">s_map_bam</span><span class="p">,</span>
                      <span class="bp">self</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">mapped_sam</span><span class="p">,</span>
                      <span class="c1"># self.iter_mapping_list[i].mapped_bam,  # for riboStack</span>
                      <span class="bp">self</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">unmapped_sam</span><span class="p">,</span>
                      <span class="bp">self</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">unmapped_bam</span><span class="p">,</span>
                      <span class="bp">self</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">sorted_mapped_bam</span><span class="p">]:</span>
                <span class="k">if</span> <span class="n">f</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">f</span><span class="p">):</span>
                        <span class="n">os</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;deleting </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span></div></div>


<div class="viewcode-block" id="NgsLib"><a class="viewcode-back" href="../riboSeed.html#riboSeed.NgsLib">[docs]</a><span class="k">class</span> <span class="nc">NgsLib</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; NgsLib objects are used to hold the sequencing data supplied by the</span>
<span class="sd">    user (master) and the seq data extracted from each iteration. Currently the</span>
<span class="sd">    software requires either a paired-end or single-end library, but this</span>
<span class="sd">    should handle more diverse library types in the future.</span>

<span class="sd">    If ngsLib is master, read lengths are determined and if smalt is used for</span>
<span class="sd">    mapping, a distance estimation file is generated.</span>

<span class="sd">    when instatiated,self.check_mands() ensure required attributes have values,</span>
<span class="sd">    self.set_libtype() sets the libtype attribute based on libraries present,</span>
<span class="sd">    self.get_readlen() determines the read length if master, and  if master,</span>
<span class="sd">    self.smalt_insert_file() creates a distance estimation file for smalt</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">master</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">readF</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">readR</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">readS0</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">readS1</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">mapping_success</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                 <span class="n">smalt_dist_path</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">readlen</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">make_dist</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                 <span class="n">libtype</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">mapper_exe</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">liblist</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">ref_fasta</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>
        <span class="c1"># Bool: whether this is a master record</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">master</span> <span class="o">=</span> <span class="n">master</span>
        <span class="c1"># holds libtype detected from non-None libraries</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">libtype</span> <span class="o">=</span> <span class="n">libtype</span>  <span class="c1"># set this dynamically</span>
        <span class="c1"># forward Fastq path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">readF</span> <span class="o">=</span> <span class="n">readF</span>
        <span class="c1"># reverse Fastq path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">readR</span> <span class="o">=</span> <span class="n">readR</span>
        <span class="c1"># singleton Fastq path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">readS0</span> <span class="o">=</span> <span class="n">readS0</span>
        <span class="c1"># other singleton fastq path (thining ahead, but not really...)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">readS1</span> <span class="o">=</span> <span class="n">readS1</span>
        <span class="c1"># detected from Forward fastq with gget_readlen below</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">readlen</span> <span class="o">=</span> <span class="n">readlen</span>  <span class="c1"># set this dynamically</span>
        <span class="c1"># bool: did mapping have errors?</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mapping_success</span> <span class="o">=</span> <span class="n">mapping_success</span>
        <span class="c1"># needed to generate distance file if master</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mapper_exe</span> <span class="o">=</span> <span class="n">mapper_exe</span>
        <span class="c1"># whether to make a distance file for smalt</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">make_dist</span> <span class="o">=</span> <span class="n">make_dist</span>
        <span class="c1"># also needed to generate distance file if master</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ref_fasta</span> <span class="o">=</span> <span class="n">ref_fasta</span>
        <span class="c1"># results of distance mapping</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">smalt_dist_path</span> <span class="o">=</span> <span class="n">smalt_dist_path</span>  <span class="c1"># set this dynamically</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">liblist</span> <span class="o">=</span> <span class="n">liblist</span>  <span class="c1"># make dynamically</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span> <span class="o">=</span> <span class="n">logger</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">check_mands</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">set_libtype</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">get_readlen</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">smalt_insert_file</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">listLibs</span><span class="p">()</span>

<div class="viewcode-block" id="NgsLib.check_mands"><a class="viewcode-back" href="../riboSeed.html#riboSeed.NgsLib.check_mands">[docs]</a>    <span class="k">def</span> <span class="nf">check_mands</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; checks that all mandatory arguments are not none</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">mandatory_other</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">ref_fasta</span><span class="p">]</span>
        <span class="k">if</span> <span class="kc">None</span> <span class="ow">in</span> <span class="n">mandatory_other</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;SeedGenome must be instantiated with name, &quot;</span>
                             <span class="s2">&quot;and ref fasta&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="NgsLib.set_libtype"><a class="viewcode-back" href="../riboSeed.html#riboSeed.NgsLib.set_libtype">[docs]</a>    <span class="k">def</span> <span class="nf">set_libtype</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;sets to either s_1, pe, pe_s based on Non-none libraries</span>
<span class="sd">        TODO: add mate support and support for multiple single libraries</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">readF</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">readR</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">readS0</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                        <span class="s2">&quot;Must have at least either a paired library or &quot;</span> <span class="o">+</span>
                        <span class="s2">&quot;single library&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">libtype</span> <span class="o">=</span> <span class="s2">&quot;s_1&quot;</span>  <span class="c1"># single library</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;cannot set library type from one PE read&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">readS0</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">libtype</span> <span class="o">=</span> <span class="s2">&quot;pe_s&quot;</span>  <span class="c1"># paired end and single</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">libtype</span> <span class="o">=</span> <span class="s2">&quot;pe&quot;</span>  <span class="c1"># just paired end</span></div>

<div class="viewcode-block" id="NgsLib.get_readlen"><a class="viewcode-back" href="../riboSeed.html#riboSeed.NgsLib.get_readlen">[docs]</a>    <span class="k">def</span> <span class="nf">get_readlen</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; If NgsLib is master, estimate the read lengh using</span>
<span class="sd">        get_ave_read_len_from_fastq.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">master</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">True</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">libtype</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;pe&#39;</span><span class="p">,</span> <span class="s1">&#39;pe_s&#39;</span><span class="p">]:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">readlen</span> <span class="o">=</span> <span class="n">get_ave_read_len_from_fastq</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">readF</span><span class="p">,</span> <span class="n">N</span><span class="o">=</span><span class="mi">36</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">readlen</span> <span class="o">=</span> <span class="n">get_ave_read_len_from_fastq</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">readS0</span><span class="p">,</span> <span class="n">N</span><span class="o">=</span><span class="mi">36</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="p">)</span></div>

<div class="viewcode-block" id="NgsLib.smalt_insert_file"><a class="viewcode-back" href="../riboSeed.html#riboSeed.NgsLib.smalt_insert_file">[docs]</a>    <span class="k">def</span> <span class="nf">smalt_insert_file</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; Smalt mapper uses a subset of mapped reads to estimate distribution</span>
<span class="sd">        of insert sizes.  This file is used along with mapping, and is created</span>
<span class="sd">        if make_dist, master, and lib_type indicates paired data.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">master</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">True</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="bp">self</span><span class="o">.</span><span class="n">make_dist</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">libtype</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;pe&#39;</span><span class="p">,</span> <span class="s1">&#39;pe_s&#39;</span><span class="p">]:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">smalt_dist_path</span> <span class="o">=</span> <span class="n">estimate_distances_smalt</span><span class="p">(</span>
                <span class="n">outfile</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ref_fasta</span><span class="p">),</span>
                                     <span class="s2">&quot;smalt_distance_est.sam&quot;</span><span class="p">),</span>
                <span class="n">smalt_exe</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">mapper_exe</span><span class="p">,</span>
                <span class="n">ref_genome</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">ref_fasta</span><span class="p">,</span>
                <span class="n">fastq1</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">readF</span><span class="p">,</span> <span class="n">fastq2</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">readR</span><span class="p">,</span>
                <span class="n">logger</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;cannot create distance estimate for lib type </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span>
                  <span class="bp">self</span><span class="o">.</span><span class="n">libtype</span><span class="p">)</span>
            <span class="k">return</span> <span class="kc">None</span></div>

<div class="viewcode-block" id="NgsLib.purge_old_files"><a class="viewcode-back" href="../riboSeed.html#riboSeed.NgsLib.purge_old_files">[docs]</a>    <span class="k">def</span> <span class="nf">purge_old_files</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">master</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; before reasigning unmapped lib, delete</span>
<span class="sd">        useless files that were used in the previous iteration</span>
<span class="sd">        return codes:</span>
<span class="sd">         0) all is well</span>
<span class="sd">         1) no files deleted due to conflict</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="n">master</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> \
            <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;Must submit your master ngs lib as a control&quot;</span> <span class="o">+</span>
                <span class="s2">&quot; to prevent undesired deletions!&quot;</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">logger</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;must use logging&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">master</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;cannot remove master NgsLib!  skipping purge&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="mi">1</span>

        <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">readF</span><span class="p">,</span>
                  <span class="bp">self</span><span class="o">.</span><span class="n">readR</span><span class="p">,</span>
                  <span class="bp">self</span><span class="o">.</span><span class="n">readS0</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">f</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">f</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">f</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">master</span><span class="o">.</span><span class="n">liblist</span><span class="p">:</span>
                        <span class="n">os</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Deleted </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> is in the master ngs object; &quot;</span> <span class="o">+</span>
                                         <span class="s2">&quot;cannot delete!&quot;</span><span class="p">),</span> <span class="n">f</span><span class="p">)</span>
                        <span class="k">return</span> <span class="mi">1</span>
        <span class="k">return</span> <span class="mi">0</span></div>

<div class="viewcode-block" id="NgsLib.listLibs"><a class="viewcode-back" href="../riboSeed.html#riboSeed.NgsLib.listLibs">[docs]</a>    <span class="k">def</span> <span class="nf">listLibs</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">liblist</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>
                        <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">readF</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">readR</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">readS0</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">readS1</span><span class="p">]</span>
                        <span class="k">if</span> <span class="n">x</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">]</span></div></div>


<div class="viewcode-block" id="LociMapping"><a class="viewcode-back" href="../riboSeed.html#riboSeed.LociMapping">[docs]</a><span class="k">class</span> <span class="nc">LociMapping</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    instantiate with iteration, mapping subdir, and ref</span>
<span class="sd">    map_ref_genome_will use it and an ngslib for mapping</span>
<span class="sd">    extract_</span>
<span class="sd">    order of operations: map to reference, extract and convert,</span>
<span class="sd">    assemble, save results here</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">iteration</span><span class="p">,</span> <span class="n">mapping_subdir</span><span class="p">,</span> <span class="n">mapping_prefix</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">assembly_success</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                 <span class="n">ref_fasta</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">pe_map_bam</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">s_map_bam</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">sorted_mapped_bam</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">mapped_bam</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">mapped_sam</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">mapped_bam_unfiltered</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">unmapped_sam</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">mapped_ids_txt</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">unmapped_bam</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">mappedS</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">assembled_contig</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">assembly_subdir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">mapped_ngsLib</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">unmapped_ngsLib</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                 <span class="n">assembly_subdir_needed</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
        <span class="c1"># int: current iteration (0 is initial)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">iteration</span> <span class="o">=</span> <span class="n">iteration</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>
        <span class="c1"># bool: did the eassembly run without errors</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">assembly_success</span> <span class="o">=</span> <span class="n">assembly_success</span>
        <span class="c1"># results for the</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mapping_subdir</span> <span class="o">=</span> <span class="n">mapping_subdir</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">assembly_subdir</span> <span class="o">=</span> <span class="n">assembly_subdir</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">assembly_subdir_needed</span> <span class="o">=</span> <span class="n">assembly_subdir_needed</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ref_fasta</span> <span class="o">=</span> <span class="n">ref_fasta</span>
        <span class="c1"># ----  do not ever name these directly ---- #</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pe_map_bam</span> <span class="o">=</span> <span class="n">pe_map_bam</span>  <span class="c1"># all reads from pe mapping</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">s_map_bam</span> <span class="o">=</span> <span class="n">s_map_bam</span>  <span class="c1"># all reads from singltons mapping</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mapping_prefix</span> <span class="o">=</span> <span class="n">mapping_prefix</span>  <span class="c1"># set dynamically</span>
        <span class="c1"># added to makes filtering with pysam easier</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mapped_bam_unfiltered</span> <span class="o">=</span> <span class="n">mapped_bam_unfiltered</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mapped_sam</span> <span class="o">=</span> <span class="n">mapped_sam</span>  <span class="c1"># mapped reads only, sam</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mapped_bam</span> <span class="o">=</span> <span class="n">mapped_bam</span>  <span class="c1"># mapped reads only, bam</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mapped_ids_txt</span> <span class="o">=</span> <span class="n">mapped_ids_txt</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">unmapped_sam</span> <span class="o">=</span> <span class="n">unmapped_sam</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">unmapped_bam</span> <span class="o">=</span> <span class="n">unmapped_bam</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sorted_mapped_bam</span> <span class="o">=</span> <span class="n">sorted_mapped_bam</span>  <span class="c1"># used with intial mapping</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mapped_ngsLib</span> <span class="o">=</span> <span class="n">mapped_ngsLib</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">unmapped_ngsLib</span> <span class="o">=</span> <span class="n">unmapped_ngsLib</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">assembled_contig</span> <span class="o">=</span> <span class="n">assembled_contig</span>
        <span class="c1">#</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">check_mands</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">make_mapping_subdir</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">make_assembly_subdir</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name_bams_and_sams</span><span class="p">()</span>

<div class="viewcode-block" id="LociMapping.check_mands"><a class="viewcode-back" href="../riboSeed.html#riboSeed.LociMapping.check_mands">[docs]</a>    <span class="k">def</span> <span class="nf">check_mands</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; checks that all mandatory arguments are not none</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">mandatory</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">iteration</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">mapping_subdir</span><span class="p">]</span>
        <span class="k">if</span> <span class="kc">None</span> <span class="ow">in</span> <span class="n">mandatory</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;mapping ob must be instantiated with name, &quot;</span>
                             <span class="s2">&quot;iteration, mapping_subdir name&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="LociMapping.make_mapping_subdir"><a class="viewcode-back" href="../riboSeed.html#riboSeed.LociMapping.make_mapping_subdir">[docs]</a>    <span class="k">def</span> <span class="nf">make_mapping_subdir</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;make a subdirectory for mapping results &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mapping_subdir</span><span class="p">):</span>
            <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">mapping_subdir</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span></div>

<div class="viewcode-block" id="LociMapping.name_bams_and_sams"><a class="viewcode-back" href="../riboSeed.html#riboSeed.LociMapping.name_bams_and_sams">[docs]</a>    <span class="k">def</span> <span class="nf">name_bams_and_sams</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; make a prefix and use it to name the future output files &quot;&quot;&quot;</span>
        <span class="n">mapping_prefix</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mapping_subdir</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pe_map_bam</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">mapping_prefix</span> <span class="o">+</span> <span class="s2">&quot;_pe.bam&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">s_map_bam</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">mapping_prefix</span> <span class="o">+</span> <span class="s2">&quot;_s.bam&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mapped_bam_unfiltered</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">mapping_prefix</span> <span class="o">+</span> <span class="s2">&quot;_unfiltered.bam&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mapped_bam</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">mapping_prefix</span> <span class="o">+</span> <span class="s2">&quot;.bam&quot;</span><span class="p">)</span>
        <span class="c1"># self.unampped_bam = str(mapping_prefix + &quot;unmapped.bam&quot;)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sorted_mapped_bam</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">mapping_prefix</span> <span class="o">+</span> <span class="s2">&quot;_sorted.bam&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mapped_sam</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">mapping_prefix</span> <span class="o">+</span> <span class="s2">&quot;.sam&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">unmapped_sam</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">mapping_prefix</span> <span class="o">+</span> <span class="s2">&quot;_unmapped.sam&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mapped_ids_txt</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">mapping_prefix</span> <span class="o">+</span> <span class="s2">&quot;_mapped.txt&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="LociMapping.make_assembly_subdir"><a class="viewcode-back" href="../riboSeed.html#riboSeed.LociMapping.make_assembly_subdir">[docs]</a>    <span class="k">def</span> <span class="nf">make_assembly_subdir</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; make a subdirectory for assembly if it is needed &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">assembly_subdir_needed</span><span class="p">:</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">assembly_subdir</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">assembly_subdir</span><span class="p">):</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">assembly_subdir</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span></div></div>


<div class="viewcode-block" id="Exes"><a class="viewcode-back" href="../riboSeed.html#riboSeed.Exes">[docs]</a><span class="k">class</span> <span class="nc">Exes</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    given the amount of system tools that riboSeed requires, this object</span>
<span class="sd">    holds the paths to the executables after expanding the user-supplied</span>
<span class="sd">    path and verifying with shutil.which that the executable is availible</span>
<span class="sd">    to the program.</span>


<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">samtools</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="n">spades</span><span class="p">,</span> <span class="n">quast</span><span class="p">,</span> <span class="n">python2_7</span><span class="p">,</span>
                 <span class="n">smalt</span><span class="p">,</span> <span class="n">bwa</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">mapper</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">samtools</span> <span class="o">=</span> <span class="n">samtools</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">method</span> <span class="o">=</span> <span class="n">method</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">mapper</span> <span class="o">=</span> <span class="n">mapper</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">spades</span> <span class="o">=</span> <span class="n">spades</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">quast</span> <span class="o">=</span> <span class="n">quast</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">smalt</span> <span class="o">=</span> <span class="n">smalt</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">bwa</span> <span class="o">=</span> <span class="n">bwa</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">python2_7</span> <span class="o">=</span> <span class="n">python2_7</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">check</span> <span class="o">=</span> <span class="n">check</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">check_mands</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">set_mapper</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">check_expand_exes</span><span class="p">()</span>

<div class="viewcode-block" id="Exes.check_mands"><a class="viewcode-back" href="../riboSeed.html#riboSeed.Exes.check_mands">[docs]</a>    <span class="k">def</span> <span class="nf">check_mands</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; checks that all mandatory arguments are not none</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">mandatory</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">spades</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">quast</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">method</span><span class="p">,</span>
                     <span class="bp">self</span><span class="o">.</span><span class="n">samtools</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">python2_7</span><span class="p">]</span>
        <span class="k">assert</span> <span class="kc">None</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">mandatory</span><span class="p">,</span> \
            <span class="s2">&quot;must instantiate with samtools, spades, method, python2_7, quast!&quot;</span></div>

<div class="viewcode-block" id="Exes.set_mapper"><a class="viewcode-back" href="../riboSeed.html#riboSeed.Exes.set_mapper">[docs]</a>    <span class="k">def</span> <span class="nf">set_mapper</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Exes.mapper attribute is set here to avoid further</span>
<span class="sd">        &quot;if method ==&#39;smalt&#39; clauses later.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;smalt&quot;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mapper</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">smalt</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;bwa&quot;</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">mapper</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">bwa</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Mapping method not found!&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="Exes.check_expand_exes"><a class="viewcode-back" href="../riboSeed.html#riboSeed.Exes.check_expand_exes">[docs]</a>    <span class="k">def</span> <span class="nf">check_expand_exes</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot; for each executable, expand wildcards and use shutil.which</span>
<span class="sd">        to get full path to executable.  If not found, throw an error</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">check</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">exe</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;mapper&quot;</span><span class="p">,</span> <span class="s2">&quot;samtools&quot;</span><span class="p">,</span> <span class="s2">&quot;spades&quot;</span><span class="p">,</span>
                        <span class="s2">&quot;quast&quot;</span><span class="p">,</span> <span class="s2">&quot;python2_7&quot;</span><span class="p">,</span> <span class="s2">&quot;mapper&quot;</span><span class="p">]:</span>
                <span class="n">exe_groomed</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">exe</span><span class="p">))</span>
                <span class="n">exe_groomed</span> <span class="o">=</span> <span class="n">shutil</span><span class="o">.</span><span class="n">which</span><span class="p">(</span><span class="n">exe_groomed</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">exe_groomed</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> not found in PATH!&quot;</span> <span class="o">%</span> <span class="n">exe</span><span class="p">)</span>
                <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">exe</span><span class="p">,</span> <span class="n">exe_groomed</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span></div></div>


<span class="c1"># --------------------------- methods --------------------------- #</span>


<div class="viewcode-block" id="get_args"><a class="viewcode-back" href="../riboSeed.html#riboSeed.get_args">[docs]</a><span class="k">def</span> <span class="nf">get_args</span><span class="p">():</span>  <span class="c1"># pragma: no cover</span>
    <span class="sd">&quot;&quot;&quot;#TODO:     for cli mods:</span>
<span class="sd">    http://stackoverflow.com/questions/18025646/</span>
<span class="sd">         python-argparse-conditional-requirements</span>
<span class="sd">    make this able to handle different library types such as two unpaired runs</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Given cluster file of rDNA regions from riboSelect and &quot;</span> <span class="o">+</span>
        <span class="s2">&quot;either paired-end or single-end reads, assembles the mapped reads &quot;</span> <span class="o">+</span>
        <span class="s2">&quot;into pseduocontig &#39;seeds&#39;, and uses those with the reads to run&quot;</span> <span class="o">+</span>
        <span class="s2">&quot;de fere novo and de novo assembly with SPAdes&quot;</span><span class="p">,</span>
        <span class="n">add_help</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>  <span class="c1"># to allow for custom help</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;clustered_loci_txt&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;output from riboSelect&quot;</span><span class="p">)</span>

    <span class="c1"># taking a hint from http://stackoverflow.com/questions/24180527</span>
    <span class="n">requiredNamed</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;required named arguments&#39;</span><span class="p">)</span>
    <span class="n">requiredNamed</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-r&quot;</span><span class="p">,</span> <span class="s2">&quot;--reference_genbank&quot;</span><span class="p">,</span>
                               <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;reference_genbank&#39;</span><span class="p">,</span>
                               <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
                               <span class="n">help</span><span class="o">=</span><span class="s2">&quot;genbank reference, used to estimate &quot;</span> <span class="o">+</span>
                               <span class="s2">&quot;insert sizes, and compare with QUAST&quot;</span><span class="p">,</span>
                               <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">requiredNamed</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-o&quot;</span><span class="p">,</span> <span class="s2">&quot;--output&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;output&#39;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
                               <span class="n">help</span><span class="o">=</span><span class="s2">&quot;output directory; &quot;</span> <span class="o">+</span>
                               <span class="s2">&quot;default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">(),</span>
                               <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># had to make this faux &quot;optional&quot; parse so that the named required ones</span>
    <span class="c1"># above get listed first</span>
    <span class="n">optional</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">add_argument_group</span><span class="p">(</span><span class="s1">&#39;optional arguments&#39;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-F&quot;</span><span class="p">,</span> <span class="s2">&quot;--fastq1&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;fastq1&#39;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;forward fastq reads, can be compressed&quot;</span><span class="p">,</span>
                          <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-R&quot;</span><span class="p">,</span> <span class="s2">&quot;--fastq2&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;fastq2&#39;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;reverse fastq reads, can be compressed&quot;</span><span class="p">,</span>
                          <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-S1&quot;</span><span class="p">,</span> <span class="s2">&quot;--fastq_single1&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;fastqS1&#39;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;single fastq reads&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
    <span class="c1"># optional.add_argument(&quot;-S2&quot;, &quot;--fastq_single2&quot;, dest=&#39;fastqS2&#39;,</span>
    <span class="c1">#                       action=&quot;store&quot;,</span>
    <span class="c1">#                       help=&quot;single fastq reads&quot;, type=str, default=None)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-n&quot;</span><span class="p">,</span> <span class="s2">&quot;--experiment_name&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;exp_name&#39;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;prefix for results files; &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="s2">&quot;riboSeed&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-l&quot;</span><span class="p">,</span> <span class="s2">&quot;--flanking_length&quot;</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;length of flanking regions, in bp; &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;flanking&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-m&quot;</span><span class="p">,</span> <span class="s2">&quot;--method_for_map&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;method&#39;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span> <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;smalt&quot;</span><span class="p">,</span> <span class="s2">&quot;bwa&quot;</span><span class="p">],</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;available mappers: smalt and bwa; &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="s1">&#39;bwa&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-c&quot;</span><span class="p">,</span> <span class="s2">&quot;--cores&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;cores&#39;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;cores used&quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;; default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--memory&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;memory&#39;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="mi">8</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;cores for multiprocessing&quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;; default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-k&quot;</span><span class="p">,</span> <span class="s2">&quot;--kmers&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;kmers&#39;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="s2">&quot;21,33,55,77,99,127&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;kmers used for final assembly&quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;, separated by commas such as&quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;21,33,55,77,99,127 . Can be set to &#39;auto&#39;, where &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;SPAdes chooses.  We ensure kmers are not &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;too big or too close to read length&quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;; default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-p&quot;</span><span class="p">,</span> <span class="s2">&quot;--pre_kmers&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;pre_kmers&#39;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="s2">&quot;21,33,55,77,99&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;kmers used during seeding assemblies, &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;separated bt commas&quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;; default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-s&quot;</span><span class="p">,</span> <span class="s2">&quot;--score_min&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;score_min&#39;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;If using smalt, this sets the &#39;-m&#39; param; &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;default with smalt is inferred from &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;read length. If using BWA, reads mapping with AS&quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;score lower than this will be rejected&quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;; default with BWA is half of read length&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-a&quot;</span><span class="p">,</span> <span class="s2">&quot;--min_assembly_len&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;min_assembly_len&#39;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="mi">6000</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;if initial SPAdes assembly largest contig &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;is not at least as long as --min_assembly_len, &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;reject. Set this to the length of the seed &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;sequence; if it is not achieved, seeding across &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;regions will likely fail; default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--include_shorts&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;include_short_contigs&#39;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;if assembled contig is smaller than  &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;--min_assembly_len, contig will still be included&quot;</span> <span class="o">+</span>
                          <span class="s2">&quot; in assembly; default: inferred&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--subtract&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;subtract&#39;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;if --subtract reads already used in previous&quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;round of subassembly will not be included in &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;subsequent rounds.  This can lead to problems &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;with multiple mapping and inflated coverage.&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--linear&quot;</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;if genome is known to not be circular and &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;a region of interest (including flanking bits) &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;extends past chromosome end, this extends the &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;seqence past chromosome origin forward by &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;--padding; &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;linear&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-d&quot;</span><span class="p">,</span> <span class="s2">&quot;--min_flank_depth&quot;</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;a subassembly will not be performed if this &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;minimum depth is not achieved on both the 3&#39; and&quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;5&#39; end of the pseudocontig. &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;min_flank_depth&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
    <span class="c1"># optional.add_argument(&quot;--padding&quot;, dest=&#39;padding&#39;, action=&quot;store&quot;,</span>
    <span class="c1">#                       default=5000, type=int,</span>
    <span class="c1">#                       help=&quot;if treating as circular, this controls the &quot; +</span>
    <span class="c1">#                       &quot;length of sequence added to the 5&#39; and 3&#39; ends &quot; +</span>
    <span class="c1">#                       &quot;to allow for selecting regions that cross the &quot; +</span>
    <span class="c1">#                       &quot;chromosome&#39;s origin; default: %(default)s&quot;)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--ref_as_contig&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;ref_as_contig&#39;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">str</span><span class="p">,</span>
                          <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;trusted&quot;</span><span class="p">,</span> <span class="s2">&quot;untrusted&quot;</span><span class="p">],</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;if &#39;trusted&#39;, SPAdes will  use the seed &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;sequences as a --trusted-contig; if &#39;untrusted&#39;, &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;SPAdes will treat as --untrusted-contig. if &#39;&#39;, &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;seeds will not be used during assembly. &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;See SPAdes docs; default: if mapping &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;percentage over 80</span><span class="si">%%</span><span class="s2">: &#39;trusted&#39;, else &#39;untrusted&#39;&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--clean_temps&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;clean_temps&#39;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;if --clean_temps, mapping files will be &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;removed once they are no no longer needed during &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;the mapping iterations to save space; &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--skip_control&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;skip_control&#39;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;if --skip_control, no de novo &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;assembly will be done; default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-i&quot;</span><span class="p">,</span> <span class="s2">&quot;--iterations&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;iterations&#39;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;if iterations&gt;1, multiple seedings will &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;occur after subassembly of seed regions; &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;if setting --target_len, seedings will continue &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;until --iterations are completed or --target_len&quot;</span>
                          <span class="s2">&quot; is matched or exceeded; &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-v&quot;</span><span class="p">,</span> <span class="s2">&quot;--verbosity&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;verbosity&#39;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span> <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">],</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Logger writes debug to file in output dir; &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;this sets verbosity level sent to stderr. &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot; 1 = debug(), 2 = info(), 3 = warning(), &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;4 = error() and 5 = critical(); &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--target_len&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;target_len&#39;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">float</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;if set, iterations will continue until &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;contigs reach this length, or max iterations (&quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;set by --iterations) have been completed. Set as &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;fraction of original seed length by giving a &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;decimal between 0 and 5, or set as an absolute &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;number of base pairs by giving an integer greater&quot;</span> <span class="o">+</span>
                          <span class="s2">&quot; than 50. Not used by default&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-t&quot;</span><span class="p">,</span> <span class="s2">&quot;--threads&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;threads&#39;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="nb">int</span><span class="p">,</span>
                          <span class="n">choices</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;if your cores are hyperthreaded, set number &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;threads to the number of threads per processer.&quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;If unsure, see &#39;cat /proc/cpuinfo&#39; under &#39;cpu &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;cores&#39;, or &#39;lscpu&#39; under &#39;Thread(s) per core&#39;.&quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-z&quot;</span><span class="p">,</span> <span class="s2">&quot;--serialize&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;serialize&#39;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;if --serialize, runs seeding and assembly &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;without multiprocessing. This is recommended for &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;machines with less than 8GB RAM: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--smalt_scoring&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;smalt_scoring&#39;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="s2">&quot;match=1,subst=-4,gapopen=-4,gapext=-3&quot;</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;if mapping with SMALT, &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;submit custom smalt scoring via smalt -S &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;scorespec option; default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--mapper_args&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;mapper_args&#39;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span>
                          <span class="n">default</span><span class="o">=</span><span class="s2">&quot;-L 0,0 -U 0 -a&quot;</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;submit custom parameters to mapper. &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;And by mapper, I mean bwa, cause we dont support &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;this option for SMALT, sorry. &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;This requires knowledge of your chosen mapper&#39;s &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;optional arguments. Proceed with caution!  &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="c1"># # had to make this explicitly to call it a faux optional arg</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;-h&quot;</span><span class="p">,</span> <span class="s2">&quot;--help&quot;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;help&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="n">argparse</span><span class="o">.</span><span class="n">SUPPRESS</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Displays this help message&quot;</span><span class="p">)</span>

    <span class="c1"># # TODO  Make these check a config file</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--spades_exe&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;spades_exe&quot;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;spades.py&quot;</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Path to SPAdes executable; &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--samtools_exe&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;samtools_exe&quot;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;samtools&quot;</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Path to samtools executable; &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--smalt_exe&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;smalt_exe&quot;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;smalt&quot;</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Path to smalt executable;&quot;</span> <span class="o">+</span>
                          <span class="s2">&quot; default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--bwa_exe&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;bwa_exe&quot;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;bwa&quot;</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Path to BWA executable;&quot;</span> <span class="o">+</span>
                          <span class="s2">&quot; default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--quast_exe&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;quast_exe&quot;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;quast.py&quot;</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Path to quast executable; &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--python2_7_exe&quot;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s2">&quot;python2_7_exe&quot;</span><span class="p">,</span>
                          <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store&quot;</span><span class="p">,</span> <span class="n">default</span><span class="o">=</span><span class="s2">&quot;python2.7&quot;</span><span class="p">,</span>
                          <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Path to python2.7 executable, cause &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;QUAST won&#39;t run on python3. default: </span><span class="si">%(default)s</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">optional</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;--version&#39;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s1">&#39;version&#39;</span><span class="p">,</span>
                          <span class="n">version</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%(prog)s</span><span class="s1"> </span><span class="si">{version}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                              <span class="n">version</span><span class="o">=</span><span class="n">__version__</span><span class="p">))</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">args</span></div>


<div class="viewcode-block" id="last_exception"><a class="viewcode-back" href="../riboSeed.html#riboSeed.last_exception">[docs]</a><span class="k">def</span> <span class="nf">last_exception</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot; Returns last exception as a string, or use in logging.</span>
<span class="sd">    stolen verbatim from pyani</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">exc_type</span><span class="p">,</span> <span class="n">exc_value</span><span class="p">,</span> <span class="n">exc_traceback</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">exc_info</span><span class="p">()</span>
    <span class="k">return</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">traceback</span><span class="o">.</span><span class="n">format_exception</span><span class="p">(</span><span class="n">exc_type</span><span class="p">,</span> <span class="n">exc_value</span><span class="p">,</span>
                                              <span class="n">exc_traceback</span><span class="p">))</span></div>


<div class="viewcode-block" id="get_rec_from_generator"><a class="viewcode-back" href="../riboSeed.html#riboSeed.get_rec_from_generator">[docs]</a><span class="k">def</span> <span class="nf">get_rec_from_generator</span><span class="p">(</span><span class="n">recordID</span><span class="p">,</span> <span class="n">gen</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; given a record ID and and SeqIO generator return sequence of</span>
<span class="sd">    genbank record that has the loci, and call method to refresh generator</span>
<span class="sd">    If on different sequences, return error</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">gen</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">recordID</span> <span class="o">==</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">method</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">method</span><span class="p">()</span>
            <span class="k">return</span> <span class="n">record</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span>
    <span class="c1"># if none found, raise error</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;no record found matching record id </span><span class="si">%s</span><span class="s2">!&quot;</span> <span class="o">%</span> <span class="n">recordID</span><span class="p">)</span></div>


<div class="viewcode-block" id="get_smalt_full_install_cmds"><a class="viewcode-back" href="../riboSeed.html#riboSeed.get_smalt_full_install_cmds">[docs]</a><span class="k">def</span> <span class="nf">get_smalt_full_install_cmds</span><span class="p">(</span><span class="n">smalt_exe</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; TODO replace this with swg tests for bambamc installation</span>
<span class="sd">    In the meantime, this looks for the files included with riboSeed</span>
<span class="sd">    (a bam file, reference, index, and fastq file), and generates the cmds</span>
<span class="sd">    to run a little test mapping</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">smalttestdir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="vm">__file__</span><span class="p">)),</span>
                                <span class="s2">&quot;sample_data&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;smalt_test&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">logger</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Must Use Logging&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;looking for smalt test dir: </span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">smalttestdir</span><span class="p">))</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">smalttestdir</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span>
            <span class="s2">&quot;Cannot find smalt_test dir containing &quot;</span> <span class="o">+</span>
            <span class="s2">&quot;files to verify bambamc install! It should here: </span><span class="se">\n</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">smalttestdir</span><span class="p">)</span>
    <span class="n">ref</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">smalttestdir</span><span class="p">,</span> <span class="s2">&quot;ref_to_test_bambamc.fasta&quot;</span><span class="p">)</span>
    <span class="n">index</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">smalttestdir</span><span class="p">,</span> <span class="s2">&quot;test_index&quot;</span><span class="p">)</span>
    <span class="n">test_bam</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">smalttestdir</span><span class="p">,</span> <span class="s2">&quot;test_mapping.bam&quot;</span><span class="p">)</span>
    <span class="n">test_reads</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">smalttestdir</span><span class="p">,</span> <span class="s2">&quot;reads_to_test_bambamc.fastq&quot;</span><span class="p">)</span>
    <span class="n">testindexcmd</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> index </span><span class="si">{1}</span><span class="s2"> </span><span class="si">{2}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">smalt_exe</span><span class="p">,</span> <span class="n">index</span><span class="p">,</span> <span class="n">ref</span><span class="p">))</span>
    <span class="n">testmapcmd</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> map -f bam -o </span><span class="si">{1}</span><span class="s2"> </span><span class="si">{2}</span><span class="s2"> </span><span class="si">{3}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">smalt_exe</span><span class="p">,</span>
                                                            <span class="n">test_bam</span><span class="p">,</span>
                                                            <span class="n">index</span><span class="p">,</span>
                                                            <span class="n">test_reads</span><span class="p">))</span>
    <span class="k">return</span><span class="p">([</span><span class="n">testindexcmd</span><span class="p">,</span> <span class="n">testmapcmd</span><span class="p">])</span></div>


<div class="viewcode-block" id="test_smalt_bam_install"><a class="viewcode-back" href="../riboSeed.html#riboSeed.test_smalt_bam_install">[docs]</a><span class="k">def</span> <span class="nf">test_smalt_bam_install</span><span class="p">(</span>
        <span class="n">cmds</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>  <span class="c1"># pragma: no cover, cause pragma, no care</span>
    <span class="sd">&quot;&quot;&quot; using test data tha tcomes with package, ensure that</span>
<span class="sd">    the bambamc library was properly installed with SMALT instaltation</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">logger</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;must use logger&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;testing instalation of SMALT and bambamc&quot;</span><span class="p">)</span>
    <span class="n">smalttestdir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="vm">__file__</span><span class="p">)),</span>
                                <span class="s2">&quot;sample_data&quot;</span><span class="p">,</span>
                                <span class="s2">&quot;smalt_test&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">test_index</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">smalttestdir</span><span class="p">,</span> <span class="s2">&quot;test_index&quot;</span><span class="p">)</span>
    <span class="n">test_bam</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">smalttestdir</span><span class="p">,</span> <span class="s2">&quot;test_mapping.bam&quot;</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">cmds</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="n">i</span><span class="p">],</span>
                           <span class="n">shell</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">platform</span> <span class="o">!=</span> <span class="s2">&quot;win32&quot;</span><span class="p">,</span>
                           <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                           <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                           <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                <span class="s2">&quot;Error running test to check bambamc lib is &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;installed! See github.com/gt1/bambamc &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;and the smalt install guide for more details.&quot;</span> <span class="o">+</span>
                <span class="s2">&quot;https://sourceforge.net/projects/smalt/files/&quot;</span><span class="p">)</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># remove the temp files</span>
    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="n">test_bam</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">test_index</span> <span class="o">+</span> <span class="s2">&quot;.sma&quot;</span><span class="p">))</span>
    <span class="n">os</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">test_index</span> <span class="o">+</span> <span class="s2">&quot;.smi&quot;</span><span class="p">))</span></div>


<div class="viewcode-block" id="estimate_distances_smalt"><a class="viewcode-back" href="../riboSeed.html#riboSeed.estimate_distances_smalt">[docs]</a><span class="k">def</span> <span class="nf">estimate_distances_smalt</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="n">smalt_exe</span><span class="p">,</span> <span class="n">ref_genome</span><span class="p">,</span>
                             <span class="n">fastq1</span><span class="p">,</span> <span class="n">fastq2</span><span class="p">,</span> <span class="n">cores</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Given fastq pair and a reference, returns path to distance estimations</span>
<span class="sd">    used by smalt to help later with mapping. if one already exists,</span>
<span class="sd">    return path to it.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">cores</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">cores</span> <span class="o">=</span> <span class="n">multiprocessing</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">outfile</span><span class="p">):</span>
        <span class="c1"># Index reference for sampling to get PE distances</span>
        <span class="k">if</span> <span class="n">logger</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Estimating insert distances with SMALT&quot;</span><span class="p">)</span>
        <span class="c1"># index with default params for genome-sized sequence</span>
        <span class="n">refindex_cmd</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">smalt_exe</span> <span class="o">+</span> <span class="s2">&quot; index -k </span><span class="si">{0}</span><span class="s2"> -s </span><span class="si">{1}</span><span class="s2"> </span><span class="si">{2}</span><span class="s2"> &quot;</span> <span class="o">+</span>
                           <span class="s2">&quot;</span><span class="si">{3}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="n">outfile</span><span class="p">,</span> <span class="n">ref_genome</span><span class="p">)</span>
        <span class="n">refsample_cmd</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">smalt_exe</span> <span class="o">+</span> <span class="s2">&quot; sample -n </span><span class="si">{0}</span><span class="s2"> -o </span><span class="si">{1}</span><span class="s2"> </span><span class="si">{2}</span><span class="s2"> </span><span class="si">{3}</span><span class="s2"> &quot;</span> <span class="o">+</span>
                            <span class="s2">&quot;</span><span class="si">{4}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cores</span><span class="p">,</span>
                                          <span class="n">outfile</span><span class="p">,</span>
                                          <span class="n">outfile</span><span class="p">,</span>
                                          <span class="n">fastq1</span><span class="p">,</span>
                                          <span class="n">fastq2</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">logger</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sampling and indexing </span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">ref_genome</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">cmd</span> <span class="ow">in</span> <span class="p">[</span><span class="n">refindex_cmd</span><span class="p">,</span> <span class="n">refsample_cmd</span><span class="p">]:</span>
            <span class="k">if</span> <span class="n">logger</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2"> command:</span><span class="se">\n\t</span><span class="s2"> </span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cmd</span><span class="p">))</span>
            <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span>
                           <span class="n">shell</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">platform</span> <span class="o">!=</span> <span class="s2">&quot;win32&quot;</span><span class="p">,</span>
                           <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                           <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                           <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">logger</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;using existing reference file&quot;</span><span class="p">)</span>
        <span class="k">pass</span>
    <span class="k">return</span> <span class="n">outfile</span></div>


<div class="viewcode-block" id="check_fastqs_len_equal"><a class="viewcode-back" href="../riboSeed.html#riboSeed.check_fastqs_len_equal">[docs]</a><span class="k">def</span> <span class="nf">check_fastqs_len_equal</span><span class="p">(</span><span class="n">file1</span><span class="p">,</span> <span class="n">file2</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; using file_len from pyutilsnrw, check that the fastqs contain</span>
<span class="sd">    the same number of lines, ie tat the pairing looks proper.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">file_len</span><span class="p">(</span><span class="n">file1</span><span class="p">)</span> <span class="o">!=</span> <span class="n">file_len</span><span class="p">(</span><span class="n">file2</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Input Fastq&#39;s are of unequal length! Try &quot;</span> <span class="o">+</span>
            <span class="s2">&quot;fixing with this script: &quot;</span> <span class="o">+</span>
            <span class="s2">&quot;github.com/enormandeau/Scripts/fastqCombinePairedEnd.py&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="nonify_empty_lib_files"><a class="viewcode-back" href="../riboSeed.html#riboSeed.nonify_empty_lib_files">[docs]</a><span class="k">def</span> <span class="nf">nonify_empty_lib_files</span><span class="p">(</span><span class="n">ngsLib</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="c1"># sometimes, if no singletons are found, we get an empty file.</span>
    <span class="c1">#  this shoudl weed out any empty read files before mapping, etc</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;checking for empty read files&quot;</span><span class="p">)</span>
    <span class="n">EMPTIES</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;readF&quot;</span><span class="p">,</span> <span class="s2">&quot;readR&quot;</span><span class="p">,</span> <span class="s2">&quot;readS0&quot;</span><span class="p">]:</span>
        <span class="c1"># ignore if lib is None, as those wont be used anyway</span>
        <span class="k">if</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">ngsLib</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> is set to None, and will be ignored&quot;</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>
            <span class="n">EMPTIES</span> <span class="o">=</span> <span class="n">EMPTIES</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="k">continue</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="n">ngsLib</span><span class="p">,</span> <span class="n">f</span><span class="p">)):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;read file </span><span class="si">%s</span><span class="s2"> not found and can not be used &quot;</span> <span class="o">+</span>
                           <span class="s2">&quot;for mapping!&quot;</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>
            <span class="n">EMPTIES</span> <span class="o">=</span> <span class="n">EMPTIES</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="c1"># set to None so mapper will ignore</span>
            <span class="nb">setattr</span><span class="p">(</span><span class="n">ngsLib</span><span class="p">,</span> <span class="n">f</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
            <span class="k">continue</span>
        <span class="c1"># if lib is not none but file is of size 0</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;size of </span><span class="si">%s</span><span class="s2">: </span><span class="si">%f</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">ngsLib</span><span class="p">,</span> <span class="n">f</span><span class="p">),</span>
                     <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">getsize</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="n">ngsLib</span><span class="p">,</span> <span class="n">f</span><span class="p">)))</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">getsize</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="n">ngsLib</span><span class="p">,</span> <span class="n">f</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;read file </span><span class="si">%s</span><span class="s2"> is empty and will not be used &quot;</span> <span class="o">+</span>
                           <span class="s2">&quot;for mapping!&quot;</span><span class="p">,</span> <span class="n">f</span><span class="p">)</span>
            <span class="n">EMPTIES</span> <span class="o">=</span> <span class="n">EMPTIES</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="c1"># set to None so mapper will ignore</span>
            <span class="nb">setattr</span><span class="p">(</span><span class="n">ngsLib</span><span class="p">,</span> <span class="n">f</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">EMPTIES</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;None of the read files hold data!&quot;</span><span class="p">)</span></div>

<span class="c1"># MapperParams = namedtuple(</span>
<span class="c1">#     &quot;MapperParams&quot;,</span>
<span class="c1">#     &quot;cores samtools_exe mapper_exe ignore_singletons &quot; +</span>
<span class="c1">#     &quot;score_minimum single_lib scoring step k smalt_scoring&quot;)</span>


<div class="viewcode-block" id="map_to_genome_ref_smalt"><a class="viewcode-back" href="../riboSeed.html#riboSeed.map_to_genome_ref_smalt">[docs]</a><span class="k">def</span> <span class="nf">map_to_genome_ref_smalt</span><span class="p">(</span><span class="n">mapping_ob</span><span class="p">,</span> <span class="n">ngsLib</span><span class="p">,</span> <span class="n">cores</span><span class="p">,</span>
                            <span class="n">samtools_exe</span><span class="p">,</span> <span class="n">smalt_exe</span><span class="p">,</span>
                            <span class="n">genome_fasta</span><span class="p">,</span>
                            <span class="n">score_minimum</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                            <span class="n">scoring</span><span class="o">=</span><span class="s2">&quot;match=1,subst=-4,gapopen=-4,gapext=-3&quot;</span><span class="p">,</span>
                            <span class="n">step</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;run smalt based on pased args</span>
<span class="sd">    #TODO rework this to read libtype of ngslib object</span>
<span class="sd">    requires at least paired end input, but can handle an additional library</span>
<span class="sd">    of singleton reads. Will not work on just singletons</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Mapping reads to reference genome with SMALT&quot;</span><span class="p">)</span>
    <span class="c1"># check min score</span>
    <span class="k">assert</span> <span class="n">score_minimum</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;must sassign score outside map function!&quot;</span>
    <span class="n">score_min</span> <span class="o">=</span> <span class="n">score_minimum</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;using a score min of &quot;</span> <span class="o">+</span>
                     <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">score_min</span><span class="p">))</span>
    <span class="c1"># index the reference</span>
    <span class="n">cmdindex</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> index -k </span><span class="si">{1}</span><span class="s2"> -s </span><span class="si">{2}</span><span class="s2"> </span><span class="si">{3}</span><span class="s2"> </span><span class="si">{3}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">smalt_exe</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="n">step</span><span class="p">,</span> <span class="n">genome_fasta</span><span class="p">)</span>
    <span class="c1"># map paired end reads to reference index</span>
    <span class="n">smaltcommands</span> <span class="o">=</span> <span class="p">[</span><span class="n">cmdindex</span><span class="p">]</span>
    <span class="k">if</span> <span class="s2">&quot;pe&quot;</span> <span class="ow">in</span> <span class="n">ngsLib</span><span class="o">.</span><span class="n">libtype</span><span class="p">:</span>
        <span class="n">cmdmap</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{0}</span><span class="s1"> map -l pe -S </span><span class="si">{1}</span><span class="s1"> &#39;</span> <span class="o">+</span>
                     <span class="s1">&#39;-m </span><span class="si">{2}</span><span class="s1"> -n </span><span class="si">{3}</span><span class="s1"> -g </span><span class="si">{4}</span><span class="s1"> -f bam -o </span><span class="si">{5}</span><span class="s1"> </span><span class="si">{6}</span><span class="s1"> </span><span class="si">{7}</span><span class="s1"> &#39;</span> <span class="o">+</span>
                     <span class="s1">&#39;</span><span class="si">{8}</span><span class="s1">&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">smalt_exe</span><span class="p">,</span> <span class="n">scoring</span><span class="p">,</span>
                                   <span class="n">score_min</span><span class="p">,</span> <span class="n">cores</span><span class="p">,</span> <span class="n">ngsLib</span><span class="o">.</span><span class="n">smalt_dist_path</span><span class="p">,</span>
                                   <span class="n">mapping_ob</span><span class="o">.</span><span class="n">pe_map_bam</span><span class="p">,</span> <span class="n">genome_fasta</span><span class="p">,</span>
                                   <span class="n">ngsLib</span><span class="o">.</span><span class="n">readF</span><span class="p">,</span>
                                   <span class="n">ngsLib</span><span class="o">.</span><span class="n">readR</span><span class="p">)</span>
        <span class="n">smaltcommands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cmdmap</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">mapping_ob</span><span class="o">.</span><span class="n">pe_map_bam</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">tempfile</span><span class="p">:</span>
            <span class="n">tempfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;@HD riboseed_dummy_file&quot;</span><span class="p">)</span>
        <span class="k">pass</span>
    <span class="c1"># if singletons are present, map those too.  Index is already made</span>
    <span class="k">if</span> <span class="n">ngsLib</span><span class="o">.</span><span class="n">readS0</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>  <span class="c1"># and not ignore_singletons:</span>
        <span class="c1"># because erros are thrown if there is no file, this</span>
        <span class="n">cmdmapS</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span>
            <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> map -S </span><span class="si">{1}</span><span class="s2"> -m </span><span class="si">{2}</span><span class="s2"> -n </span><span class="si">{3}</span><span class="s2"> -g </span><span class="si">{4}</span><span class="s2"> -f bam -o </span><span class="si">{5}</span><span class="s2"> &quot;</span> <span class="o">+</span>
            <span class="s2">&quot;</span><span class="si">{6}</span><span class="s2"> </span><span class="si">{7}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">smalt_exe</span><span class="p">,</span> <span class="n">scoring</span><span class="p">,</span> <span class="n">score_min</span><span class="p">,</span> <span class="n">cores</span><span class="p">,</span>
                              <span class="n">ngsLib</span><span class="o">.</span><span class="n">smalt_dist_path</span><span class="p">,</span> <span class="n">mapping_ob</span><span class="o">.</span><span class="n">s_map_bam</span><span class="p">,</span>
                              <span class="n">genome_fasta</span><span class="p">,</span> <span class="n">ngsLib</span><span class="o">.</span><span class="n">readS0</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">mapping_ob</span><span class="o">.</span><span class="n">s_map_bam</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">tempfile</span><span class="p">:</span>
            <span class="n">tempfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;@HD riboseed_dummy_file&quot;</span><span class="p">)</span>
        <span class="c1"># merge together the singleton and pe reads</span>
        <span class="n">cmdmergeS</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{0}</span><span class="s1"> merge -f </span><span class="si">{3}</span><span class="s1"> </span><span class="si">{1}</span><span class="s1"> </span><span class="si">{2}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">samtools_exe</span><span class="p">,</span> <span class="n">mapping_ob</span><span class="o">.</span><span class="n">pe_map_bam</span><span class="p">,</span>
            <span class="n">mapping_ob</span><span class="o">.</span><span class="n">s_map_bam</span><span class="p">,</span> <span class="n">mapping_ob</span><span class="o">.</span><span class="n">mapped_bam</span><span class="p">)</span>
        <span class="n">smaltcommands</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">cmdmapS</span><span class="p">,</span> <span class="n">cmdmergeS</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># if not already none, set to None when ignoring singleton</span>
        <span class="n">ngsLib</span><span class="o">.</span><span class="n">readS0</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="c1"># &#39;merge&#39;, but reallt just converts</span>
        <span class="n">cmdmerge</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> view -bh </span><span class="si">{1}</span><span class="s2"> &gt;&quot;</span> <span class="o">+</span>
                       <span class="s2">&quot;</span><span class="si">{2}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">samtools_exe</span><span class="p">,</span> <span class="n">mapping_ob</span><span class="o">.</span><span class="n">pe_map_bam</span><span class="p">,</span> <span class="n">mapping_ob</span><span class="o">.</span><span class="n">mapped_bam</span><span class="p">)</span>
        <span class="n">smaltcommands</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">cmdmerge</span><span class="p">])</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;running SMALT:&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;with the following SMALT commands:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">smaltcommands</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">platform</span> <span class="o">!=</span> <span class="s2">&quot;win32&quot;</span><span class="p">,</span>
                       <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                       <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># report simgpleton reads mapped</span>
    <span class="k">if</span> <span class="n">ngsLib</span><span class="o">.</span><span class="n">readS0</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;Singleton mapped reads: &quot;</span> <span class="o">+</span>
                        <span class="n">get_number_mapped</span><span class="p">(</span><span class="n">mapping_ob</span><span class="o">.</span><span class="n">s_map_bam</span><span class="p">,</span>
                                          <span class="n">samtools_exe</span><span class="o">=</span><span class="n">samtools_exe</span><span class="p">)))</span>
    <span class="c1"># report paired reads mapped</span>
    <span class="k">if</span> <span class="s2">&quot;pe&quot;</span> <span class="ow">in</span> <span class="n">ngsLib</span><span class="o">.</span><span class="n">libtype</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;PE mapped reads: &quot;</span> <span class="o">+</span>
                        <span class="n">get_number_mapped</span><span class="p">(</span><span class="n">mapping_ob</span><span class="o">.</span><span class="n">pe_map_bam</span><span class="p">,</span>
                                          <span class="n">samtools_exe</span><span class="o">=</span><span class="n">samtools_exe</span><span class="p">)))</span>
    <span class="n">combined_map_string</span> <span class="o">=</span> <span class="n">get_number_mapped</span><span class="p">(</span><span class="n">mapping_ob</span><span class="o">.</span><span class="n">mapped_bam_unfiltered</span><span class="p">,</span>
                                            <span class="n">samtools_exe</span><span class="o">=</span><span class="n">samtools_exe</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;Combined mapped reads: &quot;</span> <span class="o">+</span> <span class="n">combined_map_string</span><span class="p">))</span>
    <span class="c1"># extract overall percentage as a float</span>
    <span class="n">map_percentage</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">combined_map_string</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;(&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;%&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>

    <span class="c1"># apparently there have been no errors, so mapping success!</span>
    <span class="n">ngsLib</span><span class="o">.</span><span class="n">mapping_success</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">return</span> <span class="n">map_percentage</span></div>


<div class="viewcode-block" id="filter_bam_AS"><a class="viewcode-back" href="../riboSeed.html#riboSeed.filter_bam_AS">[docs]</a><span class="k">def</span> <span class="nf">filter_bam_AS</span><span class="p">(</span><span class="n">inbam</span><span class="p">,</span> <span class="n">outsam</span><span class="p">,</span> <span class="n">score</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; This is needed because bwa cannot filter based n alignment score</span>
<span class="sd">    for paired reads.</span>
<span class="sd">    https://sourceforge.net/p/bio-bwa/mailman/message/31968535/</span>
<span class="sd">    Given a  bam file from bwa (has &quot;AS&quot; tags), write out</span>
<span class="sd">    reads with AS higher than --score to outsam</span>
<span class="sd">    read count from https://www.biostars.org/p/1890/</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">notag</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">written</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">score_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">pysam</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">inbam</span><span class="p">)</span>
    <span class="n">bam</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">inbam</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
    <span class="n">osam</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">Samfile</span><span class="p">(</span><span class="n">outsam</span><span class="p">,</span> <span class="s1">&#39;wh&#39;</span><span class="p">,</span> <span class="n">template</span><span class="o">=</span><span class="n">bam</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">bam</span><span class="o">.</span><span class="n">fetch</span><span class="p">():</span>
        <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">has_tag</span><span class="p">(</span><span class="s1">&#39;AS&#39;</span><span class="p">):</span>
            <span class="n">score_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="s1">&#39;AS&#39;</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="s1">&#39;AS&#39;</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">score</span><span class="p">:</span>
                <span class="n">osam</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">read</span><span class="p">)</span>
                <span class="n">written</span> <span class="o">=</span> <span class="n">written</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">pass</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">notag</span> <span class="o">=</span> <span class="n">notag</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="k">pass</span>
    <span class="n">bam</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Reads after filtering: </span><span class="si">%i</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">written</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">notag</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Reads lacking alignment score: </span><span class="si">%i</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">notag</span><span class="p">)</span>
    <span class="k">return</span><span class="p">(</span><span class="n">score_list</span><span class="p">)</span></div>


<div class="viewcode-block" id="get_bam_AS"><a class="viewcode-back" href="../riboSeed.html#riboSeed.get_bam_AS">[docs]</a><span class="k">def</span> <span class="nf">get_bam_AS</span><span class="p">(</span><span class="n">inbam</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Return the mappign scores for downstream QC plotting.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">logger</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;must use logging&quot;</span>
    <span class="n">score_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">count</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">pysam</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">inbam</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">pysam</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">SamtoolsError</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;It looks like your bam file is unsorted! No good&quot;</span><span class="p">)</span>
    <span class="n">bam</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">inbam</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">bam</span><span class="o">.</span><span class="n">fetch</span><span class="p">():</span>
        <span class="n">count</span> <span class="o">=</span> <span class="n">count</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">has_tag</span><span class="p">(</span><span class="s1">&#39;AS&#39;</span><span class="p">):</span>
            <span class="n">score_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">read</span><span class="o">.</span><span class="n">get_tag</span><span class="p">(</span><span class="s1">&#39;AS&#39;</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span>
    <span class="n">bam</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">score_list</span><span class="p">)</span> <span class="o">!=</span> <span class="n">count</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%i</span><span class="s2"> reads did not have AS tags&quot;</span><span class="p">,</span>
                       <span class="n">count</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">score_list</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">score_list</span></div>


<div class="viewcode-block" id="sam_to_bam"><a class="viewcode-back" href="../riboSeed.html#riboSeed.sam_to_bam">[docs]</a><span class="k">def</span> <span class="nf">sam_to_bam</span><span class="p">(</span><span class="n">samtools_exe</span><span class="p">,</span> <span class="n">bam</span><span class="p">,</span> <span class="n">sam</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    becasue pysam doesnt like to write bams in an iterator, which makes sense</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">logger</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;must use logging&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Making mapped bam file:&quot;</span><span class="p">)</span>
    <span class="n">make_mapped_bam</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> view -o </span><span class="si">{1}</span><span class="s2"> -h </span><span class="si">{2}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">samtools_exe</span><span class="p">,</span> <span class="n">bam</span><span class="p">,</span> <span class="n">sam</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">make_mapped_bam</span><span class="p">)</span>
    <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="n">make_mapped_bam</span><span class="p">],</span>
                   <span class="n">shell</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">platform</span> <span class="o">!=</span> <span class="s2">&quot;win32&quot;</span><span class="p">,</span>
                   <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                   <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                   <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>


<div class="viewcode-block" id="map_to_genome_ref_bwa"><a class="viewcode-back" href="../riboSeed.html#riboSeed.map_to_genome_ref_bwa">[docs]</a><span class="k">def</span> <span class="nf">map_to_genome_ref_bwa</span><span class="p">(</span><span class="n">mapping_ob</span><span class="p">,</span> <span class="n">ngsLib</span><span class="p">,</span> <span class="n">cores</span><span class="p">,</span>
                          <span class="n">samtools_exe</span><span class="p">,</span> <span class="n">bwa_exe</span><span class="p">,</span> <span class="n">genome_fasta</span><span class="p">,</span>
                          <span class="n">score_minimum</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                          <span class="n">add_args</span><span class="o">=</span><span class="s1">&#39;-L 0,0 -U 0 -a&#39;</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Map to bam.  maps PE and S reads separately,</span>
<span class="sd">    then combines them into a X_mapped.bam file</span>
<span class="sd">    TODO:: break up into execution and comamnd generation</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Mapping reads to reference genome with BWA&quot;</span><span class="p">)</span>
    <span class="c1"># check min score</span>
    <span class="k">if</span> <span class="n">score_minimum</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">score_min</span> <span class="o">=</span> <span class="n">score_minimum</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
            <span class="s2">&quot;no bwa mapping score minprovided; default is 1/2 read &quot;</span> <span class="o">+</span>
            <span class="s2">&quot;length or 50, whichever is greater.&quot;</span><span class="p">)</span>
        <span class="c1"># hard minimum of 50</span>
        <span class="n">score_min</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">ngsLib</span><span class="o">.</span><span class="n">readlen</span><span class="p">)</span> <span class="o">/</span> <span class="mf">2.0</span><span class="p">)),</span>
                        <span class="mi">50</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;using a score minimum of </span><span class="si">%i</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">score_min</span><span class="p">)</span>
    <span class="c1"># index the reference</span>
    <span class="n">cmdindex</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> index </span><span class="si">{1}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">bwa_exe</span><span class="p">,</span> <span class="n">genome_fasta</span><span class="p">)</span>
    <span class="c1"># map paired end reads to reference index.</span>
    <span class="n">bwacommands</span> <span class="o">=</span> <span class="p">[</span><span class="n">cmdindex</span><span class="p">]</span>
    <span class="k">if</span> <span class="s2">&quot;pe&quot;</span> <span class="ow">in</span> <span class="n">ngsLib</span><span class="o">.</span><span class="n">libtype</span><span class="p">:</span>
        <span class="n">cmdmap</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="s1">&#39;</span><span class="si">{0}</span><span class="s1"> mem -t </span><span class="si">{1}</span><span class="s1"> </span><span class="si">{2}</span><span class="s1"> -k 15 &#39;</span> <span class="o">+</span>
                     <span class="s1">&#39;</span><span class="si">{3}</span><span class="s1"> </span><span class="si">{4}</span><span class="s1"> </span><span class="si">{5}</span><span class="s1"> | </span><span class="si">{6}</span><span class="s1"> view -bh - | &#39;</span> <span class="o">+</span>
                     <span class="s1">&#39;</span><span class="si">{6}</span><span class="s1"> sort -o &#39;</span> <span class="o">+</span>
                     <span class="s1">&#39;</span><span class="si">{7}</span><span class="s1"> - &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">bwa_exe</span><span class="p">,</span>  <span class="c1"># 0</span>
                                      <span class="n">cores</span><span class="p">,</span>  <span class="c1"># 1</span>
                                      <span class="n">add_args</span><span class="p">,</span>  <span class="c1"># 2</span>
                                      <span class="n">genome_fasta</span><span class="p">,</span>  <span class="c1"># 3</span>
                                      <span class="n">ngsLib</span><span class="o">.</span><span class="n">readF</span><span class="p">,</span>  <span class="c1"># 4</span>
                                      <span class="n">ngsLib</span><span class="o">.</span><span class="n">readR</span><span class="p">,</span>  <span class="c1"># 5</span>
                                      <span class="n">samtools_exe</span><span class="p">,</span>  <span class="c1"># 6</span>
                                      <span class="n">mapping_ob</span><span class="o">.</span><span class="n">pe_map_bam</span><span class="p">)</span>  <span class="c1"># 7)</span>
        <span class="n">bwacommands</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cmdmap</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">assert</span> <span class="n">ngsLib</span><span class="o">.</span><span class="n">readS0</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> \
            <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;No readS0 attribute found, cannot run mapping with &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;any reads in .readS0 or .readF and .readR&quot;</span><span class="p">)</span>

    <span class="c1"># if singletons are present, map those too.  Index is already made</span>
    <span class="k">if</span> <span class="n">ngsLib</span><span class="o">.</span><span class="n">readS0</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>  <span class="c1"># and not ignore_singletons:</span>
        <span class="n">cmdmapS</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span>
            <span class="s1">&#39;</span><span class="si">{0}</span><span class="s1"> mem -t </span><span class="si">{1}</span><span class="s1"> </span><span class="si">{2}</span><span class="s1"> -k 15 &#39;</span> <span class="o">+</span>
            <span class="s1">&#39;</span><span class="si">{3}</span><span class="s1"> </span><span class="si">{4}</span><span class="s1"> | </span><span class="si">{5}</span><span class="s1"> view -bh - | &#39;</span> <span class="o">+</span>
            <span class="s1">&#39;</span><span class="si">{5}</span><span class="s1"> sort -o </span><span class="si">{6}</span><span class="s1"> - &#39;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">bwa_exe</span><span class="p">,</span>  <span class="c1"># 0</span>
                                         <span class="n">cores</span><span class="p">,</span>  <span class="c1"># 1</span>
                                         <span class="n">add_args</span><span class="p">,</span>  <span class="c1"># 2</span>
                                         <span class="n">genome_fasta</span><span class="p">,</span>  <span class="c1"># 3</span>
                                         <span class="n">ngsLib</span><span class="o">.</span><span class="n">readS0</span><span class="p">,</span>  <span class="c1"># 4</span>
                                         <span class="n">samtools_exe</span><span class="p">,</span>  <span class="c1"># 5</span>
                                         <span class="n">mapping_ob</span><span class="o">.</span><span class="n">s_map_bam</span><span class="p">)</span>  <span class="c1"># 5)</span>
        <span class="c1"># merge together the singleton and pe reads, if there are any</span>
        <span class="k">if</span> <span class="s2">&quot;s&quot;</span> <span class="ow">in</span> <span class="n">ngsLib</span><span class="o">.</span><span class="n">libtype</span><span class="p">:</span>
            <span class="n">cmdmergeS</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span>
                <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> view -bh </span><span class="si">{1}</span><span class="s2"> &gt; </span><span class="si">{2}</span><span class="s2">&quot;</span>
            <span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">samtools_exe</span><span class="p">,</span> <span class="n">mapping_ob</span><span class="o">.</span><span class="n">s_map_bam</span><span class="p">,</span> <span class="n">mapping_ob</span><span class="o">.</span><span class="n">mapped_bam_unfiltered</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">cmdmergeS</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">{0}</span><span class="s1"> merge -f </span><span class="si">{3}</span><span class="s1"> </span><span class="si">{1}</span><span class="s1"> </span><span class="si">{2}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">samtools_exe</span><span class="p">,</span> <span class="n">mapping_ob</span><span class="o">.</span><span class="n">pe_map_bam</span><span class="p">,</span>
                <span class="n">mapping_ob</span><span class="o">.</span><span class="n">s_map_bam</span><span class="p">,</span> <span class="n">mapping_ob</span><span class="o">.</span><span class="n">mapped_bam_unfiltered</span><span class="p">)</span>
        <span class="n">bwacommands</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">cmdmapS</span><span class="p">,</span> <span class="n">cmdmergeS</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># if not already none, set to None when ignoring singleton</span>
        <span class="n">ngsLib</span><span class="o">.</span><span class="n">readS0</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">cmdmerge</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> view -bh </span><span class="si">{1}</span><span class="s2"> &gt; &quot;</span> <span class="o">+</span>
                       <span class="s2">&quot;</span><span class="si">{2}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">samtools_exe</span><span class="p">,</span> <span class="n">mapping_ob</span><span class="o">.</span><span class="n">pe_map_bam</span><span class="p">,</span>
                                     <span class="n">mapping_ob</span><span class="o">.</span><span class="n">mapped_bam_unfiltered</span><span class="p">)</span>
        <span class="n">bwacommands</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">cmdmerge</span><span class="p">])</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;running BWA:&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;with the following BWA commands:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">bwacommands</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
        <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">shell</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">platform</span> <span class="o">!=</span> <span class="s2">&quot;win32&quot;</span><span class="p">,</span>
                       <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                       <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span> <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># report simgpleton reads mapped</span>
    <span class="k">if</span> <span class="n">ngsLib</span><span class="o">.</span><span class="n">readS0</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;Singleton mapped reads: &quot;</span> <span class="o">+</span>
                        <span class="n">get_number_mapped</span><span class="p">(</span><span class="n">mapping_ob</span><span class="o">.</span><span class="n">s_map_bam</span><span class="p">,</span>
                                          <span class="n">samtools_exe</span><span class="o">=</span><span class="n">samtools_exe</span><span class="p">)))</span>
    <span class="c1"># report paired reads mapped</span>
    <span class="k">if</span> <span class="s2">&quot;pe&quot;</span> <span class="ow">in</span> <span class="n">ngsLib</span><span class="o">.</span><span class="n">libtype</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;PE mapped reads: &quot;</span> <span class="o">+</span>
                        <span class="n">get_number_mapped</span><span class="p">(</span><span class="n">mapping_ob</span><span class="o">.</span><span class="n">pe_map_bam</span><span class="p">,</span>
                                          <span class="n">samtools_exe</span><span class="o">=</span><span class="n">samtools_exe</span><span class="p">)))</span>

    <span class="n">combined_map_string</span> <span class="o">=</span> <span class="n">get_number_mapped</span><span class="p">(</span><span class="n">mapping_ob</span><span class="o">.</span><span class="n">mapped_bam_unfiltered</span><span class="p">,</span>
                                            <span class="n">samtools_exe</span><span class="o">=</span><span class="n">samtools_exe</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;Combined mapped reads: &quot;</span> <span class="o">+</span> <span class="n">combined_map_string</span><span class="p">))</span>
    <span class="c1"># extract overall percentage as a float</span>
    <span class="n">map_percentage</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">combined_map_string</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;(&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;%&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;filtering mapped reads with an AS score minimum of </span><span class="si">%i</span><span class="s2">&quot;</span><span class="p">,</span>
                 <span class="n">score_min</span><span class="p">)</span>
    <span class="n">score_list</span> <span class="o">=</span> <span class="n">filter_bam_AS</span><span class="p">(</span><span class="n">inbam</span><span class="o">=</span><span class="n">mapping_ob</span><span class="o">.</span><span class="n">mapped_bam_unfiltered</span><span class="p">,</span>
                               <span class="n">outsam</span><span class="o">=</span><span class="n">mapping_ob</span><span class="o">.</span><span class="n">mapped_sam</span><span class="p">,</span>
                               <span class="n">score</span><span class="o">=</span><span class="n">score_min</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
    <span class="n">sam_to_bam</span><span class="p">(</span><span class="n">bam</span><span class="o">=</span><span class="n">mapping_ob</span><span class="o">.</span><span class="n">mapped_bam</span><span class="p">,</span>
               <span class="n">sam</span><span class="o">=</span><span class="n">mapping_ob</span><span class="o">.</span><span class="n">mapped_sam</span><span class="p">,</span>
               <span class="n">samtools_exe</span><span class="o">=</span><span class="n">samtools_exe</span><span class="p">,</span>
               <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;Mapped reads after filtering: &quot;</span> <span class="o">+</span>
                    <span class="n">get_number_mapped</span><span class="p">(</span><span class="n">mapping_ob</span><span class="o">.</span><span class="n">mapped_bam</span><span class="p">,</span>
                                      <span class="n">samtools_exe</span><span class="o">=</span><span class="n">samtools_exe</span><span class="p">)))</span>
    <span class="c1"># apparently there have been no errors, so mapping success!</span>
    <span class="n">ngsLib</span><span class="o">.</span><span class="n">mapping_success</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">map_percentage</span><span class="p">,</span> <span class="n">score_list</span><span class="p">)</span></div>


<div class="viewcode-block" id="convert_bam_to_fastqs_cmd"><a class="viewcode-back" href="../riboSeed.html#riboSeed.convert_bam_to_fastqs_cmd">[docs]</a><span class="k">def</span> <span class="nf">convert_bam_to_fastqs_cmd</span><span class="p">(</span><span class="n">mapping_ob</span><span class="p">,</span> <span class="n">ref_fasta</span><span class="p">,</span> <span class="n">samtools_exe</span><span class="p">,</span>
                              <span class="n">which</span><span class="o">=</span><span class="s1">&#39;mapped&#39;</span><span class="p">,</span> <span class="n">source_ext</span><span class="o">=</span><span class="s2">&quot;_sam&quot;</span><span class="p">,</span>
                              <span class="n">single</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;generate a cmd to convert a bam file to fastq, using samtools</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">which</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;mapped&#39;</span><span class="p">,</span> <span class="s1">&#39;unmapped&#39;</span><span class="p">],</span> \
        <span class="s2">&quot;only valid options are mapped and unmapped&quot;</span>
    <span class="n">read_path_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;readF&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;readR&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span> <span class="s1">&#39;readS&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;preparing to convert extracted reads to make these files:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">read_path_dict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">read_path_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span>
            <span class="n">mapping_ob</span><span class="o">.</span><span class="n">mapped_bam</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="n">which</span> <span class="o">+</span> <span class="n">key</span> <span class="o">+</span> <span class="s1">&#39;.fastq&#39;</span><span class="p">)</span>

    <span class="k">assert</span> <span class="kc">None</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">read_path_dict</span><span class="o">.</span><span class="n">values</span><span class="p">(),</span> \
        <span class="s2">&quot;Could not properly construct fastq names!&quot;</span>
    <span class="c1"># if converting mapped reads, get them from the bam file</span>
    <span class="k">if</span> <span class="n">which</span> <span class="o">==</span> <span class="s1">&#39;mapped&#39;</span><span class="p">:</span>
        <span class="n">source_ext</span> <span class="o">=</span> <span class="s1">&#39;_bam&#39;</span>
    <span class="c1"># else, leave the defaultsource ext (sam)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">pass</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">single</span><span class="p">:</span>
        <span class="n">samfastq</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> fastq </span><span class="si">{1}</span><span class="s2"> -1 </span><span class="si">{2}</span><span class="s2"> -2 </span><span class="si">{3}</span><span class="s2"> -s </span><span class="si">{4}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">samtools_exe</span><span class="p">,</span>
            <span class="nb">getattr</span><span class="p">(</span><span class="n">mapping_ob</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">which</span> <span class="o">+</span> <span class="n">source_ext</span><span class="p">)),</span>
            <span class="n">read_path_dict</span><span class="p">[</span><span class="s1">&#39;readF&#39;</span><span class="p">],</span>
            <span class="n">read_path_dict</span><span class="p">[</span><span class="s1">&#39;readR&#39;</span><span class="p">],</span>
            <span class="n">read_path_dict</span><span class="p">[</span><span class="s1">&#39;readS&#39;</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;readF&#39;</span><span class="p">,</span> <span class="s1">&#39;readR&#39;</span><span class="p">,</span> <span class="s1">&#39;readS&#39;</span><span class="p">]:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">read_path_dict</span><span class="p">[</span><span class="n">key</span><span class="p">])</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># This option outputs all the reads in a single fastq</span>
        <span class="c1"># its needed for low coverage mappings when the F and R</span>
        <span class="c1"># file may end up empty.  Since default behaviour is to</span>
        <span class="c1"># treat F and R as single libraries anyway, this works</span>
        <span class="n">samfastq</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> fastq </span><span class="si">{1}</span><span class="s2"> &gt; </span><span class="si">{2}</span><span class="s2"> &quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">samtools_exe</span><span class="p">,</span>
            <span class="nb">getattr</span><span class="p">(</span><span class="n">mapping_ob</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">which</span> <span class="o">+</span> <span class="n">source_ext</span><span class="p">)),</span>
            <span class="n">read_path_dict</span><span class="p">[</span><span class="s1">&#39;readS&#39;</span><span class="p">])</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">read_path_dict</span><span class="p">[</span><span class="s1">&#39;readS&#39;</span><span class="p">])</span>
        <span class="c1"># Flag the others for ignoral</span>
        <span class="c1"># read_path_dict[&#39;readF&#39;] = None</span>
        <span class="c1"># read_path_dict[&#39;readR&#39;] = None</span>
    <span class="k">return</span><span class="p">(</span><span class="n">samfastq</span><span class="p">,</span> <span class="n">NgsLib</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">which</span><span class="p">,</span> <span class="n">master</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                            <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">,</span>
                            <span class="n">readF</span><span class="o">=</span><span class="n">read_path_dict</span><span class="p">[</span><span class="s1">&#39;readF&#39;</span><span class="p">],</span>
                            <span class="n">readR</span><span class="o">=</span><span class="n">read_path_dict</span><span class="p">[</span><span class="s1">&#39;readR&#39;</span><span class="p">],</span>
                            <span class="n">readS0</span><span class="o">=</span><span class="n">read_path_dict</span><span class="p">[</span><span class="s1">&#39;readS&#39;</span><span class="p">],</span>
                            <span class="n">ref_fasta</span><span class="o">=</span><span class="n">ref_fasta</span><span class="p">))</span></div>


<div class="viewcode-block" id="generate_spades_cmd"><a class="viewcode-back" href="../riboSeed.html#riboSeed.generate_spades_cmd">[docs]</a><span class="k">def</span> <span class="nf">generate_spades_cmd</span><span class="p">(</span>
        <span class="n">mapping_ob</span><span class="p">,</span> <span class="n">ngs_ob</span><span class="p">,</span> <span class="n">ref_as_contig</span><span class="p">,</span> <span class="n">as_paired</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">addLibs</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span>
        <span class="n">prelim</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="s2">&quot;21,33,55,77,99&quot;</span><span class="p">,</span> <span class="n">spades_exe</span><span class="o">=</span><span class="s2">&quot;spades.py&quot;</span><span class="p">,</span>
        <span class="n">single_lib</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">check_libs</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;return spades command so we can multiprocess the assemblies</span>
<span class="sd">    wrapper for common spades setting for long illumina reads</span>
<span class="sd">    ref_as_contig should be either blank, &#39;trusted&#39;, or &#39;untrusted&#39;</span>
<span class="sd">    prelim flag is True, only assembly is run, and without coverage corrections</span>
<span class="sd">    #TODO</span>
<span class="sd">    the seqname variable is used only for renaming the resulting contigs</span>
<span class="sd">    during iterative assembly.  It would be nice to inheirit from &quot;ref&quot;,</span>
<span class="sd">    but that is changed with each iteration. This should probably be addressed</span>
<span class="sd">    before next major version change</span>
<span class="sd">    #TODO dynamicllly choose kmers based on read len and whether prelim</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">logger</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Must Use Logging&quot;</span>

    <span class="c1"># make kmer comamnd empty if using &quot;auto&quot;, or set to something like</span>
    <span class="c1"># -k 55,77,99</span>
    <span class="k">if</span> <span class="n">k</span> <span class="ow">is</span> <span class="ow">not</span> <span class="s1">&#39;auto&#39;</span><span class="p">:</span>
        <span class="n">kmers</span> <span class="o">=</span> <span class="s2">&quot;-k &quot;</span> <span class="o">+</span> <span class="n">k</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">kmers</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="c1">#  prepare reference, if being used</span>
    <span class="k">if</span> <span class="n">ref_as_contig</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">alt_contig</span> <span class="o">=</span> <span class="s2">&quot;--</span><span class="si">{0}</span><span class="s2">-contigs </span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">ref_as_contig</span><span class="p">,</span> <span class="n">mapping_ob</span><span class="o">.</span><span class="n">ref_fasta</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">alt_contig</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    <span class="n">libs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">single_lib</span><span class="p">:</span>
        <span class="n">singles</span> <span class="o">=</span> <span class="s2">&quot;--pe1-s </span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">ngs_ob</span><span class="o">.</span><span class="n">readS0</span><span class="p">)</span>
        <span class="n">pairs</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="n">libs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ngs_ob</span><span class="o">.</span><span class="n">readS0</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">as_paired</span> <span class="ow">and</span> <span class="n">ngs_ob</span><span class="o">.</span><span class="n">readS0</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>  <span class="c1"># for lib with both</span>
        <span class="n">singles</span> <span class="o">=</span> <span class="s2">&quot;--pe1-s </span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">ngs_ob</span><span class="o">.</span><span class="n">readS0</span><span class="p">)</span>
        <span class="n">pairs</span> <span class="o">=</span> <span class="s2">&quot;--pe1-1 </span><span class="si">{0}</span><span class="s2"> --pe1-2 </span><span class="si">{1}</span><span class="s2"> &quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">ngs_ob</span><span class="o">.</span><span class="n">readF</span><span class="p">,</span> <span class="n">ngs_ob</span><span class="o">.</span><span class="n">readR</span><span class="p">)</span>
        <span class="n">libs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ngs_ob</span><span class="o">.</span><span class="n">readS0</span><span class="p">)</span>
        <span class="n">libs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ngs_ob</span><span class="o">.</span><span class="n">readF</span><span class="p">)</span>
        <span class="n">libs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ngs_ob</span><span class="o">.</span><span class="n">readR</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">as_paired</span> <span class="ow">and</span> <span class="n">ngs_ob</span><span class="o">.</span><span class="n">readS0</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>  <span class="c1"># for lib with just PE</span>
        <span class="n">singles</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="n">pairs</span> <span class="o">=</span> <span class="s2">&quot;--pe1-1 </span><span class="si">{0}</span><span class="s2"> --pe1-2 </span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">ngs_ob</span><span class="o">.</span><span class="n">readF</span><span class="p">,</span> <span class="n">ngs_ob</span><span class="o">.</span><span class="n">readR</span><span class="p">)</span>
        <span class="n">libs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ngs_ob</span><span class="o">.</span><span class="n">readF</span><span class="p">)</span>
        <span class="n">libs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ngs_ob</span><span class="o">.</span><span class="n">readR</span><span class="p">)</span>
    <span class="c1"># for libraries treating paired ends as two single-end libs</span>
    <span class="k">elif</span> <span class="ow">not</span> <span class="n">as_paired</span> <span class="ow">and</span> <span class="n">ngs_ob</span><span class="o">.</span><span class="n">readS0</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">singles</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
        <span class="n">pairs</span> <span class="o">=</span> <span class="s2">&quot;--pe1-s </span><span class="si">{0}</span><span class="s2"> --pe2-s </span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">ngs_ob</span><span class="o">.</span><span class="n">readF</span><span class="p">,</span> <span class="n">ngs_ob</span><span class="o">.</span><span class="n">readR</span><span class="p">)</span>
        <span class="n">libs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ngs_ob</span><span class="o">.</span><span class="n">readF</span><span class="p">)</span>
        <span class="n">libs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ngs_ob</span><span class="o">.</span><span class="n">readR</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>  <span class="c1"># for 3 single end libraries</span>
        <span class="n">singles</span> <span class="o">=</span> <span class="s2">&quot;--pe3-s </span><span class="si">{0}</span><span class="s2"> &quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">ngs_ob</span><span class="o">.</span><span class="n">readS0</span><span class="p">)</span>
        <span class="n">pairs</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;--pe1-s </span><span class="si">{0}</span><span class="s2"> --pe2-s </span><span class="si">{1}</span><span class="s2"> &quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">ngs_ob</span><span class="o">.</span><span class="n">readF</span><span class="p">,</span> <span class="n">ngs_ob</span><span class="o">.</span><span class="n">readR</span><span class="p">))</span>
        <span class="n">libs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ngs_ob</span><span class="o">.</span><span class="n">readS0</span><span class="p">)</span>
        <span class="n">libs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ngs_ob</span><span class="o">.</span><span class="n">readF</span><span class="p">)</span>
        <span class="n">libs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ngs_ob</span><span class="o">.</span><span class="n">readR</span><span class="p">)</span>
    <span class="n">reads</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">pairs</span> <span class="o">+</span> <span class="n">singles</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">prelim</span><span class="p">:</span>
        <span class="n">cmd</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span>
            <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> --only-assembler --cov-cutoff off --sc --careful </span><span class="si">{1}</span><span class="s2"> &quot;</span> <span class="o">+</span>
            <span class="s2">&quot;</span><span class="si">{2}</span><span class="s2"> </span><span class="si">{3}</span><span class="s2"> </span><span class="si">{4}</span><span class="s2"> -o </span><span class="si">{5}</span><span class="s2">&quot;</span>
        <span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">spades_exe</span><span class="p">,</span> <span class="n">kmers</span><span class="p">,</span> <span class="n">reads</span><span class="p">,</span> <span class="n">alt_contig</span><span class="p">,</span> <span class="n">addLibs</span><span class="p">,</span>
                 <span class="n">mapping_ob</span><span class="o">.</span><span class="n">assembly_subdir</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">cmd</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> --careful </span><span class="si">{1}</span><span class="s2"> </span><span class="si">{2}</span><span class="s2"> </span><span class="si">{3}</span><span class="s2"> </span><span class="si">{4}</span><span class="s2"> -o </span><span class="si">{5}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">spades_exe</span><span class="p">,</span> <span class="n">kmers</span><span class="p">,</span> <span class="n">reads</span><span class="p">,</span> <span class="n">alt_contig</span><span class="p">,</span> <span class="n">addLibs</span><span class="p">,</span>
            <span class="n">mapping_ob</span><span class="o">.</span><span class="n">assembly_subdir</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">check_libs</span><span class="p">:</span>
        <span class="n">spades_cmd</span> <span class="o">=</span> <span class="n">make_spades_empty_check</span><span class="p">(</span><span class="n">liblist</span><span class="o">=</span><span class="n">libs</span><span class="p">,</span> <span class="n">cmd</span><span class="o">=</span><span class="n">cmd</span><span class="p">,</span>
                                             <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">spades_cmd</span> <span class="o">=</span> <span class="n">cmd</span>
    <span class="k">return</span> <span class="n">spades_cmd</span></div>


<div class="viewcode-block" id="make_spades_empty_check"><a class="viewcode-back" href="../riboSeed.html#riboSeed.make_spades_empty_check">[docs]</a><span class="k">def</span> <span class="nf">make_spades_empty_check</span><span class="p">(</span><span class="n">liblist</span><span class="p">,</span> <span class="n">cmd</span><span class="p">,</span> <span class="n">logger</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; returns shell/spades cmd as string. All this does is make it a</span>
<span class="sd">    conditional shell cmd that depends on the presense of the file</span>
<span class="sd">    needed for assembly.  It is needed so we can bin all</span>
<span class="sd">    the cmds with multiprocessing.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;constructing shell file check for subprocess cmd&quot;</span><span class="p">)</span>
    <span class="n">prefix</span> <span class="o">=</span> <span class="s2">&quot;if &quot;</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">lib</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">liblist</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">i</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">prefix</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="s2">&quot;&amp;&amp; &quot;</span>
        <span class="n">check</span> <span class="o">=</span> <span class="s2">&quot;[ -s </span><span class="si">{0}</span><span class="s2"> ] &quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">lib</span><span class="p">)</span>
        <span class="n">prefix</span> <span class="o">=</span> <span class="n">prefix</span> <span class="o">+</span> <span class="n">check</span>
    <span class="n">suffix</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;; then </span><span class="si">{0}</span><span class="s2"> ; else echo &#39;input lib not found, &quot;</span> <span class="o">+</span>
                 <span class="s2">&quot;skipping this SPAdes call&#39; ; fi&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
    <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="n">prefix</span> <span class="o">+</span> <span class="n">suffix</span><span class="p">)</span></div>


<div class="viewcode-block" id="evaluate_spades_success"><a class="viewcode-back" href="../riboSeed.html#riboSeed.evaluate_spades_success">[docs]</a><span class="k">def</span> <span class="nf">evaluate_spades_success</span><span class="p">(</span><span class="n">clu</span><span class="p">,</span> <span class="n">mapping_ob</span><span class="p">,</span> <span class="n">proceed_to_target</span><span class="p">,</span> <span class="n">target_len</span><span class="p">,</span>
                            <span class="n">include_short_contigs</span><span class="p">,</span> <span class="n">min_assembly_len</span><span class="p">,</span> <span class="n">read_len</span><span class="p">,</span>
                            <span class="n">flank</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
                            <span class="n">min_delta</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                            <span class="n">keep_best_contig</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                            <span class="n">seqname</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;return success codes:</span>
<span class="sd">    0 = include contigs, all good</span>
<span class="sd">    DEPRECIATED! 1 = include contigs, but dont keep iterating</span>
<span class="sd">    2 = exclude contigs, and keep from iterating</span>
<span class="sd">    3 = exclude contigs, error ocurred</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># DANGEROUS_CONTIG_LENGTH_THRESHOLD_FACTOR = 6</span>
    <span class="n">prelog</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">-</span><span class="si">{1}</span><span class="s2">-iter-</span><span class="si">{2}</span><span class="s2">:&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;SEED_cluster&quot;</span><span class="p">,</span> <span class="n">clu</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>
                                        <span class="n">mapping_ob</span><span class="o">.</span><span class="n">iteration</span><span class="p">)</span>
    <span class="k">assert</span> <span class="n">logger</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Must Use Logging&quot;</span>
    <span class="k">if</span> <span class="n">seqname</span> <span class="o">==</span> <span class="s1">&#39;&#39;</span><span class="p">:</span>
        <span class="n">seqname</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">mapping_ob</span><span class="o">.</span><span class="n">ref_fasta</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c1">### deal with those excluded from assembly by lack of coverage depth</span>
    <span class="c1">###  ie  (coverage_exclusion=True)</span>
    <span class="k">if</span> <span class="n">clu</span><span class="o">.</span><span class="n">coverage_exclusion</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">assert</span> <span class="n">clu</span><span class="o">.</span><span class="n">coverage_exclusion</span><span class="p">,</span> \
            <span class="s2">&quot;this should only be set by the partition_mapping method.&quot;</span>
        <span class="c1">#  if this is the first iteration, return two</span>
        <span class="c1"># Otherwise,  UPDATED: continue with warning</span>
        <span class="k">if</span> <span class="n">mapping_ob</span><span class="o">.</span><span class="n">iteration</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">cluster</span><span class="o">.</span><span class="n">mappings</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">assembled_contig</span> <span class="o">=</span> \
                <span class="n">cluster</span><span class="o">.</span><span class="n">mappings</span><span class="p">[</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">assembled_contig</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot; the coverage is worryingly low, but we &quot;</span> <span class="o">+</span>
                           <span class="s2">&quot;will continue.&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="mi">0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;the coverage is worryingly low, and as this &quot;</span> <span class="o">+</span>
                        <span class="s2">&quot;is the first iteration, we must discard this contig&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="mi">2</span>
    <span class="n">mapping_ob</span><span class="o">.</span><span class="n">assembled_contig</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">mapping_ob</span><span class="o">.</span><span class="n">assembly_subdir</span><span class="p">,</span> <span class="s2">&quot;contigs.fasta&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;checking for the following file: </span><span class="se">\n</span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">mapping_ob</span><span class="o">.</span><span class="n">assembled_contig</span><span class="p">))</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">mapping_ob</span><span class="o">.</span><span class="n">assembled_contig</span><span class="p">)</span> <span class="ow">and</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">getsize</span><span class="p">(</span><span class="n">mapping_ob</span><span class="o">.</span><span class="n">assembled_contig</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
            <span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> No output from SPAdes this time around! return code 3&quot;</span><span class="p">,</span>
            <span class="n">prelog</span><span class="p">)</span>
        <span class="k">return</span> <span class="mi">3</span>
    <span class="k">if</span> <span class="n">keep_best_contig</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;reserving first contig&quot;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">keep_only_first_contig</span><span class="p">(</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mapping_ob</span><span class="o">.</span><span class="n">assembly_subdir</span><span class="p">,</span> <span class="s2">&quot;contigs.fasta&quot;</span><span class="p">),</span>
                <span class="n">newname</span><span class="o">=</span><span class="n">seqname</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
            <span class="k">raise</span> <span class="n">f</span>
    <span class="c1"># -------------------------- --------------------------- #</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> analyzing  mapping&quot;</span><span class="p">,</span> <span class="n">prelog</span><span class="p">)</span>
    <span class="n">seed_len</span> <span class="o">=</span> <span class="n">get_fasta_lengths</span><span class="p">(</span><span class="n">clu</span><span class="o">.</span><span class="n">mappings</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ref_fasta</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c1"># seed_len = get_fasta_lengths(mapping_ob.ref_fasta)[0]</span>
    <span class="c1"># set proceed_to_target params</span>
    <span class="k">if</span> <span class="n">proceed_to_target</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">target_len</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="mi">5</span> <span class="o">&gt;</span> <span class="n">target_len</span><span class="p">:</span>
            <span class="n">target_seed_len</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">target_len</span> <span class="o">*</span> <span class="n">seed_len</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">target_len</span> <span class="o">&gt;</span> <span class="mi">50</span><span class="p">:</span>
            <span class="n">target_seed_len</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">target_len</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> invalid target length provided; must be given &quot;</span> <span class="o">+</span>
                         <span class="s2">&quot;as fraction of total length or as an absolute &quot;</span> <span class="o">+</span>
                         <span class="s2">&quot;number of base pairs greater than 50&quot;</span><span class="p">,</span> <span class="n">prelog</span><span class="p">)</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="c1"># compare lengths of reference and freshly assembled contig</span>
    <span class="n">contig_len</span> <span class="o">=</span> <span class="n">get_fasta_lengths</span><span class="p">(</span><span class="n">mapping_ob</span><span class="o">.</span><span class="n">assembled_contig</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">ref_len</span> <span class="o">=</span> <span class="n">get_fasta_lengths</span><span class="p">(</span><span class="n">mapping_ob</span><span class="o">.</span><span class="n">ref_fasta</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">contig_length_diff</span> <span class="o">=</span> <span class="n">contig_len</span> <span class="o">-</span> <span class="n">ref_len</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> Seed length: </span><span class="si">%i</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">prelog</span><span class="p">,</span> <span class="n">seed_len</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">proceed_to_target</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Target length: </span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">target_seed_len</span><span class="p">))</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> Length of this iteration&#39;s longest contig: </span><span class="si">%i</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="n">prelog</span><span class="p">,</span> <span class="n">contig_len</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">mapping_ob</span><span class="o">.</span><span class="n">iteration</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> Length of previous longest contig: </span><span class="si">%i</span><span class="s2">&quot;</span><span class="p">,</span>
                    <span class="n">prelog</span><span class="p">,</span> <span class="n">ref_len</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> The new contig differs from the previous &quot;</span> <span class="o">+</span>
                    <span class="s2">&quot;iteration by </span><span class="si">%i</span><span class="s2"> bases&quot;</span><span class="p">,</span> <span class="n">prelog</span><span class="p">,</span> <span class="n">contig_length_diff</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> The new contig differs from the reference &quot;</span> <span class="o">+</span>
                    <span class="s2">&quot;seed by </span><span class="si">%i</span><span class="s2"> bases&quot;</span><span class="p">,</span> <span class="n">prelog</span><span class="p">,</span> <span class="n">contig_length_diff</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">contig_len</span> <span class="o">&gt;</span> <span class="p">(</span><span class="n">ref_len</span> <span class="o">+</span> <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="n">flank</span><span class="p">)):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
            <span class="s2">&quot;Contig length is exceedingly long!  We set the threshold of &quot;</span> <span class="o">+</span>
            <span class="s2">&quot;twice the flanking length as the maximum allowed long-read &quot;</span> <span class="o">+</span>
            <span class="s2">&quot;length. This may indicate  bad mapping parameters, so the &quot;</span> <span class="o">+</span>
            <span class="s2">&quot;long-read will be discarded.  Return code 2&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="mi">2</span>

    <span class="c1"># This cuts failing assemblies short</span>
    <span class="k">if</span> <span class="n">min_assembly_len</span> <span class="o">&gt;</span> <span class="n">contig_len</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;The first iteration&#39;s assembly&#39;s best contig &quot;</span> <span class="o">+</span>
                       <span class="s2">&quot;is not greater than length set by &quot;</span> <span class="o">+</span>
                       <span class="s2">&quot;--min_assembly_len. Assembly will likely fail if &quot;</span> <span class="o">+</span>
                       <span class="s2">&quot;the contig does not meet the length of the seed&quot;</span><span class="p">)</span>
        <span class="c1"># if mapping_ob.iteration &gt; 0:</span>
        <span class="k">if</span> <span class="n">include_short_contigs</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Continuing to , but if this occurs for more &quot;</span> <span class="o">+</span>
                           <span class="s2">&quot;than one seed, we reccommend  you abort and &quot;</span> <span class="o">+</span>
                           <span class="s2">&quot;retry with longer seeds, a different ref, &quot;</span> <span class="o">+</span>
                           <span class="s2">&quot;or re-examine the riboSnag clustering&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Return code 0&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="mi">0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Return code 2&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="mi">2</span>
    <span class="k">elif</span> <span class="n">proceed_to_target</span> <span class="ow">and</span> <span class="n">contig_len</span> <span class="o">&gt;=</span> <span class="n">target_seed_len</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;target length threshold! has been reached; &quot;</span> <span class="o">+</span>
                    <span class="s2">&quot;skipping future iterations.  return code 0&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="mi">0</span>
    <span class="c1"># if not first time through, ensure adequate change between iterations to</span>
    <span class="c1"># avoid problems with trying to assemble a very small number of reads</span>
    <span class="k">elif</span> <span class="n">min_delta</span> <span class="o">&gt;</span> <span class="nb">abs</span><span class="p">(</span><span class="n">contig_length_diff</span><span class="p">)</span> <span class="ow">and</span> <span class="n">mapping_ob</span><span class="o">.</span><span class="n">iteration</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span>
            <span class="s2">&quot;The length of the assembled contig didn&#39;t change more &quot;</span> <span class="o">+</span>
            <span class="s2">&quot;more than </span><span class="si">{0}</span><span class="s2">bp between rounds of iteration. Continuing &quot;</span> <span class="o">+</span>
            <span class="s2">&quot;will likely cause error; skipping future iterations. &quot;</span> <span class="o">+</span>
            <span class="s2">&quot;return code 0&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">min_delta</span><span class="p">))</span>
        <span class="k">return</span> <span class="mi">0</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;return code 0&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="mi">0</span></div>


<div class="viewcode-block" id="parse_subassembly_return_code"><a class="viewcode-back" href="../riboSeed.html#riboSeed.parse_subassembly_return_code">[docs]</a><span class="k">def</span> <span class="nf">parse_subassembly_return_code</span><span class="p">(</span><span class="n">cluster</span><span class="p">,</span> <span class="n">final_contigs_dir</span><span class="p">,</span> <span class="n">skip_copy</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                  <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; given a return code from the above spades success function,</span>
<span class="sd">    set object attributes as needed</span>
<span class="sd">    20170531 depreciated return code 1: as we have moved to using the</span>
<span class="sd">    whole library for mapping, not just the unmapped reads, it is more</span>
<span class="sd">    important to keep in even unchanging contigs to allow for</span>
<span class="sd">    competition during mapping</span>

<span class="sd">    ----------------------</span>
<span class="sd">    return success codes:</span>
<span class="sd">    0 = include contigs, all good</span>
<span class="sd">    1 = include contigs, but dont keep iterating</span>
<span class="sd">    2 = exclude contigs, and keep from iterating</span>
<span class="sd">    3 = exclude contigs, error ocurred</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">logger</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;must use logging&quot;</span>
    <span class="k">if</span> <span class="n">cluster</span><span class="o">.</span><span class="n">assembly_success</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
        <span class="c1"># TODO other error handling; make a &quot;failed&quot; counter?</span>
        <span class="n">cluster</span><span class="o">.</span><span class="n">continue_iterating</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">cluster</span><span class="o">.</span><span class="n">keep_contigs</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">elif</span> <span class="n">cluster</span><span class="o">.</span><span class="n">assembly_success</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
        <span class="n">cluster</span><span class="o">.</span><span class="n">continue_iterating</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="n">cluster</span><span class="o">.</span><span class="n">keep_contigs</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">elif</span> <span class="n">cluster</span><span class="o">.</span><span class="n">assembly_success</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;return code 1 depreciated! a warning can be &quot;</span> <span class="o">+</span>
                         <span class="s2">&quot;issued for short asssemblies, but they must &quot;</span> <span class="o">+</span>
                         <span class="s2">&quot;remain in the pseudogenome&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">cluster</span><span class="o">.</span><span class="n">assembly_success</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">cluster</span><span class="o">.</span><span class="n">continue_iterating</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">cluster</span><span class="o">.</span><span class="n">keep_contigs</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Error evaluating spades results return!&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="make_quick_quast_table"><a class="viewcode-back" href="../riboSeed.html#riboSeed.make_quick_quast_table">[docs]</a><span class="k">def</span> <span class="nf">make_quick_quast_table</span><span class="p">(</span><span class="n">pathlist</span><span class="p">,</span> <span class="n">write</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">writedir</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; given paths to two or more quast reports, this generates dictionary</span>
<span class="sd">    where the key is the field in the report and the value is a list of</span>
<span class="sd">    the values for each report.   Hand for passing to the logger function.</span>
<span class="sd">    This skips any fields not in first report, for better or worse...</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">logger</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Must Use Logging&quot;</span>
    <span class="k">assert</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pathlist</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span> <span class="ow">is</span> <span class="kc">True</span><span class="p">,</span>\
        <span class="s2">&quot;paths for quast reports must be in a list!&quot;</span>
    <span class="n">filelist</span> <span class="o">=</span> <span class="n">pathlist</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Quast reports to combine: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">filelist</span><span class="p">))</span>
    <span class="n">mainDict</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">filelist</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">counter</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">dex</span><span class="p">,</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">handle</span><span class="p">):</span>
                        <span class="n">row</span><span class="p">,</span> <span class="n">val</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="k">if</span> <span class="n">dex</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
                            <span class="k">continue</span>  <span class="c1"># skip header</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">mainDict</span><span class="p">[</span><span class="n">row</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">val</span><span class="p">]</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;error parsing </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">report_list</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">handle</span><span class="p">:</span>
                    <span class="k">for</span> <span class="n">dex</span><span class="p">,</span> <span class="n">line</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">handle</span><span class="p">):</span>
                        <span class="n">row</span><span class="p">,</span> <span class="n">val</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                        <span class="n">report_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">row</span><span class="p">,</span> <span class="n">val</span><span class="p">])</span>
                    <span class="c1"># logger.debug(&quot;report list: %s&quot;, str(report_list))</span>
                    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">mainDict</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                        <span class="k">if</span> <span class="n">k</span> <span class="ow">in</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">report_list</span><span class="p">]:</span>
                            <span class="n">mainDict</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                                <span class="nb">str</span><span class="p">([</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>
                                     <span class="n">report_list</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">k</span><span class="p">][</span><span class="mi">0</span><span class="p">]))</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">mainDict</span><span class="p">[</span><span class="n">k</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;XX&quot;</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;error parsing </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>
                <span class="k">raise</span> <span class="n">e</span>
        <span class="n">counter</span> <span class="o">=</span> <span class="n">counter</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="k">if</span> <span class="n">write</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">writedir</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;no output dir, cannot write!&quot;</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">mainDict</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">writedir</span><span class="p">,</span> <span class="s2">&quot;combined_quast_report.tsv&quot;</span><span class="p">),</span>
                      <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outfile</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">mainDict</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
                    <span class="c1"># logger.debug(&quot;{0}\t{1}\n&quot;.format(k, str(&quot;\t&quot;.join(v))))</span>
                    <span class="n">outfile</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="se">\t</span><span class="si">{1}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="nb">str</span><span class="p">(</span><span class="n">k</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">v</span><span class="p">))))</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">e</span>
    <span class="k">return</span> <span class="n">mainDict</span></div>


<div class="viewcode-block" id="check_kmer_vs_reads"><a class="viewcode-back" href="../riboSeed.html#riboSeed.check_kmer_vs_reads">[docs]</a><span class="k">def</span> <span class="nf">check_kmer_vs_reads</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">readlen</span><span class="p">,</span> <span class="n">min_diff</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="k">assert</span> <span class="n">logger</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;must use logging! &quot;</span>
    <span class="c1"># ignore this if we are letting spades</span>
    <span class="k">if</span> <span class="n">k</span> <span class="ow">is</span> <span class="s1">&#39;auto&#39;</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;no need to check k, we let spades set k&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">k</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">klist</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">k</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)]</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;error splitting kmers by comma!&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">last_exception</span><span class="p">())</span>
        <span class="k">raise</span> <span class="ne">ValueError</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">klist</span><span class="p">)</span>
    <span class="n">new_ks</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">klist</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;</span> <span class="n">readlen</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;removing </span><span class="si">%d</span><span class="s2"> from list of kmers: exceeds read length&quot;</span><span class="p">,</span>
                        <span class="n">i</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">readlen</span> <span class="o">-</span> <span class="n">i</span> <span class="o">&lt;=</span> <span class="n">min_diff</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;removing </span><span class="si">%d</span><span class="s2"> from list of kmers: too close &quot;</span> <span class="o">+</span>
                        <span class="s2">&quot;to read length&quot;</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">i</span> <span class="o">%</span> <span class="mi">2</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;removing </span><span class="si">%d</span><span class="s2"> from list of kmers: must be odd&quot;</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">new_ks</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
    <span class="k">return</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">new_ks</span><span class="p">])</span></div>


<div class="viewcode-block" id="make_samtools_depth_cmds"><a class="viewcode-back" href="../riboSeed.html#riboSeed.make_samtools_depth_cmds">[docs]</a><span class="k">def</span> <span class="nf">make_samtools_depth_cmds</span><span class="p">(</span><span class="n">exe</span><span class="p">,</span> <span class="n">bam</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">region</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">prep</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; this just makes the commands to get the depth from samtools.</span>
<span class="sd">    If prep, the sam file gets sorted and indexed first</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">prep_cmds</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># cmd = &quot;samtools depth ./iter_1_s_mappi.bam -r scannedScaffolds:5000-6000&quot;</span>
    <span class="n">sorted_bam</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">bam</span><span class="p">),</span>
        <span class="nb">str</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">bam</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;_sorted.bam&quot;</span><span class="p">))</span>
    <span class="c1"># sort that bam, just in case</span>
    <span class="n">prep_cmds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> sort </span><span class="si">{1}</span><span class="s2"> &gt; </span><span class="si">{2}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">exe</span><span class="p">,</span> <span class="n">bam</span><span class="p">,</span> <span class="n">sorted_bam</span><span class="p">))</span>
    <span class="c1"># index that bam!</span>
    <span class="n">prep_cmds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> index </span><span class="si">{1}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">exe</span><span class="p">,</span> <span class="n">sorted_bam</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">prep</span><span class="p">:</span>
        <span class="n">bamfile</span> <span class="o">=</span> <span class="n">sorted_bam</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">bamfile</span> <span class="o">=</span> <span class="n">bam</span>
    <span class="c1"># extract the depth stats for a region</span>
    <span class="k">if</span> <span class="n">region</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">depth_cmd</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> depth -r </span><span class="si">{2}</span><span class="s2">:</span><span class="si">{3}</span><span class="s2">-</span><span class="si">{4}</span><span class="s2"> </span><span class="si">{1}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">exe</span><span class="p">,</span> <span class="n">bamfile</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">depth_cmd</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> depth -r </span><span class="si">{2}</span><span class="s2"> </span><span class="si">{1}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">exe</span><span class="p">,</span> <span class="n">bamfile</span><span class="p">,</span> <span class="n">region</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">prep_cmds</span><span class="p">,</span> <span class="n">depth_cmd</span><span class="p">)</span></div>


<div class="viewcode-block" id="parse_samtools_depth_results"><a class="viewcode-back" href="../riboSeed.html#riboSeed.parse_samtools_depth_results">[docs]</a><span class="k">def</span> <span class="nf">parse_samtools_depth_results</span><span class="p">(</span><span class="n">result</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; parses out the subprocess results from samtools depth</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">splits</span> <span class="o">=</span> <span class="n">result</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">splits</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">3</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;unable to split the results from samtools depth&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="k">raise</span> <span class="n">e</span>
    <span class="n">covs</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)[</span><span class="mi">2</span><span class="p">])</span> <span class="k">for</span>
            <span class="n">x</span> <span class="ow">in</span> <span class="n">result</span><span class="o">.</span><span class="n">stdout</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">]]</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">covs</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">result</span><span class="o">.</span><span class="n">returncode</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Error parsing samtools depth results! &quot;</span> <span class="o">+</span>
                           <span class="s2">&quot;Here are the results:&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;This isn&#39;t always fatal, so we will continue. &quot;</span> <span class="o">+</span>
                           <span class="s2">&quot;but take a look with IGB or similar so there &quot;</span> <span class="o">+</span>
                           <span class="s2">&quot;aren&#39;t any suprises down the road.&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">[[</span><span class="s2">&quot;&quot;</span><span class="p">],</span> <span class="mi">0</span><span class="p">]</span>

    <span class="n">average</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">covs</span><span class="p">))</span> <span class="o">/</span> <span class="nb">float</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">covs</span><span class="p">))</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">covs</span><span class="p">,</span> <span class="n">average</span><span class="p">]</span></div>


<div class="viewcode-block" id="get_samtools_depths"><a class="viewcode-back" href="../riboSeed.html#riboSeed.get_samtools_depths">[docs]</a><span class="k">def</span> <span class="nf">get_samtools_depths</span><span class="p">(</span><span class="n">samtools_exe</span><span class="p">,</span> <span class="n">bam</span><span class="p">,</span> <span class="n">chrom</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span>
                        <span class="n">prep</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">region</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Use samtools depth and awk to get the average coverage depth of a</span>
<span class="sd">    particular region</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">prep_cmds</span><span class="p">,</span> <span class="n">depth_cmd</span> <span class="o">=</span> <span class="n">make_samtools_depth_cmds</span><span class="p">(</span>
        <span class="n">exe</span><span class="o">=</span><span class="n">samtools_exe</span><span class="p">,</span> <span class="n">bam</span><span class="o">=</span><span class="n">bam</span><span class="p">,</span> <span class="n">chrom</span><span class="o">=</span><span class="n">chrom</span><span class="p">,</span>
        <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="n">end</span><span class="p">,</span> <span class="n">region</span><span class="o">=</span><span class="n">region</span><span class="p">,</span> <span class="n">prep</span><span class="o">=</span><span class="n">prep</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;running the following commands to get coverage:&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">prep</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">prep_cmds</span><span class="p">:</span>  <span class="c1"># index and sort</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">i</span><span class="p">,</span>
                           <span class="n">shell</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">platform</span> <span class="o">!=</span> <span class="s2">&quot;win32&quot;</span><span class="p">,</span>
                           <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                           <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                           <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="c1"># get the results from the depth call</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">depth_cmd</span><span class="p">)</span>
    <span class="n">result</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">depth_cmd</span><span class="p">,</span>
                            <span class="n">shell</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">platform</span> <span class="o">!=</span> <span class="s2">&quot;win32&quot;</span><span class="p">,</span>
                            <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                            <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                            <span class="n">check</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">covs</span><span class="p">,</span> <span class="n">ave</span> <span class="o">=</span> <span class="n">parse_samtools_depth_results</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">covs</span><span class="p">,</span> <span class="n">ave</span><span class="p">]</span></div>


<div class="viewcode-block" id="prepare_next_mapping"><a class="viewcode-back" href="../riboSeed.html#riboSeed.prepare_next_mapping">[docs]</a><span class="k">def</span> <span class="nf">prepare_next_mapping</span><span class="p">(</span><span class="n">cluster</span><span class="p">,</span> <span class="n">seedGenome</span><span class="p">,</span> <span class="n">samtools_exe</span><span class="p">,</span> <span class="n">flank</span><span class="p">,</span>
                         <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;use within partition mapping funtion;</span>
<span class="sd">    makes LociMapping, get region coords, write extracted region,</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mapping_subdir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">seedGenome</span><span class="o">.</span><span class="n">output_root</span><span class="p">,</span> <span class="n">cluster</span><span class="o">.</span><span class="n">cluster_dir_name</span><span class="p">,</span>
        <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_cluster_</span><span class="si">{1}</span><span class="s2">_mapping_iteration_</span><span class="si">{2}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">cluster</span><span class="o">.</span><span class="n">sequence_id</span><span class="p">,</span> <span class="n">cluster</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">))</span>
    <span class="n">assembly_subdir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">seedGenome</span><span class="o">.</span><span class="n">output_root</span><span class="p">,</span> <span class="n">cluster</span><span class="o">.</span><span class="n">cluster_dir_name</span><span class="p">,</span>
        <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_cluster_</span><span class="si">{1}</span><span class="s2">_assembly_iteration_</span><span class="si">{2}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">cluster</span><span class="o">.</span><span class="n">sequence_id</span><span class="p">,</span> <span class="n">cluster</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">))</span>

    <span class="n">mapping0</span> <span class="o">=</span> <span class="n">LociMapping</span><span class="p">(</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_cluster_</span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">cluster</span><span class="o">.</span><span class="n">sequence_id</span><span class="p">,</span> <span class="n">cluster</span><span class="o">.</span><span class="n">index</span><span class="p">),</span>
        <span class="n">iteration</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">,</span>
        <span class="n">assembly_subdir_needed</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">mapping_subdir</span><span class="o">=</span><span class="n">mapping_subdir</span><span class="p">,</span>
        <span class="n">assembly_subdir</span><span class="o">=</span><span class="n">assembly_subdir</span><span class="p">)</span>
    <span class="c1"># if first time through, get the global start and end coords.</span>
    <span class="k">if</span> <span class="n">cluster</span><span class="o">.</span><span class="n">global_start_coord</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">cluster</span><span class="o">.</span><span class="n">global_end_coord</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;global start and end should be defined previously! Exiting&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">start_coord</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">cluster</span><span class="o">.</span><span class="n">loci_list</span><span class="p">])</span> <span class="o">!=</span> \
           <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">start_coord</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">cluster</span><span class="o">.</span><span class="n">loci_list</span><span class="p">]:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Coords are not in increasing order; &quot;</span> <span class="o">+</span>
                           <span class="s2">&quot;you&#39;ve been warned&quot;</span><span class="p">)</span>
        <span class="n">start_list</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">start_coord</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">cluster</span><span class="o">.</span><span class="n">loci_list</span><span class="p">])</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Start_list: </span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">start_list</span><span class="p">))</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Finding coords to gather reads from the following loci:&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">cluster</span><span class="o">.</span><span class="n">loci_list</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> cluster </span><span class="si">%i</span><span class="s2"> -- locus </span><span class="si">%i</span><span class="s2"> -- </span><span class="si">%s</span><span class="s2"> (</span><span class="si">%i</span><span class="s2">, </span><span class="si">%i</span><span class="s2">)(</span><span class="si">%i</span><span class="s2">) </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                         <span class="n">i</span><span class="o">.</span><span class="n">sequence_id</span><span class="p">,</span> <span class="n">cluster</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>
                         <span class="n">i</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">i</span><span class="o">.</span><span class="n">locus_tag</span><span class="p">,</span>
                         <span class="n">i</span><span class="o">.</span><span class="n">start_coord</span><span class="p">,</span> <span class="n">i</span><span class="o">.</span><span class="n">end_coord</span><span class="p">,</span> <span class="n">i</span><span class="o">.</span><span class="n">strand</span><span class="p">,</span>
                         <span class="n">i</span><span class="o">.</span><span class="n">product</span><span class="p">)</span>
        <span class="c1">#  This works as long as coords are never in reverse order</span>
        <span class="n">cluster</span><span class="o">.</span><span class="n">global_start_coord</span> <span class="o">=</span> <span class="nb">min</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">start_coord</span> <span class="k">for</span>
                                          <span class="n">x</span> <span class="ow">in</span> <span class="n">cluster</span><span class="o">.</span><span class="n">loci_list</span><span class="p">])</span> <span class="o">-</span> <span class="n">flank</span>
        <span class="c1"># if start is negative, just use 1, the beginning of the sequence</span>
        <span class="k">if</span> <span class="n">cluster</span><span class="o">.</span><span class="n">global_start_coord</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot;Caution! Cannot retrieve full flanking region, as &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;the 5&#39; flanking region extends past start of &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;sequence. If this is a problem, try using a smaller &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;--flanking region, and/or if  appropriate, run with &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;--linear.&quot;</span><span class="p">)</span>
            <span class="n">cluster</span><span class="o">.</span><span class="n">global_start_coord</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="n">cluster</span><span class="o">.</span><span class="n">global_end_coord</span> <span class="o">=</span> <span class="nb">max</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">end_coord</span> <span class="k">for</span>
                                        <span class="n">x</span> <span class="ow">in</span> <span class="n">cluster</span><span class="o">.</span><span class="n">loci_list</span><span class="p">])</span> <span class="o">+</span> <span class="n">flank</span>
        <span class="c1"># logger.debug(&quot;rec len: %i&quot;, len(cluster.seq_record.seq))</span>
        <span class="k">if</span> <span class="n">cluster</span><span class="o">.</span><span class="n">global_end_coord</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">cluster</span><span class="o">.</span><span class="n">seq_record</span><span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot;Caution! Cannot retrieve full flanking region, as &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;the 5&#39; flanking region extends past start of &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;sequence. If this is a problem, try using a smaller &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;--flanking region, and/or if  appropriate, run with &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;--linear.&quot;</span><span class="p">)</span>
            <span class="n">cluster</span><span class="o">.</span><span class="n">global_end_coord</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">cluster</span><span class="o">.</span><span class="n">seq_record</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;global start and end: </span><span class="si">%s</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                     <span class="n">cluster</span><span class="o">.</span><span class="n">global_start_coord</span><span class="p">,</span>
                     <span class="n">cluster</span><span class="o">.</span><span class="n">global_end_coord</span><span class="p">)</span>
    <span class="c1">#  if not the first time though, fuhgetaboudit.</span>
    <span class="c1">#  Ie, the coords have been reassigned by the make_faux_genome function</span>
    <span class="c1">#  WE WONT USE A FLANKING REGION BECAUSE NO FLANKING READS ARE AVAILIBLE!</span>
    <span class="c1">#  meaning, the overhang is gained from the bits that overhand the end of</span>
    <span class="c1">#  the mapping. Because both SMALT and BWA use soft-clipping by defualt, we</span>
    <span class="c1">#  recover and use the clipped regions</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;using coords from previous iterations &#39;genome&#39;:&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Coordinates for </span><span class="si">%s</span><span class="s2"> cluster </span><span class="si">%i</span><span class="s2">:  [</span><span class="si">%i</span><span class="s2"> - </span><span class="si">%i</span><span class="s2">]&quot;</span><span class="p">,</span>
                <span class="n">cluster</span><span class="o">.</span><span class="n">seq_record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span>
                <span class="n">cluster</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>
                <span class="n">cluster</span><span class="o">.</span><span class="n">global_start_coord</span><span class="p">,</span>
                <span class="n">cluster</span><span class="o">.</span><span class="n">global_end_coord</span><span class="p">)</span>

    <span class="n">cluster</span><span class="o">.</span><span class="n">extractedSeqRecord</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span>
        <span class="n">cluster</span><span class="o">.</span><span class="n">seq_record</span><span class="o">.</span><span class="n">seq</span><span class="p">[</span>
            <span class="n">cluster</span><span class="o">.</span><span class="n">global_start_coord</span><span class="p">:</span>
            <span class="n">cluster</span><span class="o">.</span><span class="n">global_end_coord</span><span class="p">])</span>

    <span class="n">mapping0</span><span class="o">.</span><span class="n">ref_fasta</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">mapping0</span><span class="o">.</span><span class="n">mapping_subdir</span><span class="p">,</span>
                                      <span class="s2">&quot;extracted_seed_sequence.fasta&quot;</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">mapping0</span><span class="o">.</span><span class="n">ref_fasta</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">writepath</span><span class="p">:</span>
        <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">cluster</span><span class="o">.</span><span class="n">extractedSeqRecord</span><span class="p">,</span> <span class="n">writepath</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">)</span>
    <span class="n">cluster</span><span class="o">.</span><span class="n">mappings</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mapping0</span><span class="p">)</span></div>


<div class="viewcode-block" id="make_mapped_partition_cmds"><a class="viewcode-back" href="../riboSeed.html#riboSeed.make_mapped_partition_cmds">[docs]</a><span class="k">def</span> <span class="nf">make_mapped_partition_cmds</span><span class="p">(</span><span class="n">cluster</span><span class="p">,</span> <span class="n">mapping_ob</span><span class="p">,</span> <span class="n">seedGenome</span><span class="p">,</span> <span class="n">samtools_exe</span><span class="p">,</span>
                               <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; returns cmds and region</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Prepare for partitioning</span>
    <span class="n">partition_cmds</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># sort our source bam</span>
    <span class="n">sort_cmd</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> sort </span><span class="si">{1}</span><span class="s2"> &gt; </span><span class="si">{2}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">samtools_exe</span><span class="p">,</span>
        <span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">]</span><span class="o">.</span><span class="n">mapped_bam</span><span class="p">,</span>
        <span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span>
            <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">]</span><span class="o">.</span><span class="n">sorted_mapped_bam</span><span class="p">)</span>
    <span class="c1"># index it</span>
    <span class="n">index_cmd</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> index </span><span class="si">{1}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">samtools_exe</span><span class="p">,</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span>
            <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">]</span><span class="o">.</span><span class="n">sorted_mapped_bam</span><span class="p">)</span>
    <span class="n">partition_cmds</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">sort_cmd</span><span class="p">,</span> <span class="n">index_cmd</span><span class="p">])</span>
    <span class="c1"># define the region to extract</span>
    <span class="n">region_to_extract</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">:</span><span class="si">{1}</span><span class="s2">-</span><span class="si">{2}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">cluster</span><span class="o">.</span><span class="n">sequence_id</span><span class="p">,</span> <span class="n">cluster</span><span class="o">.</span><span class="n">global_start_coord</span><span class="p">,</span>
        <span class="n">cluster</span><span class="o">.</span><span class="n">global_end_coord</span><span class="p">)</span>
    <span class="c1"># make a subser from of reads in that region</span>
    <span class="n">view_cmd</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> view -o </span><span class="si">{1}</span><span class="s2"> </span><span class="si">{2}</span><span class="s2"> </span><span class="si">{3}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">samtools_exe</span><span class="p">,</span> <span class="n">mapping_ob</span><span class="o">.</span><span class="n">mapped_bam</span><span class="p">,</span>
        <span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span>
            <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">]</span><span class="o">.</span><span class="n">sorted_mapped_bam</span><span class="p">,</span>
        <span class="n">region_to_extract</span><span class="p">)</span>
    <span class="n">partition_cmds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">view_cmd</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">partition_cmds</span><span class="p">,</span> <span class="n">region_to_extract</span><span class="p">)</span></div>


<div class="viewcode-block" id="make_unmapped_partition_cmds"><a class="viewcode-back" href="../riboSeed.html#riboSeed.make_unmapped_partition_cmds">[docs]</a><span class="k">def</span> <span class="nf">make_unmapped_partition_cmds</span><span class="p">(</span><span class="n">mapped_regions</span><span class="p">,</span> <span class="n">samtools_exe</span><span class="p">,</span> <span class="n">seedGenome</span><span class="p">):</span>
    <span class="n">unmapped_cmds</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="sd">&quot;&quot;&quot; given a list of regions (formatted for samtools view, etc) make a</span>
<span class="sd">    list of mapped reads (file path stored under mapped_ids_txt), and</span>
<span class="sd">    use the cgrep voodoo to make a sam file from the full library without</span>
<span class="sd">    the mapped reads. returns a cmd as a string</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># starting at second iteration, copy previous iterms mapped_ids_txt</span>
    <span class="c1"># as a starting point so we can track the reads better.</span>
    <span class="k">if</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">shutil</span><span class="o">.</span><span class="n">copyfile</span><span class="p">(</span>
            <span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span>
                <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">mapped_ids_txt</span><span class="p">,</span>
            <span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span>
                <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">]</span><span class="o">.</span><span class="n">mapped_ids_txt</span><span class="p">)</span>

    <span class="c1"># moved to filter_bam_as function</span>
    <span class="n">make_mapped_sam</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> view -o </span><span class="si">{1}</span><span class="s2"> -h </span><span class="si">{2}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">samtools_exe</span><span class="p">,</span>
        <span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">]</span><span class="o">.</span><span class="n">mapped_sam</span><span class="p">,</span>
        <span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">]</span><span class="o">.</span><span class="n">mapped_bam</span><span class="p">)</span>

    <span class="n">unmapped_cmds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">make_mapped_sam</span><span class="p">)</span>

    <span class="c1"># for each region, add read names in that region to</span>
    <span class="c1"># a list (taken from previous iteration if there has been one)</span>
    <span class="k">for</span> <span class="n">region</span> <span class="ow">in</span> <span class="n">mapped_regions</span><span class="p">:</span>
        <span class="n">unmapped_cmds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> view </span><span class="si">{1}</span><span class="s2"> </span><span class="si">{2}</span><span class="s2"> | cut -f1 &gt;&gt; </span><span class="si">{3}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">samtools_exe</span><span class="p">,</span>
                <span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span>
                    <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">]</span><span class="o">.</span><span class="n">sorted_mapped_bam</span><span class="p">,</span>
                <span class="n">region</span><span class="p">,</span>
                <span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span>
                    <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">]</span><span class="o">.</span><span class="n">mapped_ids_txt</span><span class="p">))</span>
    <span class="n">uniquify_list</span> <span class="o">=</span> <span class="s2">&quot;sort -u </span><span class="si">{0}</span><span class="s2"> -o </span><span class="si">{0}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">]</span><span class="o">.</span><span class="n">mapped_ids_txt</span><span class="p">)</span>
    <span class="n">unmapped_cmds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">uniquify_list</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">unmapped_cmds</span></div>


<div class="viewcode-block" id="pysam_extract_reads"><a class="viewcode-back" href="../riboSeed.html#riboSeed.pysam_extract_reads">[docs]</a><span class="k">def</span> <span class="nf">pysam_extract_reads</span><span class="p">(</span><span class="n">sam</span><span class="p">,</span> <span class="n">textfile</span><span class="p">,</span> <span class="n">unmapped_sam</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; This replaces the &quot;LC_ALL &quot; grep call from the above function. On macs,</span>
<span class="sd">    there is no speedup gained.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">qfile</span> <span class="o">=</span> <span class="n">textfile</span>  <span class="c1"># contains read names of mapped reads</span>
    <span class="n">sfile</span> <span class="o">=</span> <span class="n">sam</span>  <span class="c1"># mapping of all reads to genome</span>
    <span class="n">ofile</span> <span class="o">=</span> <span class="n">unmapped_sam</span>  <span class="c1"># destination sam of all reads not in regions</span>
    <span class="n">nunmapped</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">total</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="c1"># Load query fixed strings as a set</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">qfile</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">qfh</span><span class="p">:</span>
        <span class="n">queries</span> <span class="o">=</span> <span class="p">{</span><span class="n">q</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">q</span> <span class="ow">in</span> <span class="n">qfh</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">q</span><span class="o">.</span><span class="n">strip</span><span class="p">())}</span>
    <span class="c1"># Subset reads</span>
    <span class="n">samfile</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">AlignmentFile</span><span class="p">(</span><span class="n">sfile</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
    <span class="n">osam</span> <span class="o">=</span> <span class="n">pysam</span><span class="o">.</span><span class="n">Samfile</span><span class="p">(</span><span class="n">ofile</span><span class="p">,</span> <span class="s1">&#39;wh&#39;</span><span class="p">,</span> <span class="n">template</span><span class="o">=</span><span class="n">samfile</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">read</span> <span class="ow">in</span> <span class="n">samfile</span><span class="o">.</span><span class="n">fetch</span><span class="p">():</span>
        <span class="n">total</span> <span class="o">=</span> <span class="n">total</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">read</span><span class="o">.</span><span class="n">qname</span> <span class="ow">in</span> <span class="n">queries</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># write out read names not in textfile</span>
            <span class="n">nunmapped</span> <span class="o">=</span> <span class="n">nunmapped</span> <span class="o">+</span> <span class="mi">1</span>
            <span class="n">osam</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">read</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">logger</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Wrote </span><span class="si">%i</span><span class="s2"> unmapped reads of the </span><span class="si">%i</span><span class="s2"> total from </span><span class="si">%s</span><span class="s2"> to </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                    <span class="n">nunmapped</span><span class="p">,</span> <span class="n">total</span><span class="p">,</span> <span class="n">sam</span><span class="p">,</span> <span class="n">ofile</span><span class="p">)</span>
    <span class="n">osam</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>


<div class="viewcode-block" id="partition_mapping"><a class="viewcode-back" href="../riboSeed.html#riboSeed.partition_mapping">[docs]</a><span class="k">def</span> <span class="nf">partition_mapping</span><span class="p">(</span><span class="n">seedGenome</span><span class="p">,</span> <span class="n">samtools_exe</span><span class="p">,</span> <span class="n">flank</span><span class="p">,</span> <span class="n">min_flank_depth</span><span class="p">,</span>
                      <span class="n">cluster_list</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Extract interesting stuff based on coords, not a binary</span>
<span class="sd">    mapped/not_mapped condition</span>
<span class="sd">    Also, if min_flanking_depth, mark reads with low</span>
<span class="sd">    mapping coverage for exclusion</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mapped_regions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;processing mapping for iteration </span><span class="si">%i</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">cluster_list</span><span class="p">:</span>
        <span class="n">prepare_next_mapping</span><span class="p">(</span><span class="n">cluster</span><span class="o">=</span><span class="n">cluster</span><span class="p">,</span> <span class="n">seedGenome</span><span class="o">=</span><span class="n">seedGenome</span><span class="p">,</span>
                             <span class="n">samtools_exe</span><span class="o">=</span><span class="n">samtools_exe</span><span class="p">,</span> <span class="n">flank</span><span class="o">=</span><span class="n">flank</span><span class="p">,</span>
                             <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>

    <span class="n">mapped_regions</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">all_depths</span> <span class="o">=</span> <span class="p">[]</span>  <span class="c1"># each entry is a tuple (idx, start_ave, end_ave)</span>
    <span class="n">filtered_cluster_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">cluster_list</span><span class="p">:</span>
        <span class="n">mapped_partition_cmds</span><span class="p">,</span> <span class="n">reg_to_extract</span> <span class="o">=</span> <span class="n">make_mapped_partition_cmds</span><span class="p">(</span>
            <span class="n">cluster</span><span class="o">=</span><span class="n">cluster</span><span class="p">,</span> <span class="n">mapping_ob</span><span class="o">=</span><span class="n">cluster</span><span class="o">.</span><span class="n">mappings</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span>
            <span class="n">seedGenome</span><span class="o">=</span><span class="n">seedGenome</span><span class="p">,</span> <span class="n">samtools_exe</span><span class="o">=</span><span class="n">samtools_exe</span><span class="p">,</span>  <span class="c1"># flank=flank,</span>
            <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">cmd</span> <span class="ow">in</span> <span class="n">mapped_partition_cmds</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
            <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="n">cmd</span><span class="p">],</span>
                           <span class="n">shell</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">platform</span> <span class="o">!=</span> <span class="s2">&quot;win32&quot;</span><span class="p">,</span>
                           <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                           <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                           <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">start_depths</span><span class="p">,</span> <span class="n">start_ave_depth</span> <span class="o">=</span> <span class="n">get_samtools_depths</span><span class="p">(</span>
            <span class="n">bam</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span>
                <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">]</span><span class="o">.</span><span class="n">sorted_mapped_bam</span><span class="p">,</span>
            <span class="n">chrom</span><span class="o">=</span><span class="n">cluster</span><span class="o">.</span><span class="n">sequence_id</span><span class="p">,</span>
            <span class="n">start</span><span class="o">=</span><span class="n">cluster</span><span class="o">.</span><span class="n">global_start_coord</span><span class="p">,</span>
            <span class="n">end</span><span class="o">=</span><span class="n">cluster</span><span class="o">.</span><span class="n">global_start_coord</span> <span class="o">+</span> <span class="n">flank</span><span class="p">,</span>
            <span class="n">region</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">prep</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">samtools_exe</span><span class="o">=</span><span class="n">samtools_exe</span><span class="p">,</span>
            <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
        <span class="n">end_depths</span><span class="p">,</span> <span class="n">end_ave_depth</span> <span class="o">=</span> <span class="n">get_samtools_depths</span><span class="p">(</span>
            <span class="n">bam</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span>
                <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">]</span><span class="o">.</span><span class="n">sorted_mapped_bam</span><span class="p">,</span>
            <span class="n">chrom</span><span class="o">=</span><span class="n">cluster</span><span class="o">.</span><span class="n">sequence_id</span><span class="p">,</span>
            <span class="n">start</span><span class="o">=</span><span class="n">cluster</span><span class="o">.</span><span class="n">global_end_coord</span> <span class="o">-</span> <span class="n">flank</span><span class="p">,</span>
            <span class="n">end</span><span class="o">=</span><span class="n">cluster</span><span class="o">.</span><span class="n">global_end_coord</span><span class="p">,</span>
            <span class="n">region</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
            <span class="n">prep</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">samtools_exe</span><span class="o">=</span><span class="n">samtools_exe</span><span class="p">,</span>
            <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Coverage for cluster &quot;</span> <span class="o">+</span>
                    <span class="s2">&quot;</span><span class="si">%i</span><span class="s2">:</span><span class="se">\n\t</span><span class="s2">5&#39; </span><span class="si">%i</span><span class="s2">bp-region: </span><span class="si">%.2f</span><span class="s2"> </span><span class="se">\n\t</span><span class="s2">3&#39; </span><span class="si">%i</span><span class="s2">bp-region: </span><span class="si">%.2f</span><span class="s2">&quot;</span><span class="p">,</span>
                    <span class="n">cluster</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>
                    <span class="n">flank</span><span class="p">,</span>
                    <span class="n">start_ave_depth</span><span class="p">,</span>
                    <span class="n">flank</span><span class="p">,</span>
                    <span class="n">end_ave_depth</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">start_ave_depth</span> <span class="o">&lt;</span> <span class="n">min_flank_depth</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;cluster </span><span class="si">{0}</span><span class="s2"> has insufficient 5&#39; flanking &quot;</span> <span class="o">+</span>
                           <span class="s2">&quot;coverage depth for subassembly, and will be &quot;</span> <span class="o">+</span>
                           <span class="s2">&quot;removed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cluster</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
            <span class="n">cluster</span><span class="o">.</span><span class="n">coverage_exclusion</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">elif</span> <span class="n">end_ave_depth</span> <span class="o">&lt;</span> <span class="n">min_flank_depth</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;cluster </span><span class="si">{0}</span><span class="s2"> has insufficient 3&#39; flanking &quot;</span> <span class="o">+</span>
                           <span class="s2">&quot;coverage depth for subassembly, and will be &quot;</span> <span class="o">+</span>
                           <span class="s2">&quot;removed&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">cluster</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
            <span class="n">cluster</span><span class="o">.</span><span class="n">coverage_exclusion</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mapped_regions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">reg_to_extract</span><span class="p">)</span>
            <span class="n">filtered_cluster_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cluster</span><span class="p">)</span>
        <span class="c1"># regardsless, report stats here</span>
        <span class="n">all_depths</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">cluster</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">start_ave_depth</span><span class="p">,</span> <span class="n">end_ave_depth</span><span class="p">))</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;mapped regions for iteration </span><span class="si">%i</span><span class="s2">:</span><span class="se">\n</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">,</span>
                <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">mapped_regions</span><span class="p">]))</span>
    <span class="c1"># make commands to extract all the reads NOT mapping to the rDNA regions</span>
    <span class="n">unmapped_partition_cmds</span> <span class="o">=</span> <span class="n">make_unmapped_partition_cmds</span><span class="p">(</span>
        <span class="n">mapped_regions</span><span class="o">=</span><span class="n">mapped_regions</span><span class="p">,</span> <span class="n">samtools_exe</span><span class="o">=</span><span class="n">samtools_exe</span><span class="p">,</span>
        <span class="n">seedGenome</span><span class="o">=</span><span class="n">seedGenome</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">cmd</span> <span class="ow">in</span> <span class="n">unmapped_partition_cmds</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
        <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="n">cmd</span><span class="p">],</span>
                       <span class="n">shell</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">platform</span> <span class="o">!=</span> <span class="s2">&quot;win32&quot;</span><span class="p">,</span>
                       <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                       <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                       <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;using pysam to extract a subset of reads &quot;</span><span class="p">)</span>
    <span class="c1"># this may look wierd: for iteration 0, we extract from the mapping.</span>
    <span class="c1"># for each one after that, we extract from the previous mapping.</span>
    <span class="n">unmapped_reads_index</span> <span class="o">=</span> <span class="mi">0</span> <span class="k">if</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span> <span class="o">==</span> <span class="mi">0</span> \
        <span class="k">else</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span> <span class="o">-</span> <span class="mi">1</span>
    <span class="n">pysam_extract_reads</span><span class="p">(</span>
        <span class="n">sam</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span><span class="n">unmapped_reads_index</span><span class="p">]</span><span class="o">.</span><span class="n">mapped_sam</span><span class="p">,</span>
        <span class="n">textfile</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span>
            <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">]</span><span class="o">.</span><span class="n">mapped_ids_txt</span><span class="p">,</span>
        <span class="n">unmapped_sam</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span>
            <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">]</span><span class="o">.</span><span class="n">unmapped_sam</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
    <span class="c1"># sam_score_list = get_sam_AS</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">all_depths</span><span class="p">,</span> <span class="n">filtered_cluster_list</span><span class="p">)</span></div>


<div class="viewcode-block" id="add_coords_to_clusters"><a class="viewcode-back" href="../riboSeed.html#riboSeed.add_coords_to_clusters">[docs]</a><span class="k">def</span> <span class="nf">add_coords_to_clusters</span><span class="p">(</span><span class="n">seedGenome</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; given a genbank file and some locus tags, add the coordinates, etc,</span>
<span class="sd">    to the entry in the seed Genome</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">loci_clusters</span><span class="p">:</span>  <span class="c1"># for each cluster of loci</span>
        <span class="c1"># get seq record that cluster is  from</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">cluster</span><span class="o">.</span><span class="n">seq_record</span> <span class="o">=</span> <span class="n">get_rec_from_generator</span><span class="p">(</span>
                <span class="n">recordID</span><span class="o">=</span><span class="n">cluster</span><span class="o">.</span><span class="n">sequence_id</span><span class="p">,</span>
                <span class="n">gen</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">seq_records</span><span class="p">,</span>
                <span class="n">method</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">refreshSeqRecGenerator</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">e</span>
        <span class="k">try</span><span class="p">:</span>  <span class="c1"># make coord list</span>
            <span class="n">extract_coords_from_locus</span><span class="p">(</span>
                <span class="n">cluster</span><span class="o">=</span><span class="n">cluster</span><span class="p">,</span> <span class="n">feature</span><span class="o">=</span><span class="n">cluster</span><span class="o">.</span><span class="n">feat_of_interest</span><span class="p">,</span>
                <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">raise</span> <span class="n">e</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Here are the detected region,coords, strand, product, &quot;</span> <span class="o">+</span>
                     <span class="s2">&quot;locus tag, subfeatures and sequence id of the results:&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">cluster</span><span class="o">.</span><span class="vm">__dict__</span><span class="p">))</span></div>


<div class="viewcode-block" id="get_final_assemblies_cmds"><a class="viewcode-back" href="../riboSeed.html#riboSeed.get_final_assemblies_cmds">[docs]</a><span class="k">def</span> <span class="nf">get_final_assemblies_cmds</span><span class="p">(</span><span class="n">seedGenome</span><span class="p">,</span> <span class="n">exes</span><span class="p">,</span>
                              <span class="n">ref_as_contig</span><span class="p">,</span>
                              <span class="n">cores</span><span class="p">,</span>
                              <span class="n">memory</span><span class="p">,</span>
                              <span class="n">serialize</span><span class="p">,</span>
                              <span class="n">skip_control</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                              <span class="n">kmers</span><span class="o">=</span><span class="s2">&quot;21,33,55,77,99&quot;</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;make cmds for runnning of SPAdes and QUAST final assembly and analysis.</span>
<span class="sd">    if skip_control, just do the de fere novo assembly.  otherwise, do bother</span>
<span class="sd">    returns list of listed cmds</span>
<span class="sd">    ([[spades_cmd, quast_cmd], [spades_cmd2, quast_cmd2]])</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n\n</span><span class="s2">Starting Final Assemblies</span><span class="se">\n\n</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">quast_reports</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">cmd_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">final_list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;de_fere_novo&quot;</span><span class="p">]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">skip_control</span><span class="p">:</span>
        <span class="n">final_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;de_novo&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">final_list</span><span class="p">:</span>
        <span class="n">final_mapping</span> <span class="o">=</span> <span class="n">LociMapping</span><span class="p">(</span>
            <span class="n">iteration</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
            <span class="n">name</span><span class="o">=</span><span class="n">j</span><span class="p">,</span>
            <span class="n">mapping_subdir</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">seedGenome</span><span class="o">.</span><span class="n">output_root</span><span class="p">,</span>
                <span class="s2">&quot;final_</span><span class="si">{0}</span><span class="s2">_mapping&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">j</span><span class="p">)),</span>
            <span class="n">assembly_subdir_needed</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">assembly_subdir</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">seedGenome</span><span class="o">.</span><span class="n">output_root</span><span class="p">,</span>
                <span class="s2">&quot;final_</span><span class="si">{0}</span><span class="s2">_assembly&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">j</span><span class="p">)))</span>
        <span class="c1"># logger.info(&quot;\n\nRunning %s SPAdes \n&quot; % j)</span>
        <span class="k">if</span> <span class="n">j</span> <span class="o">==</span> <span class="s2">&quot;de_novo&quot;</span><span class="p">:</span>
            <span class="n">final_mapping</span><span class="o">.</span><span class="n">ref_fasta</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
            <span class="n">assembly_ref_as_contig</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">j</span> <span class="o">==</span> <span class="s2">&quot;de_fere_novo&quot;</span><span class="p">,</span> \
                <span class="s2">&quot;Only valid cases are de novo and de fere novo!&quot;</span>
            <span class="n">final_mapping</span><span class="o">.</span><span class="n">ref_fasta</span> <span class="o">=</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">assembled_seeds</span>
            <span class="n">assembly_ref_as_contig</span> <span class="o">=</span> <span class="n">ref_as_contig</span>

        <span class="c1"># remove unneeded dir</span>
        <span class="n">os</span><span class="o">.</span><span class="n">rmdir</span><span class="p">(</span><span class="n">final_mapping</span><span class="o">.</span><span class="n">mapping_subdir</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Getting commands for </span><span class="si">%s</span><span class="s2"> SPAdes&quot;</span> <span class="o">%</span> <span class="n">j</span><span class="p">)</span>
        <span class="n">spades_cmd</span> <span class="o">=</span> <span class="n">generate_spades_cmd</span><span class="p">(</span>
            <span class="n">single_lib</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">master_ngs_ob</span><span class="o">.</span><span class="n">libtype</span> <span class="o">==</span> <span class="s2">&quot;s_1&quot;</span><span class="p">,</span>
            <span class="n">check_libs</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">mapping_ob</span><span class="o">=</span><span class="n">final_mapping</span><span class="p">,</span> <span class="n">ngs_ob</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">master_ngs_ob</span><span class="p">,</span>
            <span class="n">ref_as_contig</span><span class="o">=</span><span class="n">assembly_ref_as_contig</span><span class="p">,</span> <span class="n">as_paired</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">prelim</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
            <span class="n">k</span><span class="o">=</span><span class="n">kmers</span><span class="p">,</span> <span class="n">spades_exe</span><span class="o">=</span><span class="n">exes</span><span class="o">.</span><span class="n">spades</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
        <span class="n">modest_spades_cmd</span> <span class="o">=</span> <span class="n">make_modest_spades_cmd</span><span class="p">(</span>
            <span class="n">cmd</span><span class="o">=</span><span class="n">spades_cmd</span><span class="p">,</span> <span class="n">cores</span><span class="o">=</span><span class="n">cores</span><span class="p">,</span> <span class="n">memory</span><span class="o">=</span><span class="n">memory</span><span class="p">,</span> <span class="n">split</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
            <span class="n">serialize</span><span class="o">=</span><span class="n">serialize</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
        <span class="n">ref</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;-R </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">ref_fasta</span><span class="p">)</span>
        <span class="n">quast_cmd</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2"> </span><span class="si">{1}</span><span class="s2"> </span><span class="si">{2}</span><span class="s2"> </span><span class="si">{3}</span><span class="s2"> -o </span><span class="si">{4}</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="n">exes</span><span class="o">.</span><span class="n">python2_7</span><span class="p">,</span>
            <span class="n">exes</span><span class="o">.</span><span class="n">quast</span><span class="p">,</span>
            <span class="n">ref</span><span class="p">,</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">final_mapping</span><span class="o">.</span><span class="n">assembly_subdir</span><span class="p">,</span> <span class="s2">&quot;contigs.fasta&quot;</span><span class="p">),</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">output_root</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;quast_&quot;</span> <span class="o">+</span> <span class="n">j</span><span class="p">)))</span>
        <span class="n">quast_reports</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">output_root</span><span class="p">,</span>
                                          <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;quast_&quot;</span> <span class="o">+</span> <span class="n">j</span><span class="p">),</span> <span class="s2">&quot;report.tsv&quot;</span><span class="p">))</span>
        <span class="n">cmd_list</span><span class="o">.</span><span class="n">append</span><span class="p">([</span><span class="n">modest_spades_cmd</span><span class="p">,</span> <span class="n">quast_cmd</span><span class="p">])</span>
    <span class="k">return</span><span class="p">(</span><span class="n">cmd_list</span><span class="p">,</span> <span class="n">quast_reports</span><span class="p">)</span></div>


<div class="viewcode-block" id="make_faux_genome"><a class="viewcode-back" href="../riboSeed.html#riboSeed.make_faux_genome">[docs]</a><span class="k">def</span> <span class="nf">make_faux_genome</span><span class="p">(</span><span class="n">cluster_list</span><span class="p">,</span> <span class="n">seedGenome</span><span class="p">,</span> <span class="n">iteration</span><span class="p">,</span>
                     <span class="n">output_root</span><span class="p">,</span> <span class="n">nbuff</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; stictch together viable assembled contigs.  perhaps more importnatly,</span>
<span class="sd">    this also re-write thes coords relative to the new &quot;genome&quot;</span>
<span class="sd">    returns path to new faux_genome</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;preparing extracted region genome for next round of mapping&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;using </span><span class="si">%i</span><span class="s2"> sequences&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">cluster_list</span><span class="p">))</span>
    <span class="n">nbuffer</span> <span class="o">=</span> <span class="s2">&quot;N&quot;</span> <span class="o">*</span> <span class="n">nbuff</span>
    <span class="n">faux_genome</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">new_seq_name</span> <span class="o">=</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">name</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cluster_list</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="mi">1</span>
    <span class="k">for</span> <span class="n">clu</span> <span class="ow">in</span> <span class="n">cluster_list</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">clu</span><span class="o">.</span><span class="n">keep_contigs</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">clu</span><span class="o">.</span><span class="n">continue_iterating</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">clu</span><span class="o">.</span><span class="n">global_start_coord</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">faux_genome</span><span class="p">)</span> <span class="o">+</span> <span class="n">nbuff</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">clu</span><span class="o">.</span><span class="n">mappings</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">assembled_contig</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">con</span><span class="p">:</span>
                <span class="n">contig_rec</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">con</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">faux_genome</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">faux_genome</span> <span class="o">+</span> <span class="n">nbuffer</span> <span class="o">+</span> <span class="n">contig_rec</span><span class="o">.</span><span class="n">seq</span><span class="p">)</span>
            <span class="n">clu</span><span class="o">.</span><span class="n">global_end_coord</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">faux_genome</span><span class="p">)</span>
            <span class="c1"># lastly, set cluster name to new sequence name</span>
            <span class="n">clu</span><span class="o">.</span><span class="n">sequence_id</span> <span class="o">=</span> <span class="n">new_seq_name</span>
            <span class="n">counter</span> <span class="o">=</span> <span class="n">counter</span> <span class="o">+</span> <span class="mi">1</span>
    <span class="k">if</span> <span class="n">counter</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;No viable contigs for faux genome construction!&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="mi">1</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;combined </span><span class="si">%s</span><span class="s2"> records as genome for next round of mapping&quot;</span><span class="p">,</span>
                    <span class="n">counter</span><span class="p">)</span>
    <span class="n">record</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="n">Seq</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">faux_genome</span> <span class="o">+</span> <span class="n">nbuffer</span><span class="p">),</span>
                           <span class="n">IUPAC</span><span class="o">.</span><span class="n">IUPACAmbiguousDNA</span><span class="p">()),</span>
                       <span class="nb">id</span><span class="o">=</span><span class="n">new_seq_name</span><span class="p">)</span>

    <span class="n">outpath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_root</span><span class="p">,</span>
                           <span class="s2">&quot;iter_</span><span class="si">{0}</span><span class="s2">_buffered_genome.fasta&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">iteration</span><span class="p">))</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outpath</span><span class="p">,</span> <span class="s1">&#39;w&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outf</span><span class="p">:</span>
        <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">outf</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">outpath</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="p">))</span></div>


<div class="viewcode-block" id="decide_proceed_to_target"><a class="viewcode-back" href="../riboSeed.html#riboSeed.decide_proceed_to_target">[docs]</a><span class="k">def</span> <span class="nf">decide_proceed_to_target</span><span class="p">(</span><span class="n">target_len</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="k">assert</span> <span class="n">logger</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Must use logging!&quot;</span>
    <span class="k">if</span> <span class="n">target_len</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">target_len</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">or</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">target_len</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;--target_len is set to invalid value! Must be a &quot;</span> <span class="o">+</span>
                         <span class="s2">&quot;decimal greater than zero, ie where 1.1 would be &quot;</span> <span class="o">+</span>
                         <span class="s2">&quot;110</span><span class="si">% o</span><span class="s2">f the original sequence length.&quot;</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span>
        <span class="k">elif</span> <span class="n">target_len</span> <span class="o">&gt;</span> <span class="mi">5</span> <span class="ow">and</span> <span class="mi">50</span> <span class="o">&gt;</span> <span class="n">target_len</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;We dont reccommend seeding to lengths greater than&quot;</span> <span class="o">+</span>
                         <span class="s2">&quot;5x original seed length. Try between 0.5 and 1.5.&quot;</span> <span class="o">+</span>
                         <span class="s2">&quot;  If you are setting a target number of bases, it &quot;</span> <span class="o">+</span>
                         <span class="s2">&quot; must be greater than 50&quot;</span><span class="p">)</span>
            <span class="k">raise</span> <span class="ne">ValueError</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">proceed_to_target</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">proceed_to_target</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">return</span> <span class="n">proceed_to_target</span></div>


<div class="viewcode-block" id="subprocess_run_list"><a class="viewcode-back" href="../riboSeed.html#riboSeed.subprocess_run_list">[docs]</a><span class="k">def</span> <span class="nf">subprocess_run_list</span><span class="p">(</span><span class="n">cmdlist</span><span class="p">,</span> <span class="n">hard</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; This just allows for sequential cmds with multiprocessing.</span>
<span class="sd">    It prevents the errors when future commands are looking for and not finding</span>
<span class="sd">    a needed file.</span>
<span class="sd">    Logger cant be used with multiprocessing</span>
<span class="sd">    returns 0 if all is well, otherwise returns 1</span>
<span class="sd">    if hard == True, quits instead of returning 1</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">for</span> <span class="n">cmd</span> <span class="ow">in</span> <span class="n">cmdlist</span><span class="p">:</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="n">cmd</span><span class="p">],</span>
                           <span class="n">shell</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">platform</span> <span class="o">!=</span> <span class="s2">&quot;win32&quot;</span><span class="p">,</span>
                           <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                           <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                           <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">logger</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">hard</span><span class="p">:</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">return</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="mi">0</span></div>


<div class="viewcode-block" id="copyToHandyDir"><a class="viewcode-back" href="../riboSeed.html#riboSeed.copyToHandyDir">[docs]</a><span class="k">def</span> <span class="nf">copyToHandyDir</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="n">pre</span><span class="p">,</span> <span class="n">seedGenome</span><span class="p">,</span> <span class="n">hard</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; copy the resulting contigs</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">logger</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Must use logging&quot;</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">outdir</span><span class="p">)</span>
    <span class="n">files_to_copy</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">output_root</span><span class="p">,</span>
                     <span class="s2">&quot;final_de_fere_novo_assembly&quot;</span><span class="p">,</span> <span class="s2">&quot;contigs.fasta&quot;</span><span class="p">),</span>
        <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">output_root</span><span class="p">,</span>
                     <span class="s2">&quot;final_de_novo_assembly&quot;</span><span class="p">,</span> <span class="s2">&quot;contigs.fasta&quot;</span><span class="p">),</span>
        <span class="n">args</span><span class="o">.</span><span class="n">reference_genbank</span><span class="p">]</span>
    <span class="n">new_names</span> <span class="o">=</span> <span class="p">[</span><span class="n">pre</span> <span class="o">+</span> <span class="s2">&quot;_de_fere_novo_contigs.fasta&quot;</span><span class="p">,</span>
                 <span class="n">pre</span> <span class="o">+</span> <span class="s2">&quot;_de_novo_contigs.fasta&quot;</span><span class="p">,</span>
                 <span class="n">pre</span> <span class="o">+</span> <span class="s2">&quot;.gb&quot;</span><span class="p">]</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">f</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">files_to_copy</span><span class="p">):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;copying </span><span class="si">%s</span><span class="s2"> to </span><span class="si">%s</span><span class="s2"> as </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                     <span class="n">f</span><span class="p">,</span>
                     <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">f</span><span class="p">)),</span>
                     <span class="n">new_names</span><span class="p">[</span><span class="n">idx</span><span class="p">])</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">shutil</span><span class="o">.</span><span class="n">copyfile</span><span class="p">(</span>
                <span class="n">f</span><span class="p">,</span>
                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="n">new_names</span><span class="p">[</span><span class="n">idx</span><span class="p">]))</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;unable to copy </span><span class="si">%s</span><span class="s2"> to </span><span class="si">%s</span><span class="s2">.&quot;</span><span class="p">,</span>
                           <span class="n">f</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">f</span><span class="p">)))</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">last_exception</span><span class="p">())</span>
            <span class="k">if</span> <span class="n">hard</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">FileNotFoundError</span></div>


<div class="viewcode-block" id="printPlot"><a class="viewcode-back" href="../riboSeed.html#riboSeed.printPlot">[docs]</a><span class="k">def</span> <span class="nf">printPlot</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">line</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ymax</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">xmax</span><span class="o">=</span><span class="mi">60</span><span class="p">,</span> <span class="n">tick</span><span class="o">=.</span><span class="mi">2</span><span class="p">,</span>
              <span class="n">title</span><span class="o">=</span><span class="s2">&quot;test&quot;</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; ascii not what your program can do for you...</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">logger</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;must use logging&quot;</span>
    <span class="n">data</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">xaxis</span> <span class="o">=</span> <span class="s2">&quot;|&quot;</span>
    <span class="n">yaxis</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span>
    <span class="n">avbin</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">scaledy</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># ylab = str(max(data))</span>
    <span class="n">ylab</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">))</span>
    <span class="n">sli</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">ceil</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="p">)</span> <span class="o">/</span> <span class="n">ymax</span><span class="p">)</span>
    <span class="c1">#  get rough averages for a window</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">ymax</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>
        <span class="n">avbin</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="nb">sum</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">i</span> <span class="o">*</span> <span class="n">sli</span><span class="p">:</span> <span class="p">(</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="n">sli</span><span class="p">])</span> <span class="o">/</span> <span class="n">sli</span><span class="p">))</span>

    <span class="n">avmax</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">avbin</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;scaling to max of </span><span class="si">%i</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">avmax</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">avbin</span><span class="p">:</span>
        <span class="n">scaledy</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">int</span><span class="p">((</span><span class="n">j</span> <span class="o">/</span> <span class="n">avmax</span><span class="p">)</span> <span class="o">*</span> <span class="n">xmax</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">line</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">scaled_line</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">line</span> <span class="o">/</span> <span class="n">avmax</span><span class="p">)</span> <span class="o">*</span> <span class="n">xmax</span><span class="p">)</span>
        <span class="n">lineidx</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">scaledy</span><span class="p">)</span> <span class="o">-</span> <span class="n">bisect</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="n">scaledy</span><span class="p">),</span> <span class="n">scaled_line</span><span class="p">)</span>
    <span class="n">plotlines</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">fillchar</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span> <span class="k">if</span> <span class="ow">not</span> <span class="n">fill</span> <span class="k">else</span> <span class="s2">&quot;X&quot;</span>
    <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">scaledy</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">idx</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">plotlines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot; &quot;</span> <span class="o">*</span> <span class="nb">int</span><span class="p">(</span><span class="n">xmax</span> <span class="o">*</span> <span class="o">.</span><span class="mi">25</span><span class="p">)</span> <span class="o">+</span> <span class="n">title</span><span class="p">)</span>
            <span class="n">plotlines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot; &quot;</span> <span class="o">*</span> <span class="mi">9</span> <span class="o">+</span> <span class="s2">&quot;0&quot;</span> <span class="o">+</span> <span class="s2">&quot; &quot;</span> <span class="o">*</span> <span class="p">(</span><span class="n">xmax</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span> <span class="o">+</span> <span class="n">ylab</span><span class="p">)</span>
            <span class="n">plotlines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot; &quot;</span> <span class="o">*</span> <span class="mi">10</span> <span class="o">+</span> <span class="n">xaxis</span> <span class="o">+</span> <span class="p">(</span><span class="n">yaxis</span> <span class="o">*</span> <span class="n">xmax</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;|&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">line</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">lineidx</span> <span class="o">==</span> <span class="n">idx</span><span class="p">:</span>
                <span class="n">plotlines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot; &quot;</span> <span class="o">*</span> <span class="mi">10</span> <span class="o">+</span> <span class="n">xaxis</span> <span class="o">+</span>
                                 <span class="s2">&quot;*&quot;</span> <span class="o">*</span> <span class="p">(</span><span class="n">xmax</span> <span class="o">-</span> <span class="mi">4</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;  &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">line</span><span class="p">))</span>
                <span class="n">fillchar</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span>
        <span class="k">if</span> <span class="n">idx</span> <span class="o">%</span> <span class="nb">int</span><span class="p">(</span><span class="n">tick</span> <span class="o">*</span> <span class="n">ymax</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># plot ticks at increments</span>
            <span class="n">ticlab</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="nb">int</span><span class="p">((</span><span class="n">j</span> <span class="o">/</span> <span class="n">xmax</span><span class="p">)</span> <span class="o">*</span> <span class="n">avmax</span><span class="p">))</span>
            <span class="n">plotlines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ticlab</span><span class="o">.</span><span class="n">rjust</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="s2">&quot; &quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">xaxis</span> <span class="o">+</span> <span class="n">fillchar</span> <span class="o">*</span>
                             <span class="n">j</span> <span class="o">+</span> <span class="s2">&quot;O&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">plotlines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot; &quot;</span> <span class="o">*</span> <span class="mi">10</span> <span class="o">+</span> <span class="n">xaxis</span> <span class="o">+</span> <span class="n">fillchar</span> <span class="o">*</span> <span class="n">j</span> <span class="o">+</span> <span class="s2">&quot;O&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">plotlines</span><span class="p">))</span></div>


<div class="viewcode-block" id="plotAsScores"><a class="viewcode-back" href="../riboSeed.html#riboSeed.plotAsScores">[docs]</a><span class="k">def</span> <span class="nf">plotAsScores</span><span class="p">(</span><span class="n">score_list</span><span class="p">,</span> <span class="n">score_min</span><span class="p">,</span> <span class="n">outdir</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

    <span class="k">assert</span> <span class="n">logger</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;must use logging&quot;</span>
    <span class="n">basename</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="s2">&quot;AS_score_plot&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">score_list</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">200000</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Downsampling our pltting data to 20k points&quot;</span><span class="p">)</span>
        <span class="n">score_list</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">score_list</span><span class="p">,</span> <span class="mi">200000</span><span class="p">)</span>
    <span class="n">xmax</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">score_list</span><span class="p">)</span>
    <span class="n">ymax</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">score_list</span><span class="p">),</span> <span class="n">key</span><span class="o">=</span><span class="n">score_list</span><span class="o">.</span><span class="n">count</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;score_min, xmax and ymax: </span><span class="si">%s</span><span class="s2">, </span><span class="si">%s</span><span class="s2">, </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">score_min</span><span class="p">,</span> <span class="n">xmax</span><span class="p">,</span> <span class="n">ymax</span><span class="p">)</span>
    <span class="c1"># plt.figure()  # &lt;- makes a new figure and sets it active (add this)</span>
    <span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">plt1</span><span class="p">,</span> <span class="n">plt2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">sharex</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="c1"># ,</span>
    <span class="c1">#                                  gridspec_kw={&#39;height_ratios&#39;: [1, 1]})</span>
    <span class="n">plt1</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">score_list</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;b&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.9</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Binned&#39;</span><span class="p">)</span>
    <span class="n">plt1</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span><span class="n">score_list</span><span class="p">,</span> <span class="n">bins</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
              <span class="n">cumulative</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
              <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Cumulative&#39;</span><span class="p">)</span>
    <span class="n">plt1</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Read Alignment Score Histogram&#39;</span><span class="p">)</span>
    <span class="n">plt1</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Alignemnt Score&#39;</span><span class="p">)</span>
    <span class="n">plt1</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Abundance&#39;</span><span class="p">)</span>
    <span class="n">plt1</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="n">score_min</span><span class="p">,</span> <span class="n">score_min</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">score_list</span><span class="p">)</span> <span class="o">*</span> <span class="mf">1.1</span><span class="p">],</span>
              <span class="n">color</span><span class="o">=</span><span class="s1">&#39;green&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>
    <span class="n">plt1</span><span class="o">.</span><span class="n">axis</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="n">score_list</span><span class="p">),</span> <span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">score_list</span><span class="p">)</span> <span class="o">*</span> <span class="mf">1.1</span><span class="p">])</span>
    <span class="c1"># plt1.set_yscale(&quot;log&quot;, nonposy=&#39;clip&#39;)</span>
    <span class="n">plt1</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">tight_layout</span><span class="p">()</span>
    <span class="n">plt1</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt2</span><span class="o">.</span><span class="n">grid</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">plt2</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">y</span><span class="o">=</span><span class="nb">sorted</span><span class="p">(</span><span class="n">score_list</span><span class="p">,</span> <span class="n">reverse</span><span class="o">=</span><span class="kc">True</span><span class="p">),</span>
                 <span class="n">x</span><span class="o">=</span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">score_list</span><span class="p">)))</span>
    <span class="n">plt2</span><span class="o">.</span><span class="n">plot</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">score_list</span><span class="p">)</span> <span class="o">*</span> <span class="mf">1.1</span><span class="p">],</span> <span class="p">[</span><span class="n">score_min</span><span class="p">,</span> <span class="n">score_min</span><span class="p">],</span>
              <span class="n">color</span><span class="o">=</span><span class="s1">&#39;green&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.6</span><span class="p">)</span>
    <span class="n">plt2</span><span class="o">.</span><span class="n">axis</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">score_list</span><span class="p">)</span> <span class="o">*</span> <span class="mf">1.1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="nb">max</span><span class="p">(</span><span class="n">score_list</span><span class="p">)</span> <span class="o">*</span> <span class="mf">1.1</span><span class="p">])</span>
    <span class="n">plt2</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;Read Alignment Score, Sorted&#39;</span><span class="p">)</span>
    <span class="n">plt2</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Alignment Score&#39;</span><span class="p">)</span>
    <span class="n">plt2</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Index of Sorted Read&#39;</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="mf">7.5</span><span class="p">)</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">basename</span> <span class="o">+</span> <span class="s1">&#39;.png&#39;</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="p">(</span><span class="mi">200</span><span class="p">))</span>
    <span class="n">fig</span><span class="o">.</span><span class="n">savefig</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">basename</span> <span class="o">+</span> <span class="s1">&#39;.pdf&#39;</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="p">(</span><span class="mi">200</span><span class="p">))</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Plotting alignment score of mapping:&quot;</span><span class="p">)</span>
    <span class="c1"># printPlot(data=score_list, line=score_min, ymax=30, xmax=60, tick=.2,</span>
    <span class="c1">#           fill=True,</span>
    <span class="c1">#           title=&quot;Average alignment Scores (y) by sorted read index (x)&quot;,</span>
    <span class="c1">#           logger=logger)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Filled area represents the reads retained after filtering. &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;If it looks like &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;this filtering threshold is inapporpriate, consider &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;adjusting with --score_min&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="reportRegionDepths"><a class="viewcode-back" href="../riboSeed.html#riboSeed.reportRegionDepths">[docs]</a><span class="k">def</span> <span class="nf">reportRegionDepths</span><span class="p">(</span><span class="n">inp</span><span class="p">,</span> <span class="n">logger</span><span class="p">):</span>
    <span class="k">assert</span> <span class="n">logger</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;must use logging&quot;</span>
    <span class="n">report_list</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">nclusters</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">inp</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">nclusters</span><span class="p">):</span>
        <span class="n">report_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Cluster </span><span class="si">%i</span><span class="s2">:&quot;</span> <span class="o">%</span> <span class="n">i</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">itidx</span><span class="p">,</span> <span class="n">iteration</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">inp</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">iteration</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">cluster</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">i</span><span class="p">:</span>
                    <span class="n">report_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                        <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">Iter </span><span class="si">%i</span><span class="s2"> -- 5&#39; coverage: </span><span class="si">%.2f</span><span class="s2">  3&#39; coverage </span><span class="si">%.2f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span>
                            <span class="n">itidx</span><span class="p">,</span> <span class="n">cluster</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">cluster</span><span class="p">[</span><span class="mi">2</span><span class="p">]))</span>
    <span class="k">return</span><span class="p">(</span><span class="n">report_list</span><span class="p">)</span></div>


<div class="viewcode-block" id="make_modest_spades_cmd"><a class="viewcode-back" href="../riboSeed.html#riboSeed.make_modest_spades_cmd">[docs]</a><span class="k">def</span> <span class="nf">make_modest_spades_cmd</span><span class="p">(</span><span class="n">cmd</span><span class="p">,</span> <span class="n">cores</span><span class="p">,</span> <span class="n">memory</span><span class="p">,</span> <span class="n">split</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
                           <span class="n">serialize</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; adjust spades commands to use set amounts of cores and memory</span>
<span class="sd">    returns the command, split on &quot;--careful&quot;, cause why would you run</span>
<span class="sd">    SPAdes with &quot;--reckless&quot;?</span>
<span class="sd">    if you need to split resouces between, say, two commands, use split 2</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">assert</span> <span class="n">logger</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">,</span> <span class="s2">&quot;Must use logging&quot;</span>
    <span class="n">cmdA</span><span class="p">,</span> <span class="n">cmdB</span> <span class="o">=</span> <span class="n">cmd</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;--careful&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">serialize</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Allocating SPAdes </span><span class="si">%d</span><span class="s2">gb of memory&quot;</span><span class="p">,</span> <span class="n">memory</span><span class="p">)</span>
        <span class="n">mem_each</span> <span class="o">=</span> <span class="n">memory</span>
        <span class="n">cores_each</span> <span class="o">=</span> <span class="n">cores</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">split</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">split</span> <span class="o">&gt;</span> <span class="n">cores</span> <span class="ow">or</span> <span class="n">split</span> <span class="o">&gt;</span> <span class="n">memory</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;cannot split cores or memory resources into &quot;</span> <span class="o">+</span>
                             <span class="s2">&quot;values less than 1!&quot;</span><span class="p">)</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
            <span class="n">mem_each</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">memory</span> <span class="o">/</span> <span class="n">split</span><span class="p">)</span>
            <span class="n">cores_each</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cores</span> <span class="o">/</span> <span class="n">split</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running spades with </span><span class="si">%d</span><span class="s2"> cores and </span><span class="si">%d</span><span class="s2"> gb memory&quot;</span><span class="p">,</span>
                        <span class="n">cores_each</span><span class="p">,</span> <span class="n">mem_each</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># make sure spades doesnt hog processors or ram</span>
            <span class="n">mem_each</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">memory</span> <span class="o">/</span> <span class="n">cores</span><span class="p">)</span>  <span class="c1"># should be floor</span>
            <span class="n">cores_each</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;Allocating SPAdes </span><span class="si">%d</span><span class="s2">gb of memory for each of </span><span class="si">%d</span><span class="s2"> cores&quot;</span><span class="p">,</span>
                <span class="n">memory</span><span class="p">,</span> <span class="n">cores</span><span class="p">)</span>
    <span class="k">return</span> <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">-t </span><span class="si">{1}</span><span class="s2"> -m </span><span class="si">{2}</span><span class="s2"> --careful</span><span class="si">{3}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">cmdA</span><span class="p">,</span>
        <span class="n">cores_each</span><span class="p">,</span>
        <span class="n">mem_each</span><span class="p">,</span>
        <span class="n">cmdB</span><span class="p">)</span></div>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">&quot;__main__&quot;</span><span class="p">:</span>
    <span class="n">args</span> <span class="o">=</span> <span class="n">get_args</span><span class="p">()</span>
    <span class="c1"># allow user to give relative paths</span>
    <span class="n">output_root</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">output</span><span class="p">))</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">output_root</span><span class="p">,</span> <span class="n">exist_ok</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">OSError</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Output directory already exists; exiting...&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">t0</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="n">log_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_root</span><span class="p">,</span> <span class="s2">&quot;riboSeed.log&quot;</span><span class="p">)</span>

    <span class="n">logger</span> <span class="o">=</span> <span class="n">set_up_logging</span><span class="p">(</span><span class="n">verbosity</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">verbosity</span><span class="p">,</span>
                            <span class="n">outfile</span><span class="o">=</span><span class="n">log_path</span><span class="p">,</span>
                            <span class="n">name</span><span class="o">=</span><span class="vm">__name__</span><span class="p">)</span>
    <span class="c1"># # log version of riboSeed, commandline options, and all settings</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;riboSeed pipeline package version: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="n">__version__</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Usage:</span><span class="se">\n</span><span class="si">{0}</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">])))</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;All settings used:&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">vars</span><span class="p">(</span><span class="n">args</span><span class="p">)</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">: </span><span class="si">{1}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="n">v</span><span class="p">))</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;current PATH:&quot;</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">environ</span><span class="p">[</span><span class="s1">&#39;PATH&#39;</span><span class="p">])</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;no PATH variable found in system environment.&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">cores</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">args</span><span class="o">.</span><span class="n">cores</span> <span class="o">=</span> <span class="n">multiprocessing</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Using </span><span class="si">%i</span><span class="s2"> cores&quot;</span><span class="p">,</span> <span class="n">multiprocessing</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">())</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;checking for installations of all required external tools&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;creating an Exes object&quot;</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">sys_exes</span> <span class="o">=</span> <span class="n">Exes</span><span class="p">(</span><span class="n">samtools</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">samtools_exe</span><span class="p">,</span>
                        <span class="n">spades</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">spades_exe</span><span class="p">,</span>
                        <span class="n">bwa</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">bwa_exe</span><span class="p">,</span>
                        <span class="n">smalt</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">smalt_exe</span><span class="p">,</span>
                        <span class="n">quast</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">quast_exe</span><span class="p">,</span>
                        <span class="n">python2_7</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">python2_7_exe</span><span class="p">,</span>
                        <span class="n">method</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">method</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;All needed system executables found!&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">sys_exes</span><span class="o">.</span><span class="vm">__dict__</span><span class="p">))</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">samtools_verison</span> <span class="o">=</span> <span class="n">check_version_from_cmd</span><span class="p">(</span>
            <span class="n">exe</span><span class="o">=</span><span class="n">sys_exes</span><span class="o">.</span><span class="n">samtools</span><span class="p">,</span> <span class="n">cmd</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">line</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">where</span><span class="o">=</span><span class="s1">&#39;stderr&#39;</span><span class="p">,</span>
            <span class="n">pattern</span><span class="o">=</span><span class="sa">r</span><span class="s2">&quot;\s*Version: (?P&lt;version&gt;[^(]+)&quot;</span><span class="p">,</span>
            <span class="n">min_version</span><span class="o">=</span><span class="n">SAMTOOLS_MIN_VERSION</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;samtools version: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">samtools_verison</span><span class="p">)</span>
    <span class="c1"># check bambamc is installed proper if using smalt</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;smalt&quot;</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;SMALT is the selected mapper&quot;</span><span class="p">)</span>
        <span class="n">test_smalt_cmds</span> <span class="o">=</span> <span class="n">get_smalt_full_install_cmds</span><span class="p">(</span><span class="n">smalt_exe</span><span class="o">=</span><span class="n">sys_exes</span><span class="o">.</span><span class="n">smalt</span><span class="p">,</span>
                                                      <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
        <span class="n">test_smalt_bam_install</span><span class="p">(</span><span class="n">cmds</span><span class="o">=</span><span class="n">test_smalt_cmds</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;BWA is the selected mapper&quot;</span><span class="p">)</span>

    <span class="c1"># if the target_len is set. set needed params</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">proceed_to_target</span> <span class="o">=</span> <span class="n">decide_proceed_to_target</span><span class="p">(</span>
            <span class="n">target_len</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">target_len</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Exiting&quot;</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># check and warn user about potential RAM issues</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">memory</span> <span class="o">&lt;</span> <span class="mi">6</span> <span class="ow">or</span> <span class="nb">int</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">memory</span> <span class="o">/</span> <span class="n">args</span><span class="o">.</span><span class="n">cores</span><span class="p">)</span> <span class="o">&lt;</span> <span class="mi">6</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;Danger!  We recommend that you have a minimum of &quot;</span> <span class="o">+</span>
                       <span class="s2">&quot;6GB memory (or 6GB memory per core is using &quot;</span> <span class="o">+</span>
                       <span class="s2">&quot;multiprocessing) available. If you have less &quot;</span> <span class="o">+</span>
                       <span class="s2">&quot;than that per core, SPAdes may run out of memory.&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;You can continue as configured if needed, and if a &quot;</span> <span class="o">+</span>
                       <span class="s2">&quot;SPAdes error occurs, you can still use the long &quot;</span> <span class="o">+</span>
                       <span class="s2">&quot;reads generated by riboSeed in a standalone assembly&quot;</span><span class="p">)</span>
<span class="c1"># --------------------------------------------------------------------------- #</span>
<span class="c1"># --------------------------------------------------------------------------- #</span>

    <span class="c1"># make seedGenome object</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;constructing the seedGenome object&quot;</span><span class="p">)</span>
    <span class="n">seedGenome</span> <span class="o">=</span> <span class="n">SeedGenome</span><span class="p">(</span>
        <span class="n">name</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">splitext</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">reference_genbank</span><span class="p">)[</span><span class="mi">0</span><span class="p">]),</span>
        <span class="c1"># this needs to be zero indexed to access mappings by iter</span>
        <span class="n">this_iteration</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
        <span class="n">iter_mapping_list</span><span class="o">=</span><span class="p">[],</span>
        <span class="n">max_iterations</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">iterations</span><span class="p">,</span>
        <span class="n">clustered_loci_txt</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">clustered_loci_txt</span><span class="p">,</span>
        <span class="n">output_root</span><span class="o">=</span><span class="n">output_root</span><span class="p">,</span>
        <span class="n">unmapped_mapping_list</span><span class="o">=</span><span class="p">[],</span>
        <span class="n">genbank_path</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">reference_genbank</span><span class="p">,</span>
        <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>

    <span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">ref_fasta</span> <span class="o">=</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">ref_fasta</span>
    <span class="c1"># add ngslib object for user supplied NGS data</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;adding the sequencing libraries to the seedGenome&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">fastqS1</span><span class="p">)</span>
    <span class="n">seedGenome</span><span class="o">.</span><span class="n">master_ngs_ob</span> <span class="o">=</span> <span class="n">NgsLib</span><span class="p">(</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;master&quot;</span><span class="p">,</span>
        <span class="n">master</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="n">make_dist</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;smalt&quot;</span><span class="p">,</span>
        <span class="n">readF</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">fastq1</span><span class="p">,</span>
        <span class="n">readR</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">fastq2</span><span class="p">,</span>
        <span class="n">readS0</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">fastqS1</span><span class="p">,</span>
        <span class="c1"># readS1=args.fastqS2,</span>
        <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">,</span>
        <span class="n">mapper_exe</span><span class="o">=</span><span class="n">sys_exes</span><span class="o">.</span><span class="n">mapper</span><span class="p">,</span>
        <span class="n">ref_fasta</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">ref_fasta</span><span class="p">)</span>

    <span class="n">checked_k</span> <span class="o">=</span> <span class="n">check_kmer_vs_reads</span><span class="p">(</span>
        <span class="n">k</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">kmers</span><span class="p">,</span>
        <span class="n">readlen</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">master_ngs_ob</span><span class="o">.</span><span class="n">readlen</span><span class="p">,</span>
        <span class="n">min_diff</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
    <span class="n">checked_prek</span> <span class="o">=</span> <span class="n">check_kmer_vs_reads</span><span class="p">(</span>
        <span class="n">k</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">pre_kmers</span><span class="p">,</span>
        <span class="n">readlen</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">master_ngs_ob</span><span class="o">.</span><span class="n">readlen</span><span class="p">,</span>
        <span class="n">min_diff</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Using the following kmer values pre and final assemblies:&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">checked_k</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">checked_prek</span><span class="p">)</span>

    <span class="k">if</span> <span class="s2">&quot;pe&quot;</span> <span class="ow">in</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">master_ngs_ob</span><span class="o">.</span><span class="n">libtype</span><span class="p">:</span>
        <span class="c1"># check equal length fastq.  This doesnt actually check propper pairs</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Checking that the fastq pair have equal number of reads&quot;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">check_fastqs_len_equal</span><span class="p">(</span><span class="n">file1</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">fastq1</span><span class="p">,</span> <span class="n">file2</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">fastq2</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="c1"># not just value error, whatever file_len throws</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">last_exception</span><span class="p">())</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="c1"># read in riboSelect clusters, make a lociCluster ob for each,</span>
    <span class="c1"># which get placed in seedGenome.loci_clusters</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">seedGenome</span><span class="o">.</span><span class="n">loci_clusters</span> <span class="o">=</span> <span class="n">parse_clustered_loci_file</span><span class="p">(</span>
            <span class="n">filepath</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">clustered_loci_txt</span><span class="p">,</span>
            <span class="n">gb_filepath</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">genbank_path</span><span class="p">,</span>
            <span class="n">output_root</span><span class="o">=</span><span class="n">output_root</span><span class="p">,</span>
            <span class="n">circular</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">linear</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">,</span>
            <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">last_exception</span><span class="p">())</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># add coordinates to lociCluster.loci_list</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">add_coords_to_clusters</span><span class="p">(</span><span class="n">seedGenome</span><span class="o">=</span><span class="n">seedGenome</span><span class="p">,</span>
                               <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">last_exception</span><span class="p">())</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;padding genbank by </span><span class="si">%i</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">args</span><span class="o">.</span><span class="n">flanking</span> <span class="o">*</span> <span class="mi">3</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;old ref_fasta: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">ref_fasta</span><span class="p">)</span>
        <span class="n">seedGenome</span><span class="o">.</span><span class="n">pad_genbank</span><span class="p">(</span><span class="n">pad</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">flanking</span> <span class="o">*</span> <span class="mi">3</span><span class="p">,</span>
                               <span class="n">circular</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">linear</span> <span class="ow">is</span> <span class="kc">False</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;new ref_fasta: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">ref_fasta</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">last_exception</span><span class="p">())</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># make first iteration look like future iterations:</span>
    <span class="c1"># this should also ensure the mapper uses the padded version</span>
    <span class="n">seedGenome</span><span class="o">.</span><span class="n">next_reference_path</span> <span class="o">=</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">ref_fasta</span>
    <span class="c1">#</span>
    <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">loci_clusters</span><span class="p">:</span>
        <span class="n">cluster</span><span class="o">.</span><span class="n">master_ngs_ob</span> <span class="o">=</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">master_ngs_ob</span>
<span class="c1"># ---------------------------------------------------------------------------</span>
    <span class="c1"># Performance summary lists</span>
    <span class="n">mapping_percentages</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">region_depths</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c1"># now, we need to assemble each mapping object</span>
    <span class="c1"># this should exclude any failures</span>
    <span class="k">while</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span> <span class="o">&lt;</span> <span class="n">args</span><span class="o">.</span><span class="n">iterations</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;processing iteration </span><span class="si">%i</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;with new reference: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">next_reference_path</span><span class="p">)</span>
        <span class="n">clusters_to_process</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">loci_clusters</span> <span class="k">if</span>
                               <span class="n">x</span><span class="o">.</span><span class="n">continue_iterating</span> <span class="ow">and</span>
                               <span class="n">x</span><span class="o">.</span><span class="n">keep_contigs</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">clusters_to_process</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;No clusters had sufficient mapping! Exiting&quot;</span><span class="p">)</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">clusters_to_process</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">loci_clusters</span><span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot;clusters excluded from this iteration </span><span class="se">\n</span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span>
                          <span class="n">seedGenome</span><span class="o">.</span><span class="n">loci_clusters</span> <span class="k">if</span>
                          <span class="n">x</span><span class="o">.</span><span class="n">index</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="n">y</span><span class="o">.</span><span class="n">index</span> <span class="k">for</span>
                                          <span class="n">y</span> <span class="ow">in</span> <span class="n">clusters_to_process</span><span class="p">]]))</span>
        <span class="c1"># For each (non-inital) iteration</span>
        <span class="k">if</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span> <span class="o">!=</span> <span class="mi">1</span><span class="p">:</span>
                <span class="c1"># clear out old .sam files to save space</span>
                <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">clean_temps</span><span class="p">:</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;removing uneeded file:&quot;</span><span class="p">)</span>
                    <span class="c1"># delete the read files from the last mapping</span>
                    <span class="c1"># dont do this on first iteration cause those be the reads!</span>
                    <span class="c1"># and if they aren&#39;t backed up you are up a creek and</span>
                    <span class="c1"># probably very upset with me.</span>
                    <span class="n">seedGenome</span><span class="o">.</span><span class="n">purge_old_files</span><span class="p">(</span><span class="n">all_iters</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>

            <span class="c1"># seqrecords for the clusters to be gen.next_reference_path</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">next_reference_path</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">nextref</span><span class="p">:</span>
                <span class="n">next_seqrec</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">nextref</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>  <span class="c1"># next?</span>
            <span class="k">for</span> <span class="n">clu</span> <span class="ow">in</span> <span class="n">clusters_to_process</span><span class="p">:</span>
                <span class="n">clu</span><span class="o">.</span><span class="n">seq_record</span> <span class="o">=</span> <span class="n">next_seqrec</span>
            <span class="c1"># print qualities of mapped reads</span>
            <span class="k">for</span> <span class="n">clu</span> <span class="ow">in</span> <span class="n">clusters_to_process</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;getting mapping scores for cluster </span><span class="si">%i</span><span class="s2"> from </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
                             <span class="n">clu</span><span class="o">.</span><span class="n">index</span><span class="p">,</span> <span class="n">clu</span><span class="o">.</span><span class="n">mappings</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">mapped_bam</span><span class="p">)</span>
                <span class="n">mapped_scores</span> <span class="o">=</span> <span class="n">get_bam_AS</span><span class="p">(</span>
                    <span class="n">inbam</span><span class="o">=</span><span class="n">clu</span><span class="o">.</span><span class="n">mappings</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">mapped_bam</span><span class="p">,</span>
                    <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">mapped_scores</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">200000</span><span class="p">:</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Downsampling our pltting data to 20k points&quot;</span><span class="p">)</span>
                    <span class="n">mapped_scores</span> <span class="o">=</span> <span class="n">random</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="n">mapped_scores</span><span class="p">,</span> <span class="mi">200000</span><span class="p">)</span>
                <span class="n">printPlot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">mapped_scores</span><span class="p">,</span> <span class="n">line</span><span class="o">=</span><span class="n">score_minimum</span><span class="p">,</span>
                          <span class="n">ymax</span><span class="o">=</span><span class="mi">18</span><span class="p">,</span> <span class="n">xmax</span><span class="o">=</span><span class="mi">60</span><span class="p">,</span> <span class="n">tick</span><span class="o">=.</span><span class="mi">2</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                          <span class="n">title</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;Average alignment Scores for cluster &quot;</span> <span class="o">+</span>
                                    <span class="s2">&quot;</span><span class="si">%i</span><span class="se">\n</span><span class="s2"> &quot;</span> <span class="o">%</span> <span class="n">clu</span><span class="o">.</span><span class="n">index</span><span class="p">),</span>
                          <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="s2">&quot;-&quot;</span> <span class="o">*</span> <span class="mi">72</span><span class="p">))</span>

            <span class="c1"># make new ngslib from unampped reads</span>
            <span class="n">convert_cmd</span><span class="p">,</span> <span class="n">unmapped_ngsLib</span> <span class="o">=</span> <span class="n">convert_bam_to_fastqs_cmd</span><span class="p">(</span>
                <span class="n">mapping_ob</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span>
                    <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span> <span class="o">-</span> <span class="mi">1</span><span class="p">],</span>
                <span class="n">samtools_exe</span><span class="o">=</span><span class="n">sys_exes</span><span class="o">.</span><span class="n">samtools</span><span class="p">,</span> <span class="n">single</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="c1"># ref fasta is used to make index cmd</span>
                <span class="n">ref_fasta</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">next_reference_path</span><span class="p">,</span>
                <span class="n">which</span><span class="o">=</span><span class="s1">&#39;unmapped&#39;</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
            <span class="c1"># unless subtract arg is used, use all reads each mapping</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">subtract</span><span class="p">:</span>
                <span class="n">unmapped_ngsLib</span> <span class="o">=</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">master_ngs_ob</span>

            <span class="n">unmapped_ngsLib</span><span class="o">.</span><span class="n">readlen</span> <span class="o">=</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">master_ngs_ob</span><span class="o">.</span><span class="n">readlen</span>
            <span class="n">unmapped_ngsLib</span><span class="o">.</span><span class="n">smalt_dist_path</span> <span class="o">=</span> \
                <span class="n">seedGenome</span><span class="o">.</span><span class="n">master_ngs_ob</span><span class="o">.</span><span class="n">smalt_dist_path</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;converting unmapped bam into reads:&quot;</span><span class="p">)</span>
            <span class="n">seedGenome</span><span class="o">.</span><span class="n">master_ngs_ob</span><span class="o">.</span><span class="n">ref_fasta</span> <span class="o">=</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">next_reference_path</span>
            <span class="c1"># dont worry about wasting time making these libraries if</span>
            <span class="c1"># not subtracting previously mapped reads</span>
            <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">subtract</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">cmd</span> <span class="ow">in</span> <span class="p">[</span><span class="n">convert_cmd</span><span class="p">]:</span>  <span class="c1"># may have more cmds here in future</span>
                    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
                    <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="n">cmd</span><span class="p">],</span>
                                   <span class="n">shell</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">platform</span> <span class="o">!=</span> <span class="s2">&quot;win32&quot;</span><span class="p">,</span>
                                   <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                                   <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                                   <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># start with whole lib if first time through</span>
            <span class="n">unmapped_ngsLib</span> <span class="o">=</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">master_ngs_ob</span>
        <span class="c1"># Run commands to map to the genome</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">score_min</span><span class="p">:</span>
            <span class="c1"># This makes it such that score minimum is now more stringent</span>
            <span class="c1"># with each mapping.</span>
            <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;smalt&#39;</span><span class="p">:</span>
                <span class="n">scaling_factor</span> <span class="o">=</span> <span class="mf">1.0</span> <span class="o">-</span> <span class="p">(</span>
                    <span class="mf">1.0</span> <span class="o">/</span> <span class="p">(</span><span class="mf">2.0</span> <span class="o">+</span> <span class="nb">float</span><span class="p">(</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">)))</span>
                <span class="n">score_minimum</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">unmapped_ngsLib</span><span class="o">.</span><span class="n">readlen</span> <span class="o">*</span> <span class="n">scaling_factor</span><span class="p">)</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                    <span class="s2">&quot;Mapping with min_score of </span><span class="si">%f</span><span class="s2">2 (</span><span class="si">%f</span><span class="s2">2 of read length, </span><span class="si">%f</span><span class="s2">2)&quot;</span><span class="p">,</span>
                    <span class="n">scaling_factor</span><span class="p">,</span> <span class="n">score_minimum</span><span class="p">,</span> <span class="n">unmapped_ngsLib</span><span class="o">.</span><span class="n">readlen</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">assert</span> <span class="n">args</span><span class="o">.</span><span class="n">method</span> <span class="o">==</span> <span class="s1">&#39;bwa&#39;</span><span class="p">,</span> <span class="s2">&quot;must be wither smalt or bwa&quot;</span>
                <span class="c1"># score_minimum = int(.15 * unmapped_ngsLib.readlen)</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;using the default minimum score for BWA&quot;</span><span class="p">)</span>
                <span class="n">score_minimum</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">score_minimum</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">score_min</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;Mapping with min_score of </span><span class="si">%f</span><span class="s2">2 (read length: </span><span class="si">%f</span><span class="s2">2)&quot;</span><span class="p">,</span>
                <span class="n">score_minimum</span><span class="p">,</span> <span class="n">unmapped_ngsLib</span><span class="o">.</span><span class="n">readlen</span><span class="p">)</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">nonify_empty_lib_files</span><span class="p">(</span><span class="n">unmapped_ngsLib</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                <span class="s2">&quot;No reads mapped for this iteration. This could be to an &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;error from samtools or elevated mapping stringency.&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot; proceeding to final assemblies&quot;</span><span class="p">)</span>
                <span class="k">break</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot; Exiting!&quot;</span><span class="p">)</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
        <span class="c1"># the exe argument is Exes.mapper because that is what is checked</span>
        <span class="c1"># during object instantiation</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;smalt&quot;</span><span class="p">:</span>
            <span class="c1"># # get rid of bwa mapper default args</span>
            <span class="c1"># if args.mapper_args == &#39;-L 0,0 -U 0&#39;:</span>
            <span class="c1">#     args.mapper_args =</span>
            <span class="n">map_percent</span> <span class="o">=</span> <span class="n">map_to_genome_ref_smalt</span><span class="p">(</span>
                <span class="n">mapping_ob</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span>
                    <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">],</span>
                <span class="n">ngsLib</span><span class="o">=</span><span class="n">unmapped_ngsLib</span><span class="p">,</span>
                <span class="n">cores</span><span class="o">=</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">cores</span> <span class="o">*</span> <span class="n">args</span><span class="o">.</span><span class="n">threads</span><span class="p">),</span>
                <span class="n">samtools_exe</span><span class="o">=</span><span class="n">sys_exes</span><span class="o">.</span><span class="n">samtools</span><span class="p">,</span>
                <span class="n">genome_fasta</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">next_reference_path</span><span class="p">,</span>
                <span class="n">smalt_exe</span><span class="o">=</span><span class="n">sys_exes</span><span class="o">.</span><span class="n">mapper</span><span class="p">,</span>
                <span class="n">score_minimum</span><span class="o">=</span><span class="n">score_minimum</span><span class="p">,</span>
                <span class="n">step</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
                <span class="n">scoring</span><span class="o">=</span><span class="s2">&quot;match=1,subst=-4,gapopen=-4,gapext=-3&quot;</span><span class="p">,</span>
                <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">args</span><span class="o">.</span><span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;bwa&quot;</span><span class="p">,</span> <span class="s2">&quot;must be either bwa or smalt&quot;</span>
            <span class="n">map_percent</span><span class="p">,</span> <span class="n">score_list</span> <span class="o">=</span> <span class="n">map_to_genome_ref_bwa</span><span class="p">(</span>
                <span class="n">mapping_ob</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">iter_mapping_list</span><span class="p">[</span>
                    <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">],</span>
                <span class="n">ngsLib</span><span class="o">=</span><span class="n">unmapped_ngsLib</span><span class="p">,</span>
                <span class="n">cores</span><span class="o">=</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">cores</span> <span class="o">*</span> <span class="n">args</span><span class="o">.</span><span class="n">threads</span><span class="p">),</span>
                <span class="n">genome_fasta</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">next_reference_path</span><span class="p">,</span>
                <span class="n">samtools_exe</span><span class="o">=</span><span class="n">sys_exes</span><span class="o">.</span><span class="n">samtools</span><span class="p">,</span>
                <span class="n">bwa_exe</span><span class="o">=</span><span class="n">sys_exes</span><span class="o">.</span><span class="n">mapper</span><span class="p">,</span>
                <span class="n">score_minimum</span><span class="o">=</span><span class="n">score_minimum</span><span class="p">,</span>
                <span class="c1"># add_args=&#39;-L 0,0 -U 0&#39;,</span>
                <span class="n">add_args</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">mapper_args</span><span class="p">,</span>
                <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
        <span class="n">mapping_percentages</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Iteration </span><span class="si">%i</span><span class="s2">: </span><span class="si">%f</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="p">(</span>
            <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">,</span> <span class="n">map_percent</span><span class="p">))</span>
        <span class="c1"># if things go really bad on the first mapping, get out while you can</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">score_list</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                <span class="s2">&quot;No reads mapped for this iteration. This could be to an &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;error from samtools, bwa mem, a bad reference, &quot;</span> <span class="o">+</span>
                <span class="s2">&quot; or elevated mapping stringency. &quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot; proceeding to final assemblies&quot;</span><span class="p">)</span>
                <span class="k">break</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Exiting!&quot;</span><span class="p">)</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

        <span class="c1"># on first time through, infer ref_as_contig if not</span>
        <span class="c1">#   provided via commandline</span>
        <span class="k">if</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="c1"># do info for smalt mapping</span>
            <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;bwa&quot;</span><span class="p">:</span>
                <span class="n">fig_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_root</span><span class="p">,</span> <span class="s2">&quot;figs&quot;</span><span class="p">)</span>
                <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="n">fig_dir</span><span class="p">)</span>
                <span class="c1"># either use defined min or use the smae heuristic as mapping</span>
                <span class="k">if</span> <span class="n">PLOT</span><span class="p">:</span>
                    <span class="n">plotAsScores</span><span class="p">(</span>
                        <span class="n">score_list</span><span class="p">,</span>
                        <span class="n">score_min</span><span class="o">=</span><span class="n">score_minimum</span> <span class="k">if</span> <span class="n">score_minimum</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span>
                        <span class="nb">int</span><span class="p">(</span><span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">master_ngs_ob</span><span class="o">.</span><span class="n">readlen</span><span class="p">)</span> <span class="o">/</span> <span class="mf">2.0</span><span class="p">)),</span>
                        <span class="n">outdir</span><span class="o">=</span><span class="n">fig_dir</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
                <span class="n">printPlot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">score_list</span><span class="p">,</span> <span class="n">line</span><span class="o">=</span><span class="n">score_minimum</span><span class="p">,</span> <span class="n">ymax</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">xmax</span><span class="o">=</span><span class="mi">60</span><span class="p">,</span>
                          <span class="n">tick</span><span class="o">=.</span><span class="mi">2</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                          <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Average alignment Scores (y) by sorted &quot;</span> <span class="o">+</span>
                          <span class="s2">&quot;read index (x)&quot;</span><span class="p">,</span>
                          <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">ref_as_contig</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">map_percent</span> <span class="o">&gt;</span> <span class="mi">80</span><span class="p">:</span>
                        <span class="n">ref_as_contig</span> <span class="o">=</span> <span class="s2">&quot;trusted&quot;</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">ref_as_contig</span> <span class="o">=</span> <span class="s2">&quot;untrusted&quot;</span>
                        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                            <span class="nb">str</span><span class="p">(</span><span class="s2">&quot;unfiltered mapping percentage is </span><span class="si">%f</span><span class="s2">2 so &quot;</span> <span class="o">+</span>
                                <span class="s2">&quot;&#39;ref_as_contigs&#39; is set to </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">),</span>
                            <span class="n">map_percent</span><span class="p">,</span> <span class="n">ref_as_contig</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">ref_as_contig</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">ref_as_contig</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">ref_as_contig</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">ref_as_contig</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">pass</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="c1"># again, this is [(idx, start_depth, end_depth)]</span>
            <span class="c1"># clusters_post_partition are the clusters passing the minimum</span>
            <span class="c1"># depth on the flanking regions.</span>
            <span class="n">iter_depths</span><span class="p">,</span> <span class="n">clusters_to_subassemble</span> <span class="o">=</span> <span class="n">partition_mapping</span><span class="p">(</span>
                <span class="n">seedGenome</span><span class="o">=</span><span class="n">seedGenome</span><span class="p">,</span>
                <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">,</span>
                <span class="n">samtools_exe</span><span class="o">=</span><span class="n">sys_exes</span><span class="o">.</span><span class="n">samtools</span><span class="p">,</span>
                <span class="n">flank</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">flanking</span><span class="p">,</span>
                <span class="n">min_flank_depth</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">min_flank_depth</span><span class="p">,</span>
                <span class="n">cluster_list</span><span class="o">=</span><span class="n">clusters_to_process</span><span class="p">)</span>

        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Error while partitioning reads from iteration </span><span class="si">%i</span><span class="s2">&quot;</span><span class="p">,</span>
                         <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">last_exception</span><span class="p">())</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>

        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">iter_depths</span><span class="p">)</span>
        <span class="n">region_depths</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">iter_depths</span><span class="p">)</span>
        <span class="n">extract_convert_assemble_cmds</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1"># generate spades cmds (cannot be multiprocessed becuase of python&#39;s</span>
        <span class="c1">#  inability to pass objects to multiprocessing)</span>
        <span class="c1"># ref_as_contig must be &#39;trusted&#39; here because of the multimapping/</span>
        <span class="c1">#  coverage issues</span>
        <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">clusters_to_subassemble</span><span class="p">:</span>
            <span class="n">cmdlist</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;generating commands to convert bam to fastqs &quot;</span> <span class="o">+</span>
                         <span class="s2">&quot;and assemble long reads&quot;</span><span class="p">)</span>
            <span class="n">convert_cmds</span><span class="p">,</span> <span class="n">new_ngslib</span> <span class="o">=</span> <span class="n">convert_bam_to_fastqs_cmd</span><span class="p">(</span>
                <span class="n">mapping_ob</span><span class="o">=</span><span class="n">cluster</span><span class="o">.</span><span class="n">mappings</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">which</span><span class="o">=</span><span class="s1">&#39;mapped&#39;</span><span class="p">,</span>
                <span class="n">single</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">samtools_exe</span><span class="o">=</span><span class="n">sys_exes</span><span class="o">.</span><span class="n">samtools</span><span class="p">,</span>
                <span class="n">ref_fasta</span><span class="o">=</span><span class="n">cluster</span><span class="o">.</span><span class="n">mappings</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">ref_fasta</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
            <span class="n">cmdlist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">convert_cmds</span><span class="p">)</span>
            <span class="n">spades_cmd</span> <span class="o">=</span> <span class="n">generate_spades_cmd</span><span class="p">(</span>
                <span class="n">mapping_ob</span><span class="o">=</span><span class="n">cluster</span><span class="o">.</span><span class="n">mappings</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span>
                <span class="n">ngs_ob</span><span class="o">=</span><span class="n">new_ngslib</span><span class="p">,</span> <span class="n">single_lib</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">ref_as_contig</span><span class="o">=</span><span class="s2">&quot;trusted&quot;</span><span class="p">,</span>
                <span class="n">check_libs</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">as_paired</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">prelim</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">k</span><span class="o">=</span><span class="n">checked_prek</span><span class="p">,</span>
                <span class="n">spades_exe</span><span class="o">=</span><span class="n">sys_exes</span><span class="o">.</span><span class="n">spades</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
            <span class="c1"># setting some thread limits here</span>
            <span class="n">modest_spades_cmd</span> <span class="o">=</span> <span class="n">make_modest_spades_cmd</span><span class="p">(</span>
                <span class="n">cmd</span><span class="o">=</span><span class="n">spades_cmd</span><span class="p">,</span> <span class="n">cores</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">cores</span><span class="p">,</span> <span class="n">memory</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">memory</span><span class="p">,</span>
                <span class="n">serialize</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">serialize</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
            <span class="n">cmdlist</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">modest_spades_cmd</span><span class="p">)</span>

            <span class="n">cluster</span><span class="o">.</span><span class="n">mappings</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">mapped_ngslib</span> <span class="o">=</span> <span class="n">new_ngslib</span>
            <span class="n">extract_convert_assemble_cmds</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cmdlist</span><span class="p">)</span>

        <span class="c1"># run all those commands!</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
            <span class="s2">&quot;</span><span class="se">\n</span><span class="s2"> running </span><span class="si">%i</span><span class="s2"> cmds: </span><span class="se">\n</span><span class="s2"> </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="nb">len</span><span class="p">([</span><span class="n">j</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">extract_convert_assemble_cmds</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">i</span><span class="p">]),</span>
            <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">j</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">extract_convert_assemble_cmds</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">i</span><span class="p">]))</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">serialize</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;running without multiprocessing!&quot;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">cmd</span> <span class="ow">in</span> <span class="p">[</span><span class="n">j</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">extract_convert_assemble_cmds</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">i</span><span class="p">]:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
                <span class="c1"># subprocess_run_list(cmdlist=cmds, hard=False, logger=logger)</span>
                <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="n">cmd</span><span class="p">],</span>
                               <span class="n">shell</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">platform</span> <span class="o">!=</span> <span class="s2">&quot;win32&quot;</span><span class="p">,</span>
                               <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                               <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                               <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">pool</span> <span class="o">=</span> <span class="n">multiprocessing</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">cores</span><span class="p">)</span>
            <span class="n">results</span> <span class="o">=</span> <span class="p">[</span>
                <span class="n">pool</span><span class="o">.</span><span class="n">apply_async</span><span class="p">(</span><span class="n">subprocess_run_list</span><span class="p">,</span>
                                 <span class="p">(</span><span class="n">cmds</span><span class="p">,),</span>
                                 <span class="p">{</span><span class="s2">&quot;logger&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                                  <span class="s2">&quot;hard&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">})</span>
                <span class="k">for</span> <span class="n">cmds</span> <span class="ow">in</span> <span class="n">extract_convert_assemble_cmds</span><span class="p">]</span>
            <span class="n">pool</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
            <span class="n">pool</span><span class="o">.</span><span class="n">join</span><span class="p">()</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sum of return codes (should be 0):&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="nb">sum</span><span class="p">([</span><span class="n">r</span><span class="o">.</span><span class="n">get</span><span class="p">()</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">results</span><span class="p">]))</span>

        <span class="c1"># evaluate mapping (cant be multiprocessed)</span>
        <span class="k">for</span> <span class="n">cluster</span> <span class="ow">in</span> <span class="n">clusters_to_process</span><span class="p">:</span>
            <span class="n">cluster</span><span class="o">.</span><span class="n">assembly_success</span> <span class="o">=</span> <span class="n">evaluate_spades_success</span><span class="p">(</span>
                <span class="n">clu</span><span class="o">=</span><span class="n">cluster</span><span class="p">,</span>
                <span class="n">read_len</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">master_ngs_ob</span><span class="o">.</span><span class="n">readlen</span><span class="p">,</span>
                <span class="n">mapping_ob</span><span class="o">=</span><span class="n">cluster</span><span class="o">.</span><span class="n">mappings</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span>
                <span class="n">include_short_contigs</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">include_short_contigs</span><span class="p">,</span>
                <span class="n">keep_best_contig</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">min_delta</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                <span class="n">flank</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">flanking</span><span class="p">,</span>
                <span class="n">seqname</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">,</span>
                <span class="n">min_assembly_len</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">min_assembly_len</span><span class="p">,</span>
                <span class="n">proceed_to_target</span><span class="o">=</span><span class="n">proceed_to_target</span><span class="p">,</span>
                <span class="n">target_len</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">target_len</span><span class="p">)</span>
            <span class="n">parse_subassembly_return_code</span><span class="p">(</span>
                <span class="n">cluster</span><span class="o">=</span><span class="n">cluster</span><span class="p">,</span>
                <span class="n">final_contigs_dir</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">final_long_reads_dir</span><span class="p">,</span>
                <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
        <span class="n">clusters_for_pseudogenome</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">loci_clusters</span> <span class="k">if</span>
            <span class="n">x</span><span class="o">.</span><span class="n">continue_iterating</span> <span class="ow">and</span> <span class="n">x</span><span class="o">.</span><span class="n">keep_contigs</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">clusters_for_pseudogenome</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">faux_genome_path</span><span class="p">,</span> <span class="n">faux_genome_len</span> <span class="o">=</span> <span class="n">make_faux_genome</span><span class="p">(</span>
                <span class="n">seedGenome</span><span class="o">=</span><span class="n">seedGenome</span><span class="p">,</span>
                <span class="n">iteration</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">,</span>
                <span class="n">output_root</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">output_root</span><span class="p">,</span>
                <span class="n">nbuff</span><span class="o">=</span><span class="mi">5000</span><span class="p">,</span>
                <span class="n">cluster_list</span><span class="o">=</span><span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">clusters_for_pseudogenome</span> <span class="k">if</span>
                              <span class="n">x</span><span class="o">.</span><span class="n">continue_iterating</span><span class="p">],</span>
                <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Length of buffered &#39;genome&#39; for mapping: </span><span class="si">%i</span><span class="s2">&quot;</span><span class="p">,</span>
                        <span class="n">faux_genome_len</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">faux_genome_path</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span> <span class="o">=</span> <span class="n">args</span><span class="o">.</span><span class="n">iterations</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span> <span class="o">=</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="n">seedGenome</span><span class="o">.</span><span class="n">next_reference_path</span> <span class="o">=</span> <span class="n">faux_genome_path</span>
        <span class="k">if</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span> <span class="o">&gt;=</span> <span class="n">args</span><span class="o">.</span><span class="n">iterations</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;moving on to final assemblies!&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Moving on to iteration: </span><span class="si">%i</span><span class="s2">&quot;</span><span class="p">,</span>
                        <span class="n">seedGenome</span><span class="o">.</span><span class="n">this_iteration</span><span class="p">)</span>

<span class="c1"># --------------------------------------------------------------------------- #</span>
<span class="c1"># --------------------------------------------------------------------------- #</span>
    <span class="c1"># done with the iterations!  Lets free up some space</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">clean_temps</span><span class="p">:</span>
        <span class="n">seedGenome</span><span class="o">.</span><span class="n">purge_old_files</span><span class="p">(</span><span class="n">all_iters</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
    <span class="c1"># And add the remaining final contigs to the directory for combination</span>
    <span class="n">contigs_moved_before_list_iter</span> <span class="o">=</span> \
        <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">final_contig_path</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">loci_clusters</span>
         <span class="k">if</span> <span class="ow">not</span> <span class="n">x</span><span class="o">.</span><span class="n">keep_contigs</span><span class="p">]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Contigs moved prior to final iteration:&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">contigs_moved_before_list_iter</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">contigs_moved_before_list_iter</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">loci_clusters</span><span class="p">):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;all contigs already copied to long_reads dir&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># this assumes that all the uable clusters not in</span>
        <span class="c1"># clusters_for_pseudogenome have already been copied over</span>
        <span class="k">for</span> <span class="n">clu</span> <span class="ow">in</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">clusters_for_pseudogenome</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">keep_contigs</span><span class="p">]:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">shutil</span><span class="o">.</span><span class="n">copyfile</span><span class="p">(</span><span class="n">clu</span><span class="o">.</span><span class="n">mappings</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">assembled_contig</span><span class="p">,</span>
                                <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                                    <span class="n">seedGenome</span><span class="o">.</span><span class="n">final_long_reads_dir</span><span class="p">,</span>
                                    <span class="s2">&quot;</span><span class="si">{0}</span><span class="s2">_cluster_</span><span class="si">{1}</span><span class="s2">_iter_</span><span class="si">{2}</span><span class="s2">.fasta&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                                        <span class="n">clu</span><span class="o">.</span><span class="n">sequence_id</span><span class="p">,</span>
                                        <span class="n">clu</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>
                                        <span class="n">clu</span><span class="o">.</span><span class="n">mappings</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">iteration</span><span class="p">)))</span>

            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">last_exception</span><span class="p">())</span>
                <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">dirpath</span><span class="p">,</span> <span class="n">dirnames</span><span class="p">,</span> <span class="n">files</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">walk</span><span class="p">(</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">final_long_reads_dir</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">files</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                <span class="s2">&quot;No pseudocontigs found in the long reads output &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;directory; it appears &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;that the subassemblies did not yield pseudocontigs &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;of sufficient quality.  Exiting with code 0&quot;</span><span class="p">)</span>
            <span class="n">sys</span><span class="o">.</span><span class="n">exit</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;combining contigs from </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">final_long_reads_dir</span><span class="p">)</span>
    <span class="n">seedGenome</span><span class="o">.</span><span class="n">assembled_seeds</span> <span class="o">=</span> <span class="n">combine_contigs</span><span class="p">(</span>
        <span class="n">contigs_dir</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">final_long_reads_dir</span><span class="p">,</span>
        <span class="n">contigs_name</span><span class="o">=</span><span class="s2">&quot;riboSeedContigs&quot;</span><span class="p">,</span>
        <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Combined Seed Contigs: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">assembled_seeds</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Time taken to run seeding: </span><span class="si">%.2f</span><span class="s2">m&quot;</span> <span class="o">%</span> <span class="p">((</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">t0</span><span class="p">)</span> <span class="o">/</span> <span class="mi">60</span><span class="p">))</span>
    <span class="c1"># Diagnostics</span>
    <span class="c1"># logger.info(&quot;Mapping percentages per iteration: \n&quot; +</span>
    <span class="c1">#             &quot;(Iteration 0, which maps to entire reference, should &quot; +</span>
    <span class="c1">#             &quot;have a high value; other iterations are percentage of &quot; +</span>
    <span class="c1">#             &quot;previously unmapped reads which now map to the seeded &quot; +</span>
    <span class="c1">#             &quot;flanking regions, which may be very low)\n&quot; +</span>
    <span class="c1">#             &quot;\n&quot;.join(mapping_percentages))</span>
    <span class="n">report</span> <span class="o">=</span> <span class="n">reportRegionDepths</span><span class="p">(</span><span class="n">inp</span><span class="o">=</span><span class="n">region_depths</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Average depths of mapping for each cluster, by iteration:&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">report</span><span class="p">))</span>

    <span class="c1"># run final contigs</span>
    <span class="n">spades_quast_cmds</span><span class="p">,</span> <span class="n">quast_reports</span> <span class="o">=</span> <span class="n">get_final_assemblies_cmds</span><span class="p">(</span>
        <span class="n">seedGenome</span><span class="o">=</span><span class="n">seedGenome</span><span class="p">,</span> <span class="n">exes</span><span class="o">=</span><span class="n">sys_exes</span><span class="p">,</span>
        <span class="n">cores</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">cores</span><span class="p">,</span>
        <span class="n">memory</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">memory</span><span class="p">,</span>
        <span class="n">serialize</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">serialize</span><span class="p">,</span>
        <span class="n">ref_as_contig</span><span class="o">=</span><span class="n">ref_as_contig</span><span class="p">,</span>
        <span class="n">skip_control</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">skip_control</span><span class="p">,</span> <span class="n">kmers</span><span class="o">=</span><span class="n">checked_k</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">serialize</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;running without multiprocessing!&quot;</span><span class="p">)</span>
        <span class="c1"># unpack nested spades quast list</span>
        <span class="k">for</span> <span class="n">cmd</span> <span class="ow">in</span> <span class="p">[</span><span class="n">j</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">spades_quast_cmds</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">i</span><span class="p">]:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">cmd</span><span class="p">)</span>
            <span class="n">subprocess</span><span class="o">.</span><span class="n">run</span><span class="p">([</span><span class="n">cmd</span><span class="p">],</span>
                           <span class="n">shell</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">platform</span> <span class="o">!=</span> <span class="s2">&quot;win32&quot;</span><span class="p">,</span>
                           <span class="n">stdout</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                           <span class="n">stderr</span><span class="o">=</span><span class="n">subprocess</span><span class="o">.</span><span class="n">PIPE</span><span class="p">,</span>
                           <span class="n">check</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># split the processors based on how many spades_cmds are on the list</span>
        <span class="c1"># dont correct for threads, as Spades defaults to lots of threads</span>
        <span class="n">split_cores</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">cores</span> <span class="o">/</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">spades_quast_cmds</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">))</span>
        <span class="k">if</span> <span class="n">split_cores</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">split_cores</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="n">pool</span> <span class="o">=</span> <span class="n">multiprocessing</span><span class="o">.</span><span class="n">Pool</span><span class="p">(</span><span class="n">processes</span><span class="o">=</span><span class="n">split_cores</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;running the following commands:&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">j</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">spades_quast_cmds</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">i</span><span class="p">]))</span>
        <span class="n">results</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">pool</span><span class="o">.</span><span class="n">apply_async</span><span class="p">(</span><span class="n">subprocess_run_list</span><span class="p">,</span>
                             <span class="p">(</span><span class="n">cmds</span><span class="p">,),</span>
                             <span class="p">{</span><span class="s2">&quot;logger&quot;</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
                              <span class="s2">&quot;hard&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">})</span>
            <span class="k">for</span> <span class="n">cmds</span> <span class="ow">in</span> <span class="n">spades_quast_cmds</span><span class="p">]</span>
        <span class="n">pool</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="n">pool</span><span class="o">.</span><span class="n">join</span><span class="p">()</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Sum of return codes (should be 0):&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="nb">sum</span><span class="p">([</span><span class="n">r</span><span class="o">.</span><span class="n">get</span><span class="p">()</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">results</span><span class="p">]))</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">skip_control</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;writing combined quast reports&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Comparing de novo and de fere novo assemblies:&quot;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">quast_comp</span> <span class="o">=</span> <span class="n">make_quick_quast_table</span><span class="p">(</span>
                <span class="n">quast_reports</span><span class="p">,</span>
                <span class="n">write</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                <span class="n">writedir</span><span class="o">=</span><span class="n">seedGenome</span><span class="o">.</span><span class="n">output_root</span><span class="p">,</span>
                <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">quast_comp</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;</span><span class="si">%s</span><span class="s2">: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">k</span><span class="p">,</span> <span class="s2">&quot;  &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">v</span><span class="p">))</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;Error writing out combined quast report&quot;</span><span class="p">)</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
    <span class="c1"># make dir for easy downloading from cluster</span>
    <span class="n">copyToHandyDir</span><span class="p">(</span><span class="n">outdir</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_root</span><span class="p">,</span> <span class="s2">&quot;mauve&quot;</span><span class="p">),</span>
                   <span class="n">pre</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">exp_name</span><span class="p">,</span>
                   <span class="n">seedGenome</span><span class="o">=</span><span class="n">seedGenome</span><span class="p">,</span>
                   <span class="n">hard</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">logger</span><span class="o">=</span><span class="n">logger</span><span class="p">)</span>
    <span class="c1"># Report that we&#39;ve finished</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Done: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">time</span><span class="o">.</span><span class="n">asctime</span><span class="p">())</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;riboSeed Assembly: </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">output_root</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Combined Contig Seeds (for validation or alternate &quot;</span> <span class="o">+</span>
                <span class="s2">&quot;assembly): </span><span class="si">%s</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">seedGenome</span><span class="o">.</span><span class="n">assembled_seeds</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Time taken: </span><span class="si">%.2f</span><span class="s2">m&quot;</span> <span class="o">%</span> <span class="p">((</span><span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span> <span class="o">-</span> <span class="n">t0</span><span class="p">)</span> <span class="o">/</span> <span class="mi">60</span><span class="p">))</span>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">riboSeed 0.4.9 documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="index.html" >Module code</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2017, Nicholas Waters, Florence Abram, Fiona Brennan, Ashleigh Holmes, and Leighton Pritchard.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.3.
    </div>
  </body>
</html>