
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>riboSeed Pipeline &#8212; riboSeed 0.4.9 documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.4.9',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Welcome to riboSeed’s documentation!" href="index.html" /> 
  </head>
  <body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to riboSeed’s documentation!"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">riboSeed 0.4.9 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <p><a class="reference external" href="https://travis-ci.org/nickp60/riboSeed"><img alt="Build Status" src="https://travis-ci.org/nickp60/riboSeed.svg?branch=master" /></a> <a class="reference external" href="https://badge.fury.io/py/riboSeed"><img alt="PyPI version" src="https://badge.fury.io/py/riboSeed.svg" /></a> <a class="reference external" href="https://coveralls.io/github/nickp60/riboSeed?branch=master"><img alt="Coverage Status" src="https://coveralls.io/repos/github/nickp60/riboSeed/badge.svg?branch=master" /></a> <a class="reference external" href="https://doi.org/10.5281/zenodo.832131"><img alt="DOI" src="https://zenodo.org/badge/DOI/10.5281/zenodo.832131.svg" /></a></p>
<p><a class="reference external" href="https://github.com/nickp60/riboSeed/blob/master/icon/logo_1.png"><img alt="ribologo" src="https://github.com/nickp60/riboSeed/blob/master/icon/logo_1.png" /></a></p>
<div class="section" id="riboseed-pipeline">
<h1>riboSeed Pipeline<a class="headerlink" href="#riboseed-pipeline" title="Permalink to this headline">¶</a></h1>
<p>Impatient? See our <a class="reference external" href="https://github.com/nickp60/riboSeed/blob/master/quickstart.md">Quickstart Guide</a></p>
<p>A brief overview of the theory can be found
<a class="reference external" href="https://nickp60.github.io/riboSeed.html">here</a>.</p>
<p>Preprint of the riboSeed manuscript can be found
<a class="reference external" href="http://www.biorxiv.org/content/early/2017/07/14/159798">here</a>;
comments welcome!</p>
<div class="section" id="description">
<h2>Description<a class="headerlink" href="#description" title="Permalink to this headline">¶</a></h2>
<p>riboSeed is an supplemental assembly refinement method to try to address
the issue of multiple ribosomal regions in a genome, as these create
repeates unresolvable by short read sequencing. It takes advantage of
the fact that while each region is identical, the regions flanking are
unique, and therefore can potentially be used to seed an assembly in
such a way that rDNA regions are bridged.</p>
<p>The pipeline consists of 3 main stages: preprocessing, de fere novo
assembly, and visualization/assessment.</p>
</div>
<div class="section" id="before-we-start">
<h2>Before We Start<a class="headerlink" href="#before-we-start" title="Permalink to this headline">¶</a></h2>
<p>This pipeline is still very much in testing. Please back up any and all
data used, and work within a virtualenv.</p>
<p>Genome assembly gobbles RAM. If you, like me, are working on a 4gb RAM
lappy, don’t run riboSeed in parallel and instead run in series by using
the <code class="docutils literal"><span class="pre">--serialize</span></code> option. That should prevent you from running out of
RAM during the final SPAdes calls.</p>
</div>
<div class="section" id="preprocessing">
<h2>0: Preprocessing<a class="headerlink" href="#preprocessing" title="Permalink to this headline">¶</a></h2>
<div class="section" id="riboscan-py">
<h3><code class="docutils literal"><span class="pre">riboScan.py</span></code><a class="headerlink" href="#riboscan-py" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal"><span class="pre">riboScan.py</span></code> preprocesses sequences straight from a multifasta or one
or more fasta. The issue with many legacy annotations, assemblies, and
scaffold collections is rDNAs are often poorly annotated at best, and
unannotated at worst. This is shortcut to happiness without using the
full Prokka annotation scheme. It requires
<code class="docutils literal"><span class="pre">`barrnap</span></code> &lt;<a class="reference external" href="http://www.vicbioinformatics.com/software.barrnap.shtml">http://www.vicbioinformatics.com/software.barrnap.shtml</a>&gt;`__
and seqret (from <code class="docutils literal"><span class="pre">`emboss</span></code> &lt;<a class="reference external" href="http://www.ebi.ac.uk/Tools/emboss/">http://www.ebi.ac.uk/Tools/emboss/</a>&gt;`__) to
be available in your path. #### Usage</p>
<p>riboScan can either use a directory of fastas or one (multi)fasta file.
If using a directory of fastas, provide the appropriate extension using
the <code class="docutils literal"><span class="pre">-e</span></code> option. If using a (multi)fasta as input, it write out each
entry to its own fasta in the <code class="docutils literal"><span class="pre">contigs</span></code> subdirectory that it makes in
the output. For each of the fastas, the script renames complex headers
(sketchy), scans with barrnap and captures the output gff. It then edits
the gff to add fake incrementing locus_tags, and uses the original
sequence file through seqret to make a GenBank file that contains just
annotated rRNA features. The last step is a concatenation which, whether
or not there are multiple files, makes a single (possibly multi-entry)
genbank file perfect for riboSeed-ing.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">riboScan</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span> <span class="p">[</span><span class="o">-</span><span class="n">e</span> <span class="n">EXT</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">k</span> <span class="p">{</span><span class="n">bac</span><span class="p">,</span><span class="n">euk</span><span class="p">,</span><span class="n">arc</span><span class="p">,</span><span class="n">mito</span><span class="p">}]</span> <span class="p">[</span><span class="o">-</span><span class="n">t</span> <span class="n">ID_THRESH</span><span class="p">]</span>
                   <span class="p">[</span><span class="o">-</span><span class="n">b</span> <span class="n">BARRNAP_EXE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">n</span> <span class="n">NAME</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span> <span class="p">{</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">16</span><span class="p">}]</span>
                   <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SEQRET_EXE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span> <span class="p">{</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">}]</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span>
                   <span class="n">contigs</span>

<span class="n">Given</span> <span class="n">a</span> <span class="n">directory</span> <span class="n">of</span> <span class="n">one</span> <span class="ow">or</span> <span class="n">more</span> <span class="n">chromosomes</span> <span class="k">as</span> <span class="n">fasta</span> <span class="n">files</span><span class="p">,</span> <span class="n">this</span> <span class="n">facilitates</span>
<span class="n">reannotation</span> <span class="n">of</span> <span class="n">rDNA</span> <span class="n">regions</span> <span class="k">with</span> <span class="n">Barrnap</span> <span class="ow">and</span> <span class="n">outputs</span> <span class="nb">all</span> <span class="n">sequences</span> <span class="k">as</span> <span class="n">a</span>
<span class="n">single</span><span class="p">,</span> <span class="n">annotated</span> <span class="n">genbank</span> <span class="n">file</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">contigs_dir</span>           <span class="n">directory</span> <span class="n">containing</span> <span class="n">one</span> <span class="ow">or</span> <span class="n">more</span> <span class="n">chromosomal</span> <span class="n">sequences</span>
                        <span class="ow">in</span> <span class="n">fasta</span> <span class="nb">format</span>

<span class="n">required</span> <span class="n">named</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">,</span> <span class="o">--</span><span class="n">output</span> <span class="n">OUTPUT</span>
                        <span class="n">output</span> <span class="n">directory</span><span class="p">;</span> <span class="n">default</span><span class="p">:</span>
                        <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">nicholas</span><span class="o">/</span><span class="n">GitHub</span><span class="o">/</span><span class="n">riboSeed</span>


<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">k</span> <span class="p">{</span><span class="n">bac</span><span class="p">,</span><span class="n">euk</span><span class="p">,</span><span class="n">arc</span><span class="p">,</span><span class="n">mito</span><span class="p">},</span> <span class="o">--</span><span class="n">kingdom</span> <span class="p">{</span><span class="n">bac</span><span class="p">,</span><span class="n">euk</span><span class="p">,</span><span class="n">arc</span><span class="p">,</span><span class="n">mito</span><span class="p">}</span>
                        <span class="n">whether</span> <span class="n">to</span> <span class="n">look</span> <span class="k">for</span> <span class="n">eukaryotic</span><span class="p">,</span> <span class="n">archaeal</span><span class="p">,</span> <span class="ow">or</span> <span class="n">bacterial</span>
                        <span class="n">rDNA</span><span class="p">;</span> <span class="n">default</span><span class="p">:</span> <span class="n">bac</span>
  <span class="o">-</span><span class="n">e</span> <span class="n">extension</span>          <span class="n">extension</span> <span class="n">of</span> <span class="n">the</span> <span class="n">chromosomal</span> <span class="n">sequences</span><span class="p">,</span> <span class="n">usually</span>
                        <span class="s1">&#39;.fa&#39;</span><span class="p">,</span> <span class="s1">&#39;.fasta&#39;</span> <span class="ow">or</span> <span class="n">similar</span><span class="p">;</span> <span class="n">default</span><span class="p">:</span> <span class="o">.</span><span class="n">fa</span>
  <span class="o">-</span><span class="n">t</span> <span class="n">ID_THRESH</span><span class="p">,</span> <span class="o">--</span><span class="n">id_thresh</span> <span class="n">ID_THRESH</span>
                        <span class="n">partial</span> <span class="n">rRNA</span> <span class="n">hits</span> <span class="n">below</span> <span class="n">this</span> <span class="n">threshold</span> <span class="n">will</span> <span class="n">be</span>
                        <span class="n">ignored</span><span class="o">.</span> <span class="n">default</span><span class="p">:</span> <span class="mf">0.5</span>
  <span class="o">-</span><span class="n">b</span> <span class="n">BARRNAP_EXE</span><span class="p">,</span> <span class="o">--</span><span class="n">barrnap_exe</span> <span class="n">BARRNAP_EXE</span>
                        <span class="n">path</span> <span class="n">to</span> <span class="n">barrnap</span> <span class="n">executable</span><span class="p">;</span> <span class="n">default</span><span class="p">:</span> <span class="n">barrnap</span>
  <span class="o">-</span><span class="n">n</span> <span class="n">NAME</span><span class="p">,</span> <span class="o">--</span><span class="n">name</span> <span class="n">NAME</span>  <span class="n">name</span> <span class="n">to</span> <span class="n">give</span> <span class="n">the</span> <span class="n">contig</span> <span class="n">files</span><span class="p">;</span> <span class="n">default</span><span class="p">:</span> <span class="n">infered</span> <span class="kn">from</span>
                        <span class="nn">file</span>
  <span class="o">-</span><span class="n">s</span> <span class="n">SEQRET_EXE</span><span class="p">,</span> <span class="o">--</span><span class="n">seqret_exe</span> <span class="n">SEQRET_EXE</span>
                        <span class="n">path</span> <span class="n">to</span> <span class="n">seqret</span> <span class="n">executable</span><span class="p">,</span> <span class="n">usually</span> <span class="n">installed</span> <span class="k">with</span>
                        <span class="n">emboss</span><span class="p">;</span> <span class="n">default</span><span class="p">:</span> <span class="n">seqret</span>
  <span class="o">-</span><span class="n">v</span> <span class="p">{</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">},</span> <span class="o">--</span><span class="n">verbosity</span> <span class="p">{</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">}</span>
                        <span class="n">Logger</span> <span class="n">writes</span> <span class="n">debug</span> <span class="n">to</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">output</span> <span class="nb">dir</span><span class="p">;</span> <span class="n">this</span> <span class="n">sets</span>
                        <span class="n">verbosity</span> <span class="n">level</span> <span class="n">sent</span> <span class="n">to</span> <span class="n">stderr</span><span class="o">.</span> <span class="mi">1</span> <span class="o">=</span> <span class="n">debug</span><span class="p">(),</span> <span class="mi">2</span> <span class="o">=</span>
                        <span class="n">info</span><span class="p">(),</span> <span class="mi">3</span> <span class="o">=</span> <span class="n">warning</span><span class="p">(),</span> <span class="mi">4</span> <span class="o">=</span> <span class="n">error</span><span class="p">()</span> <span class="ow">and</span> <span class="mi">5</span> <span class="o">=</span> <span class="n">critical</span><span class="p">();</span>
                        <span class="n">default</span><span class="p">:</span> <span class="mi">2</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">Displays</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span>
</pre></div>
</div>
<p>NOTE: If using a reference with long names or containing special
characters, use the –name argument to rename the contigs to something a
bit more convenient and less prone to errors when piping results.</p>
</div>
<div class="section" id="riboselect-py">
<h3><code class="docutils literal"><span class="pre">riboSelect.py</span></code><a class="headerlink" href="#riboselect-py" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal"><span class="pre">riboSelect.py</span></code> searches the genome for rRNA annotations, clusters
them into likely ribosomal groups, and outputs a colon-separated list of
clustered rRNA locus tags by record id.</p>
<p>You will probably want to preview your file to figure out the syntax
used. (ie, 16s vs 16S, rRNA vs RRNA, etc…)</p>
<p>If not using <code class="docutils literal"><span class="pre">riboScan.py</span></code> or if not working with a prokaryotic
genome, you will need to change <code class="docutils literal"><span class="pre">--specific_features</span></code> appropriately to
reflect the annotations in your reference (ie, for a fungal genome, use
<code class="docutils literal"><span class="pre">--specific_features</span> <span class="pre">5_8S:18S:28S</span></code>).</p>
<p>NOTE: the format of the output text file is very simple, and due to the
relatively small number of such coding sequences in bacterial genomes,
this can be constructed by hand if the clusters do not look appropriate.
The format is <code class="docutils literal"><span class="pre">genome_sequence_id</span> <span class="pre">locus_tag1:locus_tag2</span></code>, where each
line represents a cluster. See example below, where 14 rRNAs are
clustered into 6 groups:</p>
<p>NOTE 2: In order to streamline things, as of version 0.0.3 there will be
a commented header line with the feature type in the format “#$ FEATURE
“, such as <code class="docutils literal"><span class="pre">#$</span> <span class="pre">FEATURE</span> <span class="pre">rRNA</span></code>.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="c1">#$ FEATURE rRNA</span>
<span class="n">CM000577</span><span class="o">.</span><span class="mi">1</span> <span class="n">FGSG_20052</span><span class="p">:</span><span class="n">FGSG_20051</span><span class="p">:</span><span class="n">FGSG_20053</span>
<span class="n">CM000577</span><span class="o">.</span><span class="mi">1</span> <span class="n">FGSG_20048</span><span class="p">:</span><span class="n">FGSG_20047</span>
<span class="n">CM000577</span><span class="o">.</span><span class="mi">1</span> <span class="n">FGSG_20049</span><span class="p">:</span><span class="n">FGSG_20050</span>
<span class="n">CM000577</span><span class="o">.</span><span class="mi">1</span> <span class="n">FGSG_20054</span><span class="p">:</span><span class="n">FGSG_20056</span><span class="p">:</span><span class="n">FGSG_20055</span>
<span class="n">CM000577</span><span class="o">.</span><span class="mi">1</span> <span class="n">FGSG_20058</span><span class="p">:</span><span class="n">FGSG_20057</span>
<span class="n">CM000577</span><span class="o">.</span><span class="mi">1</span> <span class="n">FGSG_20075</span><span class="p">:</span><span class="n">FGSG_20074</span>
</pre></div>
</div>
<div class="section" id="usage">
<h4>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h4>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">riboSelect</span><span class="o">.</span><span class="n">py</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">f</span> <span class="n">FEATURE</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">s</span> <span class="n">SPECIFIC_FEATURES</span><span class="p">]</span>
                     <span class="p">[</span><span class="o">--</span><span class="n">keep_temps</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">clobber</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span> <span class="n">CLUSTERS</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span> <span class="n">VERBOSITY</span><span class="p">]</span>
                     <span class="p">[</span><span class="o">--</span><span class="n">debug</span><span class="p">]</span>
                     <span class="n">genbank_genome</span>

<span class="n">This</span> <span class="ow">is</span> <span class="n">used</span> <span class="n">to</span> <span class="n">identify</span> <span class="ow">and</span> <span class="n">cluster</span> <span class="n">rRNA</span> <span class="n">regions</span> <span class="kn">from</span> <span class="nn">a</span> <span class="n">gb</span> <span class="n">file</span><span class="p">,</span> <span class="n">returns</span> <span class="n">a</span>
<span class="n">text</span> <span class="n">file</span> <span class="k">with</span> <span class="n">the</span> <span class="n">clusters</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">genbank_genome</span>        <span class="n">Genbank</span> <span class="n">file</span> <span class="p">(</span><span class="n">WITH</span> <span class="n">SEQUENCE</span><span class="p">)</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">show</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span> <span class="ow">and</span> <span class="n">exit</span>

<span class="n">required</span> <span class="n">named</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">,</span> <span class="o">--</span><span class="n">output</span> <span class="n">OUTPUT</span>
                        <span class="n">output</span> <span class="n">directory</span><span class="p">;</span><span class="n">default</span><span class="p">:</span>
                        <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">nicholas</span><span class="o">/</span><span class="n">GitHub</span><span class="o">/</span><span class="n">riboSeed</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">f</span> <span class="n">FEATURE</span><span class="p">,</span> <span class="o">--</span><span class="n">feature</span> <span class="n">FEATURE</span>
                        <span class="n">Feature</span><span class="p">,</span> <span class="n">rRNA</span> <span class="ow">or</span> <span class="n">RRNA</span><span class="p">;</span> <span class="n">default</span><span class="p">:</span> <span class="n">rRNA</span>
  <span class="o">-</span><span class="n">s</span> <span class="n">SPECIFIC_FEATURES</span><span class="p">,</span> <span class="o">--</span><span class="n">specific_features</span> <span class="n">SPECIFIC_FEATURES</span>
                        <span class="n">colon</span><span class="p">:</span><span class="n">separated</span> <span class="o">--</span> <span class="n">specific</span> <span class="n">features</span><span class="p">;</span> <span class="n">default</span><span class="p">:</span>
                        <span class="mi">16</span><span class="n">S</span><span class="p">:</span><span class="mi">23</span><span class="n">S</span><span class="p">:</span><span class="mi">5</span><span class="n">S</span>
  <span class="o">--</span><span class="n">keep_temps</span>          <span class="n">view</span> <span class="n">intermediate</span> <span class="n">clustering</span> <span class="n">filesdefault</span><span class="p">:</span> <span class="kc">False</span>
  <span class="o">--</span><span class="n">clobber</span>             <span class="n">overwrite</span> <span class="n">previous</span> <span class="n">output</span> <span class="n">files</span><span class="p">:</span> <span class="n">default</span><span class="p">:</span> <span class="kc">False</span>
  <span class="o">-</span><span class="n">c</span> <span class="n">CLUSTERS</span><span class="p">,</span> <span class="o">--</span><span class="n">clusters</span> <span class="n">CLUSTERS</span>
                        <span class="n">number</span> <span class="n">of</span> <span class="n">rDNA</span> <span class="n">clusters</span><span class="p">;</span><span class="k">if</span> <span class="n">submitting</span> <span class="n">multiple</span>
                        <span class="n">records</span><span class="p">,</span> <span class="n">must</span> <span class="n">be</span> <span class="n">a</span> <span class="n">colon</span><span class="p">:</span><span class="n">separated</span> <span class="nb">list</span> <span class="n">whose</span> <span class="n">length</span>
                        <span class="n">matches</span> <span class="n">number</span> <span class="n">of</span> <span class="n">genbank</span> <span class="n">records</span><span class="o">.</span> <span class="n">Default</span> <span class="ow">is</span> <span class="n">inferred</span>
                        <span class="kn">from</span> <span class="nn">specific</span> <span class="n">feature</span> <span class="k">with</span> <span class="n">fewest</span> <span class="n">hits</span>
  <span class="o">-</span><span class="n">v</span> <span class="n">VERBOSITY</span><span class="p">,</span> <span class="o">--</span><span class="n">verbosity</span> <span class="n">VERBOSITY</span>
                        <span class="mi">1</span> <span class="o">=</span> <span class="n">debug</span><span class="p">(),</span> <span class="mi">2</span> <span class="o">=</span> <span class="n">info</span><span class="p">(),</span> <span class="mi">3</span> <span class="o">=</span> <span class="n">warning</span><span class="p">(),</span> <span class="mi">4</span> <span class="o">=</span> <span class="n">error</span><span class="p">()</span>
                        <span class="ow">and</span> <span class="mi">5</span> <span class="o">=</span> <span class="n">critical</span><span class="p">();</span> <span class="n">default</span><span class="p">:</span> <span class="mi">2</span>
  <span class="o">--</span><span class="n">debug</span>               <span class="n">Enable</span> <span class="n">debug</span> <span class="n">messages</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="de-fere-novo-assembly">
<h2>2: <em>De fere novo</em> Assembly<a class="headerlink" href="#de-fere-novo-assembly" title="Permalink to this headline">¶</a></h2>
<div class="section" id="riboseed-py">
<h3><code class="docutils literal"><span class="pre">riboSeed.py</span></code><a class="headerlink" href="#riboseed-py" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal"><span class="pre">riboSeed.py</span></code> maps reads to a genome and (1) extracts reads mapping to
rDNA regions, (2) perfoms subassemblies on each pool of extracted reads
to recover the rDNA complete with flanking regions (resulting in a
pseudocontig) (3) concatenates a;; pseudocontigs into them into a
pseudogenome with 5kb spacers of N’s in between, (5) map remaining reads
to the pseudogenome, and (6) repeat steps 1-5 for a given number of
iterations (default 3 iterations). Finally, riboSeed runs SPAdes
assemblied with and without the pseudocontigs and the resulting
assemblies are assessed with QUAST.</p>
<div class="section" id="output">
<h4>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h4>
<p>The results directory will contain a ‘final_long_reads’ directory with
all the pseudocontigs, the mapped fastq files, and
<code class="docutils literal"><span class="pre">final_de_novo_assembly</span></code> and <code class="docutils literal"><span class="pre">final_de_fere_novo_assembly</span></code> folders,
containing the SPAdes results.</p>
<div class="section" id="note">
<h5>NOTE:<a class="headerlink" href="#note" title="Permalink to this headline">¶</a></h5>
<p>If using a consumer-grade computer, it will be advantagous to run with
<code class="docutils literal"><span class="pre">-z/--serialize</span></code> enabled to run asseblies in serial rather than
parallel.</p>
</div>
</div>
<div class="section" id="id1">
<h4>Usage:<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h4>
<p>minimal usage:
<code class="docutils literal"><span class="pre">riboSeed.py</span> <span class="pre">clustered_accession\list.txt</span> <span class="pre">-F</span> <span class="pre">FASTQ1</span> <span class="pre">-R</span> <span class="pre">FASTQ2</span> <span class="pre">-r</span> <span class="pre">REFERENCE_GENOME</span> <span class="pre">-o</span> <span class="pre">OUTPUT</span></code></p>
<div class="highlight-default"><div class="highlight"><pre><span></span>usage: riboSeed.py -r REFERENCE_GENBANK -o OUTPUT [-F FASTQ1] [-R FASTQ2]
                   [-S1 FASTQS1] [-n EXP_NAME] [-l FLANKING] [-m {smalt,bwa}]
                   [-c CORES] [-k KMERS] [-p PRE_KMERS] [-s SCORE_MIN]
                   [-a MIN_ASSEMBLY_LEN] [--include_shorts] [--linear]
                   [--ref_as_contig {None,trusted,untrusted}] [--keep_temps]
                   [--skip_control] [-i ITERATIONS] [-v {1,2,3,4,5}]
                   [--target_len TARGET_LEN] [-t {1,2,4}] [-z]
                   [--smalt_scoring SMALT_SCORING] [--mapper_args MAPPER_ARGS]
                   [-h] [--spades_exe SPADES_EXE]
                   [--samtools_exe SAMTOOLS_EXE] [--smalt_exe SMALT_EXE]
                   [--bwa_exe BWA_EXE] [--quast_exe QUAST_EXE]
                   [--python2_7_exe PYTHON2_7_EXE]
                   clustered_loci_txt

Given cluster file of rDNA regions from riboSelect and either paired-end or
single-end reads, assembles the mapped reads into pseduocontig &#39;seeds&#39;, and
uses those with the reads to runde fere novo and de novo assembly with SPAdes

positional arguments:
  clustered_loci_txt    output from riboSelect

required named arguments:
  -r REFERENCE_GENBANK, --reference_genbank REFERENCE_GENBANK
                        genbank reference, used to estimate insert sizes, and
                        compare with QUAST
  -o OUTPUT, --output OUTPUT
                        output directory; default:
                        /home/nicholas/GitHub/riboSeed

optional arguments:
  -F FASTQ1, --fastq1 FASTQ1
                        forward fastq reads, can be compressed
  -R FASTQ2, --fastq2 FASTQ2
                        reverse fastq reads, can be compressed
  -S1 FASTQS1, --fastq_single1 FASTQS1
                        single fastq reads
  -n EXP_NAME, --experiment_name EXP_NAME
                        prefix for results files; default: riboSeed
  -l FLANKING, --flanking_length FLANKING
                        length of flanking regions, in bp; default: 1000
  -m {smalt,bwa}, --method_for_map {smalt,bwa}
                        available mappers: smalt and bwa; default: bwa
  -c CORES, --cores CORES
                        cores for multiprocessing; default: None
  -k KMERS, --kmers KMERS
                        kmers used for final assembly, separated by commas;
                        default: 21,33,55,77,99,127
  -p PRE_KMERS, --pre_kmers PRE_KMERS
                        kmers used during seeding assemblies, separated bt
                        commas; default: 21,33,55,77,99
 -s SCORE_MIN, --score_min SCORE_MIN
                        If using smalt, this sets the &#39;-m&#39; param; default with
                        smalt is inferred from read length. If using BWA,
                        reads mapping with ASscore lower than this will be
                        rejected; default with SWA is half of read length
   -a MIN_ASSEMBLY_LEN, --min_assembly_len MIN_ASSEMBLY_LEN
                        if initial SPAdes assembly largest contig is not at
                        least as long as --min_assembly_len, exit. Set this to
                        the length of the seed sequence; if it is not
                        achieved, seeding across regions will likely fail;
                        default: 6000
  --include_shorts      if assembled contig is smaller than
                        --min_assembly_len, contig will still be included in
                        assembly; default: inferred
  --linear              if genome is known to not be circular and a region of
                        interest (including flanking bits) extends past
                        chromosome end, this extends the seqence past
                        chromosome origin forward by --padding; default: False
  --ref_as_contig {None,trusted,untrusted}
                        if &#39;trusted&#39;, SPAdes will use the seed sequences as a
                        --trusted-contig; if &#39;untrusted&#39;, SPAdes will treat as
                        --untrusted-contig. if &#39;&#39;, seeds will not be used
                        during assembly. See SPAdes docs; default: untrusted
  --keep_temps          if not --keep_temps, mapping files will be removed
                        once they are no no longer needed during the
                        iterations; default: False
  --skip_control        if --skip_control, no de novo assembly will be done;
                        default: False
  -i ITERATIONS, --iterations ITERATIONS
                        if iterations&gt;1, multiple seedings will occur after
                        subassembly of seed regions; if setting --target_len,
                        seedings will continue until --iterations are
                        completed or --target_len is matched or exceeded;
                        default: 3
  -v {1,2,3,4,5}, --verbosity {1,2,3,4,5}
                        Logger writes debug to file in output dir; this sets
                        verbosity level sent to stderr. 1 = debug(), 2 =
                        info(), 3 = warning(), 4 = error() and 5 = critical();
                        default: 2
  --target_len TARGET_LEN
                        if set, iterations will continue until contigs reach
                        this length, or max iterations (set by --iterations)
                        have been completed. Set as fraction of original seed
                        length by giving a decimal between 0 and 5, or set as
                        an absolute number of base pairs by giving an integer
                        greater than 50. Not used by default
  -t {1,2,4}, --threads {1,2,4}
                        if your cores are hyperthreaded, set number threads to
                        the number of threads per processer.If unsure, see
                        &#39;cat /proc/cpuinfo&#39; under &#39;cpu cores&#39;, or &#39;lscpu&#39;
                        under &#39;Thread(s) per core&#39;.: 1
  -z, --serialize       if --serialize, runs seeding and assembly without
                        multiprocessing. This is recommended for machines with
                        less than 8GB RAM: False
  --smalt_scoring SMALT_SCORING
                        if mapping with SMALT, submit custom smalt scoring via
                        smalt -S scorespec option; default:
                        match=1,subst=-4,gapopen=-4,gapext=-3
  --mapper_args MAPPER_ARGS
                        submit custom parameters to mapper. And by mapper, I
                        mean bwa, cause we dont support this option for SMALT,
                        sorry. This requires knowledge of your chosen mapper&#39;s
                        optional arguments. Proceed with caution! default: -L
                        0,0 -U 0
  -h, --help            Displays this help message
  --spades_exe SPADES_EXE
                        Path to SPAdes executable; default: spades.py
  --samtools_exe SAMTOOLS_EXE
                        Path to samtools executable; default: samtools
  --smalt_exe SMALT_EXE
                        Path to smalt executable; default: smalt
  --bwa_exe BWA_EXE     Path to BWA executable; default: bwa
  --quast_exe QUAST_EXE
                        Path to quast executable; default: quast.py
  --python2_7_exe PYTHON2_7_EXE
                        Path to python2.7 executable, cause QUAST won&#39;t run on
                        python3. default: python2.7
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="key-parameters">
<h2>Key Parameters<a class="headerlink" href="#key-parameters" title="Permalink to this headline">¶</a></h2>
<p>Results can be tuned by changing several of the default parameters.</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">--score_min</span></code>: This can be used to set the minimum mapping score.
If using BWA, the default is not to supply a minimum and to rely on
the BWA default. If submitting a <code class="docutils literal"><span class="pre">--score_min</span></code> to BWA, double check
that it is appropriate. It appears to be extremely sensitive to read
length, and having a too-low threshold for minimum mapping can
seriously ruin ones day. Check out IGB or similar to view your
mappings if greater than, say, 5% or the reads are mapping in
subsequent iterations.</li>
<li><code class="docutils literal"><span class="pre">-l,</span> <span class="pre">--flanking_length</span></code>: Default is 2000. That seems to be a good
compromise between gaining unique sequence and not relying too much
on the reference.</li>
<li><code class="docutils literal"><span class="pre">--kmers</span></code> and <code class="docutils literal"><span class="pre">--pre_kmers</span></code>: Adjust these as you otherwise would
for a <em>de novo</em> assembly.</li>
<li><code class="docutils literal"><span class="pre">--min_assembly_len</span></code>: For bacteria, this is about 7000bp, as the
rDNA regions for a typical operon of 16S 23S and 5S coding sequences
combined usually are about that long. If you are using non-standard
rDNA regions, this should be adjusted to prevent spurious assemblies.</li>
<li><code class="docutils literal"><span class="pre">--ref_as_contig</span></code>: This can be used to guide how SPAdes treats the
long read sequences during the assembly (<code class="docutils literal"><span class="pre">trusted</span></code> or
<code class="docutils literal"><span class="pre">untrusted</span></code>). By default, this is infered from mapping percentage
(<code class="docutils literal"><span class="pre">trusted</span></code> if over 85% of reads map to the reference)</li>
<li><code class="docutils literal"><span class="pre">--iterations</span></code>: Each iteration typically increases the length of
the long read by approximately 5%.</li>
</ul>
</div>
<div class="section" id="visualization-assessment">
<h2>3: Visualization/Assessment<a class="headerlink" href="#visualization-assessment" title="Permalink to this headline">¶</a></h2>
<div class="section" id="ribosnag-py">
<h3><code class="docutils literal"><span class="pre">riboSnag.py</span></code><a class="headerlink" href="#ribosnag-py" title="Permalink to this headline">¶</a></h3>
<p><code class="docutils literal"><span class="pre">riboSnag.py</span></code> takes the list of clustered locus tags and extracts
their sequences with flanking regions, optionally turning the coding
sequences to N’s to minimize bias towards reference. Is used to pull out
regions of interest from a Genbank file. Outputs a directory with a
fasta file for each clustered region (and a log file).</p>
<p>Additionally, it does a lot of plotting to visualize the Shannon
entropy, coverage, occurrences, and other useful metrics.</p>
<div class="section" id="id2">
<h4>Usage:<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h4>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">riboSnag</span><span class="o">.</span><span class="n">py</span> <span class="p">[</span><span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">n</span> <span class="n">NAME</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">l</span> <span class="n">FLANKING</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">msa_kmers</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">c</span><span class="p">]</span>
                   <span class="p">[</span><span class="o">-</span><span class="n">p</span> <span class="n">PADDING</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span> <span class="n">VERBOSITY</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">clobber</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">no_revcomp</span><span class="p">]</span>
                   <span class="p">[</span><span class="o">--</span><span class="n">skip_check</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">msa_tool</span> <span class="p">{</span><span class="n">mafft</span><span class="p">,</span><span class="n">prank</span><span class="p">}]</span>
                   <span class="p">[</span><span class="o">--</span><span class="n">prank_exe</span> <span class="n">PRANK_EXE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">mafft_exe</span> <span class="n">MAFFT_EXE</span><span class="p">]</span>
                   <span class="p">[</span><span class="o">--</span><span class="n">barrnap_exe</span> <span class="n">BARRNAP_EXE</span><span class="p">]</span>
                   <span class="p">[</span><span class="o">--</span><span class="n">makeblastdb_exe</span> <span class="n">MAKEBLASTDB_EXE</span><span class="p">]</span>
                   <span class="p">[</span><span class="o">--</span><span class="n">kingdom</span> <span class="p">{</span><span class="n">mito</span><span class="p">,</span><span class="n">euk</span><span class="p">,</span><span class="n">arc</span><span class="p">,</span><span class="n">bac</span><span class="p">}]</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span>
                   <span class="n">genbank_genome</span> <span class="n">clustered_loci</span>

<span class="n">Use</span> <span class="n">to</span> <span class="n">extract</span> <span class="n">regions</span> <span class="n">of</span> <span class="n">interest</span> <span class="n">based</span> <span class="n">on</span> <span class="n">supplied</span> <span class="n">Locus</span> <span class="n">tags</span> <span class="ow">and</span> <span class="n">evaluate</span>
<span class="n">the</span> <span class="n">extracted</span> <span class="n">regions</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">genbank_genome</span>        <span class="n">Genbank</span> <span class="n">file</span> <span class="p">(</span><span class="n">WITH</span> <span class="n">SEQUENCE</span><span class="p">)</span>
  <span class="n">clustered_loci</span>        <span class="n">output</span> <span class="kn">from</span> <span class="nn">riboSelect</span>

<span class="n">required</span> <span class="n">named</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">,</span> <span class="o">--</span><span class="n">output</span> <span class="n">OUTPUT</span>
                        <span class="n">output</span> <span class="n">directory</span><span class="p">;</span> <span class="n">default</span><span class="p">:</span>
                        <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">nicholas</span><span class="o">/</span><span class="n">GitHub</span><span class="o">/</span><span class="n">riboSeed</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">n</span> <span class="n">NAME</span><span class="p">,</span> <span class="o">--</span><span class="n">name</span> <span class="n">NAME</span>  <span class="n">rename</span> <span class="n">the</span> <span class="n">contigs</span> <span class="k">with</span> <span class="n">this</span> <span class="n">prefixdefault</span><span class="p">:</span> <span class="n">date</span>
                        <span class="p">(</span><span class="n">YYYYMMDD</span><span class="p">)</span>
  <span class="o">-</span><span class="n">l</span> <span class="n">FLANKING</span><span class="p">,</span> <span class="o">--</span><span class="n">flanking_length</span> <span class="n">FLANKING</span>
                        <span class="n">length</span> <span class="n">of</span> <span class="n">flanking</span> <span class="n">regions</span><span class="p">,</span> <span class="ow">in</span> <span class="n">bp</span><span class="p">;</span> <span class="n">default</span><span class="p">:</span> <span class="mi">1000</span>
  <span class="o">--</span><span class="n">msa_kmers</span>           <span class="n">calculate</span> <span class="n">kmer</span> <span class="n">similarity</span> <span class="n">based</span> <span class="n">on</span> <span class="n">aligned</span> <span class="n">sequences</span>
                        <span class="n">instead</span> <span class="n">of</span> <span class="n">raw</span> <span class="n">sequences</span><span class="p">;</span><span class="n">default</span><span class="p">:</span> <span class="kc">False</span>
  <span class="o">-</span><span class="n">c</span><span class="p">,</span> <span class="o">--</span><span class="n">circular</span>        <span class="k">if</span> <span class="n">the</span> <span class="n">genome</span> <span class="ow">is</span> <span class="n">known</span> <span class="n">to</span> <span class="n">be</span> <span class="n">circular</span><span class="p">,</span> <span class="ow">and</span> <span class="n">an</span> <span class="n">region</span>
                        <span class="n">of</span> <span class="n">interest</span> <span class="p">(</span><span class="n">including</span> <span class="n">flanking</span> <span class="n">bits</span><span class="p">)</span> <span class="n">extends</span> <span class="n">past</span>
                        <span class="n">chromosome</span> <span class="n">end</span><span class="p">,</span> <span class="n">this</span> <span class="n">extends</span> <span class="n">the</span> <span class="n">seqence</span> <span class="n">past</span>
                        <span class="n">chromosome</span> <span class="n">origin</span> <span class="n">forward</span> <span class="n">by</span> <span class="mi">5</span><span class="n">kb</span><span class="p">;</span> <span class="n">default</span><span class="p">:</span> <span class="kc">False</span>
  <span class="o">-</span><span class="n">p</span> <span class="n">PADDING</span><span class="p">,</span> <span class="o">--</span><span class="n">padding</span> <span class="n">PADDING</span>
                        <span class="k">if</span> <span class="n">treating</span> <span class="k">as</span> <span class="n">circular</span><span class="p">,</span> <span class="n">this</span> <span class="n">controls</span> <span class="n">the</span> <span class="n">length</span> <span class="n">of</span>
                        <span class="n">sequence</span> <span class="n">added</span> <span class="n">to</span> <span class="n">the</span> <span class="mi">5</span><span class="s1">&#39; and 3&#39;</span> <span class="n">ends</span> <span class="n">to</span> <span class="n">allow</span> <span class="k">for</span>
                        <span class="n">selecting</span> <span class="n">regions</span> <span class="n">that</span> <span class="n">cross</span> <span class="n">the</span> <span class="n">chromosom</span><span class="s1">&#39;s origin;</span>
                        <span class="n">default</span><span class="p">:</span> <span class="mi">5000</span>
  <span class="o">-</span><span class="n">v</span> <span class="n">VERBOSITY</span><span class="p">,</span> <span class="o">--</span><span class="n">verbosity</span> <span class="n">VERBOSITY</span>
                        <span class="mi">1</span> <span class="o">=</span> <span class="n">debug</span><span class="p">(),</span> <span class="mi">2</span> <span class="o">=</span> <span class="n">info</span><span class="p">(),</span> <span class="mi">3</span> <span class="o">=</span> <span class="n">warning</span><span class="p">(),</span> <span class="mi">4</span> <span class="o">=</span> <span class="n">error</span><span class="p">()</span>
                        <span class="ow">and</span> <span class="mi">5</span> <span class="o">=</span> <span class="n">critical</span><span class="p">();</span> <span class="n">default</span><span class="p">:</span> <span class="mi">2</span>
  <span class="o">--</span><span class="n">clobber</span>             <span class="n">overwrite</span> <span class="n">previous</span> <span class="n">output</span> <span class="n">filesdefault</span><span class="p">:</span> <span class="kc">False</span>
  <span class="o">--</span><span class="n">no_revcomp</span>          <span class="n">default</span> <span class="n">returns</span> <span class="n">reverse</span> <span class="n">complimented</span> <span class="n">seq</span> <span class="k">if</span> <span class="n">majority</span>
                        <span class="n">of</span> <span class="n">regions</span> <span class="n">on</span> <span class="n">reverse</span> <span class="n">strand</span><span class="o">.</span> <span class="k">if</span> <span class="o">--</span><span class="n">no_revcomp</span><span class="p">,</span> <span class="n">this</span> <span class="ow">is</span>
                        <span class="n">overwriddendefault</span><span class="p">:</span> <span class="kc">False</span>
  <span class="o">--</span><span class="n">skip_check</span>          <span class="n">Dont</span> <span class="n">bother</span> <span class="n">calculating</span> <span class="n">Shannon</span> <span class="n">Entropy</span><span class="p">;</span> <span class="n">default</span><span class="p">:</span>
                        <span class="kc">False</span>
  <span class="o">--</span><span class="n">msa_tool</span> <span class="p">{</span><span class="n">mafft</span><span class="p">,</span><span class="n">prank</span><span class="p">}</span>
                        <span class="n">Path</span> <span class="n">to</span> <span class="n">PRANK</span> <span class="n">executable</span><span class="p">;</span> <span class="n">default</span><span class="p">:</span> <span class="n">mafft</span>
  <span class="o">--</span><span class="n">prank_exe</span> <span class="n">PRANK_EXE</span>
                        <span class="n">Path</span> <span class="n">to</span> <span class="n">PRANK</span> <span class="n">executable</span><span class="p">;</span> <span class="n">default</span><span class="p">:</span> <span class="n">prank</span>
  <span class="o">--</span><span class="n">mafft_exe</span> <span class="n">MAFFT_EXE</span>
                        <span class="n">Path</span> <span class="n">to</span> <span class="n">MAFFT</span> <span class="n">executable</span><span class="p">;</span> <span class="n">default</span><span class="p">:</span> <span class="n">mafft</span>
  <span class="o">--</span><span class="n">barrnap_exe</span> <span class="n">BARRNAP_EXE</span>
                        <span class="n">Path</span> <span class="n">to</span> <span class="n">barrnap</span> <span class="n">executable</span><span class="p">;</span> <span class="n">default</span><span class="p">:</span> <span class="n">barrnap</span>
  <span class="o">--</span><span class="n">makeblastdb_exe</span> <span class="n">MAKEBLASTDB_EXE</span>
                        <span class="n">Path</span> <span class="n">to</span> <span class="n">makeblastdb</span> <span class="n">executable</span><span class="p">;</span> <span class="n">default</span><span class="p">:</span> <span class="n">makeblastdb</span>
  <span class="o">--</span><span class="n">kingdom</span> <span class="p">{</span><span class="n">mito</span><span class="p">,</span><span class="n">euk</span><span class="p">,</span><span class="n">arc</span><span class="p">,</span><span class="n">bac</span><span class="p">}</span>
                        <span class="n">kingdom</span> <span class="k">for</span> <span class="n">barrnap</span><span class="p">;</span> <span class="n">default</span><span class="p">:</span> <span class="n">bac</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">Displays</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="ribostack-py">
<h3><code class="docutils literal"><span class="pre">riboStack.py</span></code><a class="headerlink" href="#ribostack-py" title="Permalink to this headline">¶</a></h3>
<p>Decause assembly using short reads often collases rDNA repeats, it is
not uncommon to find a reference genome that has less than the actual
number of rDNAs. riboStack uses <code class="docutils literal"><span class="pre">bedtools</span></code> and <code class="docutils literal"><span class="pre">samtools</span></code> to
determine the coverage across rDNA regiosn, adn compares that coverage
depth to 10 sets of randomly selected non-rDNA regions. If the number of
rDNAs in the reference matches the number of rDNAs in your sequecned
isolate, the coverage should be pretty similar. However, if the coverage
in your rDNA regions is significantly higher, than there are likely more
rDNAs in your sequenced isoalte that there are in the reference, which
is something to be aware of.</p>
<p>It requires a mapping BAM file and the riboScan output directory as
input.</p>
</div>
<div class="section" id="riboswap-py">
<h3><code class="docutils literal"><span class="pre">riboSwap.py</span></code><a class="headerlink" href="#riboswap-py" title="Permalink to this headline">¶</a></h3>
<p>Infrequently, <code class="docutils literal"><span class="pre">riboSeed</span></code> has joined together contigs that appear
incorrect according to your reference. If you are at all unhappy with a
bridging, <code class="docutils literal"><span class="pre">riboSwap.py</span></code> allows swapping of a “bad” contig for one or
more syntenic contigs from the <em>de novo</em> assembly. #### USAGE</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">riboSwap</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span> <span class="p">{</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">}]</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span>
                   <span class="n">de_novo_file</span> <span class="n">de_fere_novo_file</span> <span class="n">bad_contig</span> <span class="n">good_contigs</span>

<span class="n">Given</span> <span class="n">de</span> <span class="n">novo</span> <span class="ow">and</span> <span class="n">de</span> <span class="n">fere</span> <span class="n">novo</span> <span class="n">contigs</span> <span class="n">files</span><span class="p">,</span> <span class="n">a</span> <span class="n">misjoined</span> <span class="n">de</span> <span class="n">fere</span> <span class="n">novo</span> <span class="n">contig</span>
<span class="n">name</span><span class="p">,</span> <span class="ow">and</span> <span class="n">a</span> <span class="n">colon</span><span class="p">:</span><span class="n">separated</span> <span class="nb">list</span> <span class="n">of</span> <span class="n">de</span> <span class="n">novo</span> <span class="n">contig</span> <span class="n">names</span><span class="p">,</span> <span class="n">replace</span> <span class="n">the</span>
<span class="n">offending</span> <span class="n">contig</span> <span class="k">with</span> <span class="n">the</span> <span class="n">de</span> <span class="n">novo</span> <span class="n">contig</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>

<span class="n">positional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="n">de_novo_file</span>          <span class="n">multifasta</span> <span class="n">containing</span> <span class="n">de</span> <span class="n">novo</span> <span class="n">contigs</span>
  <span class="n">de_fere_novo_file</span>     <span class="n">multifasta</span> <span class="n">containing</span> <span class="n">de</span> <span class="n">fere</span> <span class="n">novo</span> <span class="n">contigs</span>
  <span class="n">bad_contig</span>            <span class="n">name</span> <span class="n">of</span> <span class="n">the</span> <span class="n">bad</span> <span class="n">contig</span>
  <span class="n">good_contigs</span>          <span class="n">colon</span> <span class="n">separated</span> <span class="n">good</span> <span class="n">contigs</span> <span class="k">for</span> <span class="n">replacement</span>

<span class="n">required</span> <span class="n">named</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">o</span> <span class="n">OUTPUT</span><span class="p">,</span> <span class="o">--</span><span class="n">output</span> <span class="n">OUTPUT</span>
                        <span class="n">output</span> <span class="n">directory</span><span class="p">;</span> <span class="n">default</span><span class="p">:</span>
                        <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">nicholas</span><span class="o">/</span><span class="n">GitHub</span><span class="o">/</span><span class="n">riboSeed</span>

<span class="n">optional</span> <span class="n">arguments</span><span class="p">:</span>
  <span class="o">-</span><span class="n">v</span> <span class="p">{</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">},</span> <span class="o">--</span><span class="n">verbosity</span> <span class="p">{</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">}</span>
                        <span class="n">Logger</span> <span class="n">writes</span> <span class="n">debug</span> <span class="n">to</span> <span class="n">file</span> <span class="ow">in</span> <span class="n">output</span> <span class="nb">dir</span><span class="p">;</span> <span class="n">this</span> <span class="n">sets</span>
                        <span class="n">verbosity</span> <span class="n">level</span> <span class="n">sent</span> <span class="n">to</span> <span class="n">stderr</span><span class="o">.</span> <span class="mi">1</span> <span class="o">=</span> <span class="n">debug</span><span class="p">(),</span> <span class="mi">2</span> <span class="o">=</span>
                        <span class="n">info</span><span class="p">(),</span> <span class="mi">3</span> <span class="o">=</span> <span class="n">warning</span><span class="p">(),</span> <span class="mi">4</span> <span class="o">=</span> <span class="n">error</span><span class="p">()</span> <span class="ow">and</span> <span class="mi">5</span> <span class="o">=</span> <span class="n">critical</span><span class="p">();</span>
                        <span class="n">default</span><span class="p">:</span> <span class="mi">2</span>
  <span class="o">-</span><span class="n">h</span><span class="p">,</span> <span class="o">--</span><span class="n">help</span>            <span class="n">Displays</span> <span class="n">this</span> <span class="n">help</span> <span class="n">message</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="known-bugs">
<h2>Known Bugs<a class="headerlink" href="#known-bugs" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li>Submitting <code class="docutils literal"><span class="pre">--smalt_scoring</span></code> with vastly different scoring schemes
usually causes an error.</li>
</ul>
</div>
<div class="section" id="running-tests">
<h2>Running Tests<a class="headerlink" href="#running-tests" title="Permalink to this headline">¶</a></h2>
<p>The tests for the module can be found under the <code class="docutils literal"><span class="pre">tests</span></code> directory. I
run them with the unittests module. The tests assume the installation of
all the recommended tools.</p>
</div>
<div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<p>riboSeed is on Pypi, so you can install with pip, preferably within a
virtualenv (recommended):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">virtualenv</span> <span class="o">-</span><span class="n">p</span> <span class="n">python3</span><span class="o">.</span><span class="mi">5</span> <span class="n">venv</span><span class="o">-</span><span class="n">riboSeed</span>
<span class="n">source</span> <span class="n">venv</span><span class="o">-</span><span class="n">riboSeed</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">activate</span>
<span class="n">pip3</span><span class="o">.</span><span class="mi">5</span> <span class="n">install</span> <span class="n">riboSeed</span>
</pre></div>
</div>
<p>To install the bleeding-edge version, install from testpypi:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">virtualenv</span> <span class="o">-</span><span class="n">p</span> <span class="n">python3</span><span class="o">.</span><span class="mi">5</span> <span class="n">venv</span><span class="o">-</span><span class="n">riboSeed</span>
<span class="n">source</span> <span class="n">venv</span><span class="o">-</span><span class="n">riboSeed</span><span class="o">/</span><span class="nb">bin</span><span class="o">/</span><span class="n">activate</span>
<span class="n">pip</span> <span class="n">install</span> <span class="o">--</span><span class="n">extra</span><span class="o">-</span><span class="n">index</span><span class="o">-</span><span class="n">url</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">testpypi</span><span class="o">.</span><span class="n">python</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="n">pypi</span> <span class="n">riboSeed</span>
</pre></div>
</div>
<p>You can also clone this repository, and run
<code class="docutils literal"><span class="pre">python3.5</span> <span class="pre">setup.py</span> <span class="pre">install</span></code>.</p>
<p>In the scripts directory, there is a script called <code class="docutils literal"><span class="pre">runme.py</span></code> which
run the pipeline on a small sample dataset. it should output a folder to
your current working dir called <code class="docutils literal"><span class="pre">integration_tests</span></code>.</p>
<blockquote>
<div>Dependencies</div></blockquote>
<hr class="docutils" />
<div class="section" id="python-requirements">
<h3>Python Requirements:<a class="headerlink" href="#python-requirements" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>Python &gt;= v3.5</li>
<li>Biopython v1.68</li>
<li>pysam v0.9.1.4,</li>
<li>pyutilsnrw &gt;= 0.0.768</li>
<li>matplotlib v1.5.3</li>
<li>pandas v0.18.1</li>
</ul>
</div>
<div class="section" id="external-requirements">
<h3>External Requirements<a class="headerlink" href="#external-requirements" title="Permalink to this headline">¶</a></h3>
<p>riboScan.py</p>
<ul class="simple">
<li>Barrnap (must be 0.7 or above)</li>
<li>Seqret</li>
</ul>
<p>riboSelect.py</p>
<ul class="simple">
<li>None</li>
</ul>
<p>riboSnag.py</p>
<ul class="simple">
<li>PRANK or Mafft</li>
<li>BLAST+ suite</li>
<li>Barrnap (must be 0.7 or above)</li>
</ul>
<p>riboSeed.py</p>
<ul class="simple">
<li>SPAdes v3.8 or higher</li>
<li>BWA (tested with 0.7.12-r1039)</li>
<li>SAMTools (must be 1.3.1 or above)</li>
<li>QUAST (tested with 4.1)</li>
</ul>
<p>NOTE: barrnap has certain Perl requirements that may not be included on
your machine. Ensure barrnap runs fine before trying <code class="docutils literal"><span class="pre">riboSnag.py</span></code>. Or
try <a class="reference external" href="https://github.com/nickp60/barrnap/">python barrnap</a>.</p>
</div>
</div>
<div class="section" id="suggested-running">
<h2>Suggested Running<a class="headerlink" href="#suggested-running" title="Permalink to this headline">¶</a></h2>
<div class="section" id="ribobatch-sh">
<h3><code class="docutils literal"><span class="pre">riboBatch.sh</span></code><a class="headerlink" href="#ribobatch-sh" title="Permalink to this headline">¶</a></h3>
<p>NOTE: You must have riboSeed installed in a <code class="docutils literal"><span class="pre">virtualenv</span></code> to use this
script</p>
<p>There are a lot of commandline options for riboSeed, so it can help to
run the pipeline as script. The included <code class="docutils literal"><span class="pre">riboBatch.sh</span></code> script is an
example.</p>
<p>If you have access to a hpc, you can set up a python virtualenv, edit
the 7 fields in this script, make any other modifications needed to fit
your job, and submit with <code class="docutils literal"><span class="pre">qsub</span></code>.</p>
<p>We recommend copying this file to your project directory, and
customizing it as needed.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">riboSeed Pipeline</a><ul>
<li><a class="reference internal" href="#description">Description</a></li>
<li><a class="reference internal" href="#before-we-start">Before We Start</a></li>
<li><a class="reference internal" href="#preprocessing">0: Preprocessing</a><ul>
<li><a class="reference internal" href="#riboscan-py"><code class="docutils literal"><span class="pre">riboScan.py</span></code></a></li>
<li><a class="reference internal" href="#riboselect-py"><code class="docutils literal"><span class="pre">riboSelect.py</span></code></a><ul>
<li><a class="reference internal" href="#usage">Usage</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#de-fere-novo-assembly">2: <em>De fere novo</em> Assembly</a><ul>
<li><a class="reference internal" href="#riboseed-py"><code class="docutils literal"><span class="pre">riboSeed.py</span></code></a><ul>
<li><a class="reference internal" href="#output">Output</a><ul>
<li><a class="reference internal" href="#note">NOTE:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#id1">Usage:</a></li>
</ul>
</li>
</ul>
</li>
<li><a class="reference internal" href="#key-parameters">Key Parameters</a></li>
<li><a class="reference internal" href="#visualization-assessment">3: Visualization/Assessment</a><ul>
<li><a class="reference internal" href="#ribosnag-py"><code class="docutils literal"><span class="pre">riboSnag.py</span></code></a><ul>
<li><a class="reference internal" href="#id2">Usage:</a></li>
</ul>
</li>
<li><a class="reference internal" href="#ribostack-py"><code class="docutils literal"><span class="pre">riboStack.py</span></code></a></li>
<li><a class="reference internal" href="#riboswap-py"><code class="docutils literal"><span class="pre">riboSwap.py</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#known-bugs">Known Bugs</a></li>
<li><a class="reference internal" href="#running-tests">Running Tests</a></li>
<li><a class="reference internal" href="#installation">Installation</a><ul>
<li><a class="reference internal" href="#python-requirements">Python Requirements:</a></li>
<li><a class="reference internal" href="#external-requirements">External Requirements</a></li>
</ul>
</li>
<li><a class="reference internal" href="#suggested-running">Suggested Running</a><ul>
<li><a class="reference internal" href="#ribobatch-sh"><code class="docutils literal"><span class="pre">riboBatch.sh</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Welcome to riboSeed’s documentation!</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/README.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="index.html" title="Welcome to riboSeed’s documentation!"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">riboSeed 0.4.9 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2017, Nicholas Waters, Florence Abram, Fiona Brennan, Ashleigh Holmes, and Leighton Pritchard.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.3.
    </div>
  </body>
</html>