
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>riboSeed package &#8212; riboSeed 0.4.9 documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.4.9',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
  </head>
  <body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">riboSeed 0.4.9 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="riboseed-package">
<h1>riboSeed package<a class="headerlink" href="#riboseed-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="riboSeed.integration_data.html">riboSeed.integration_data package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="riboSeed.integration_data.html#module-contents">Module contents</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="riboseed-plotmsa-module">
<h2>riboSeed.plotMSA module<a class="headerlink" href="#riboseed-plotmsa-module" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="riboseed-riboscan-module">
<h2>riboSeed.riboScan module<a class="headerlink" href="#riboseed-riboscan-module" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="riboseed-riboscore-module">
<h2>riboSeed.riboScore module<a class="headerlink" href="#riboseed-riboscore-module" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="riboseed-riboseed-module">
<h2>riboSeed.riboSeed module<a class="headerlink" href="#riboseed-riboseed-module" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="riboseed-riboselect-module">
<h2>riboSeed.riboSelect module<a class="headerlink" href="#riboseed-riboselect-module" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="riboseed-ribosim-module">
<h2>riboSeed.riboSim module<a class="headerlink" href="#riboseed-ribosim-module" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="riboseed-ribosketch-module">
<h2>riboSeed.riboSketch module<a class="headerlink" href="#riboseed-ribosketch-module" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="riboseed-ribosnag-module">
<h2>riboSeed.riboSnag module<a class="headerlink" href="#riboseed-ribosnag-module" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="riboseed-ribostack-module">
<h2>riboSeed.riboStack module<a class="headerlink" href="#riboseed-ribostack-module" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="riboseed-riboswap-module">
<h2>riboSeed.riboSwap module<a class="headerlink" href="#riboseed-riboswap-module" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-riboSeed">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-riboSeed" title="Permalink to this headline">¶</a></h2>
<p>Created on Sun Jul 24 19:33:37 2016</p>
<p>See README.md for more info and usage</p>
<dl class="class">
<dt id="riboSeed.Exes">
<em class="property">class </em><code class="descclassname">riboSeed.</code><code class="descname">Exes</code><span class="sig-paren">(</span><em>samtools</em>, <em>method</em>, <em>spades</em>, <em>quast</em>, <em>python2_7</em>, <em>smalt</em>, <em>bwa</em>, <em>check=True</em>, <em>mapper=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#Exes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.Exes" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>given the amount of system tools that riboSeed requires, this object
holds the paths to the executables after expanding the user-supplied
path and verifying with shutil.which that the executable is availible
to the program.</p>
<dl class="method">
<dt id="riboSeed.Exes.check_expand_exes">
<code class="descname">check_expand_exes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#Exes.check_expand_exes"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.Exes.check_expand_exes" title="Permalink to this definition">¶</a></dt>
<dd><p>for each executable, expand wildcards and use shutil.which
to get full path to executable.  If not found, throw an error</p>
</dd></dl>

<dl class="method">
<dt id="riboSeed.Exes.check_mands">
<code class="descname">check_mands</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#Exes.check_mands"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.Exes.check_mands" title="Permalink to this definition">¶</a></dt>
<dd><p>checks that all mandatory arguments are not none</p>
</dd></dl>

<dl class="method">
<dt id="riboSeed.Exes.set_mapper">
<code class="descname">set_mapper</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#Exes.set_mapper"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.Exes.set_mapper" title="Permalink to this definition">¶</a></dt>
<dd><p>Exes.mapper attribute is set here to avoid further
“if method ==’smalt’ clauses later.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="riboSeed.LociMapping">
<em class="property">class </em><code class="descclassname">riboSeed.</code><code class="descname">LociMapping</code><span class="sig-paren">(</span><em>name</em>, <em>iteration</em>, <em>mapping_subdir</em>, <em>mapping_prefix=None</em>, <em>assembly_success=False</em>, <em>ref_fasta=None</em>, <em>pe_map_bam=None</em>, <em>s_map_bam=None</em>, <em>sorted_mapped_bam=None</em>, <em>mapped_bam=None</em>, <em>mapped_sam=None</em>, <em>mapped_bam_unfiltered=None</em>, <em>unmapped_sam=None</em>, <em>mapped_ids_txt=None</em>, <em>unmapped_bam=None</em>, <em>mappedS=None</em>, <em>assembled_contig=None</em>, <em>assembly_subdir=None</em>, <em>mapped_ngsLib=None</em>, <em>unmapped_ngsLib=None</em>, <em>assembly_subdir_needed=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#LociMapping"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.LociMapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>instantiate with iteration, mapping subdir, and ref
map_ref_genome_will use it and an ngslib for mapping
<a href="#id1"><span class="problematic" id="id2">extract_</span></a>
order of operations: map to reference, extract and convert,
assemble, save results here</p>
<dl class="method">
<dt id="riboSeed.LociMapping.check_mands">
<code class="descname">check_mands</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#LociMapping.check_mands"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.LociMapping.check_mands" title="Permalink to this definition">¶</a></dt>
<dd><p>checks that all mandatory arguments are not none</p>
</dd></dl>

<dl class="method">
<dt id="riboSeed.LociMapping.make_assembly_subdir">
<code class="descname">make_assembly_subdir</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#LociMapping.make_assembly_subdir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.LociMapping.make_assembly_subdir" title="Permalink to this definition">¶</a></dt>
<dd><p>make a subdirectory for assembly if it is needed</p>
</dd></dl>

<dl class="method">
<dt id="riboSeed.LociMapping.make_mapping_subdir">
<code class="descname">make_mapping_subdir</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#LociMapping.make_mapping_subdir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.LociMapping.make_mapping_subdir" title="Permalink to this definition">¶</a></dt>
<dd><p>make a subdirectory for mapping results</p>
</dd></dl>

<dl class="method">
<dt id="riboSeed.LociMapping.name_bams_and_sams">
<code class="descname">name_bams_and_sams</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#LociMapping.name_bams_and_sams"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.LociMapping.name_bams_and_sams" title="Permalink to this definition">¶</a></dt>
<dd><p>make a prefix and use it to name the future output files</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="riboSeed.NgsLib">
<em class="property">class </em><code class="descclassname">riboSeed.</code><code class="descname">NgsLib</code><span class="sig-paren">(</span><em>name</em>, <em>master=False</em>, <em>readF=None</em>, <em>readR=None</em>, <em>readS0=None</em>, <em>readS1=None</em>, <em>mapping_success=False</em>, <em>smalt_dist_path=None</em>, <em>readlen=None</em>, <em>make_dist=False</em>, <em>libtype=None</em>, <em>logger=None</em>, <em>mapper_exe=None</em>, <em>liblist=None</em>, <em>ref_fasta=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#NgsLib"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.NgsLib" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>NgsLib objects are used to hold the sequencing data supplied by the
user (master) and the seq data extracted from each iteration. Currently the
software requires either a paired-end or single-end library, but this
should handle more diverse library types in the future.</p>
<p>If ngsLib is master, read lengths are determined and if smalt is used for
mapping, a distance estimation file is generated.</p>
<p>when instatiated,self.check_mands() ensure required attributes have values,
self.set_libtype() sets the libtype attribute based on libraries present,
self.get_readlen() determines the read length if master, and  if master,
self.smalt_insert_file() creates a distance estimation file for smalt</p>
<dl class="method">
<dt id="riboSeed.NgsLib.check_mands">
<code class="descname">check_mands</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#NgsLib.check_mands"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.NgsLib.check_mands" title="Permalink to this definition">¶</a></dt>
<dd><p>checks that all mandatory arguments are not none</p>
</dd></dl>

<dl class="method">
<dt id="riboSeed.NgsLib.get_readlen">
<code class="descname">get_readlen</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#NgsLib.get_readlen"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.NgsLib.get_readlen" title="Permalink to this definition">¶</a></dt>
<dd><p>If NgsLib is master, estimate the read lengh using
get_ave_read_len_from_fastq.</p>
</dd></dl>

<dl class="method">
<dt id="riboSeed.NgsLib.listLibs">
<code class="descname">listLibs</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#NgsLib.listLibs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.NgsLib.listLibs" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="riboSeed.NgsLib.purge_old_files">
<code class="descname">purge_old_files</code><span class="sig-paren">(</span><em>master</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#NgsLib.purge_old_files"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.NgsLib.purge_old_files" title="Permalink to this definition">¶</a></dt>
<dd><p>before reasigning unmapped lib, delete
useless files that were used in the previous iteration
return codes:</p>
<blockquote>
<div><ol class="arabic simple" start="0">
<li>all is well</li>
<li>no files deleted due to conflict</li>
</ol>
</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="riboSeed.NgsLib.set_libtype">
<code class="descname">set_libtype</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#NgsLib.set_libtype"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.NgsLib.set_libtype" title="Permalink to this definition">¶</a></dt>
<dd><p>sets to either s_1, pe, pe_s based on Non-none libraries
TODO: add mate support and support for multiple single libraries</p>
</dd></dl>

<dl class="method">
<dt id="riboSeed.NgsLib.smalt_insert_file">
<code class="descname">smalt_insert_file</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#NgsLib.smalt_insert_file"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.NgsLib.smalt_insert_file" title="Permalink to this definition">¶</a></dt>
<dd><p>Smalt mapper uses a subset of mapped reads to estimate distribution
of insert sizes.  This file is used along with mapping, and is created
if make_dist, master, and lib_type indicates paired data.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="riboSeed.SeedGenome">
<em class="property">class </em><code class="descclassname">riboSeed.</code><code class="descname">SeedGenome</code><span class="sig-paren">(</span><em>genbank_path</em>, <em>final_long_reads_dir=None</em>, <em>this_iteration=0</em>, <em>ref_fasta=None</em>, <em>next_reference_path=None</em>, <em>loci_clusters=None</em>, <em>output_root=None</em>, <em>initial_map_bam=None</em>, <em>unmapped_ngsLib=None</em>, <em>name=None</em>, <em>iter_mapping_list=None</em>, <em>reads_mapped_txt=None</em>, <em>unmapped_mapping_list=None</em>, <em>max_iterations=None</em>, <em>initial_map_sam=None</em>, <em>unmapped_sam=None</em>, <em>clustered_loci_txt=None</em>, <em>seq_records=None</em>, <em>master_ngs_ob=None</em>, <em>initial_map_sorted_bam=None</em>, <em>initial_map_prefix=None</em>, <em>assembled_seeds=None</em>, <em>seq_records_count=None</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#SeedGenome"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.SeedGenome" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>This object is the “master” object which holds slots for each
cluster’s mappings, the sequencing library, and keeps track of the
current iteration as execution progresses.</p>
<p>When instantiated, self.check_mands() checks that all required attributes
are present, self.attach_genome_seqRecords() parses and loads the gb data,
self.write_fasta_genome() writes a .fasta version of the genome for later
use, and self.make_map_paths_and_dir() sets up the required directories for
each iteration</p>
<dl class="method">
<dt id="riboSeed.SeedGenome.check_mands">
<code class="descname">check_mands</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#SeedGenome.check_mands"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.SeedGenome.check_mands" title="Permalink to this definition">¶</a></dt>
<dd><p>check that all mandatory arguments are not none</p>
</dd></dl>

<dl class="method">
<dt id="riboSeed.SeedGenome.countSeqRecords">
<code class="descname">countSeqRecords</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#SeedGenome.countSeqRecords"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.SeedGenome.countSeqRecords" title="Permalink to this definition">¶</a></dt>
<dd><p>Nothing fancy; just a quick way to get the number of records</p>
</dd></dl>

<dl class="method">
<dt id="riboSeed.SeedGenome.make_map_paths_and_dir">
<code class="descname">make_map_paths_and_dir</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#SeedGenome.make_map_paths_and_dir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.SeedGenome.make_map_paths_and_dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a output root, prepare all the needed subdirs and paths</p>
</dd></dl>

<dl class="method">
<dt id="riboSeed.SeedGenome.pad_genbank">
<code class="descname">pad_genbank</code><span class="sig-paren">(</span><em>pad=5000</em>, <em>circular=False</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#SeedGenome.pad_genbank"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.SeedGenome.pad_genbank" title="Permalink to this definition">¶</a></dt>
<dd><p>if the genome is circular (which it is, by default) adjust the
cluster coordinates and rewrite the reference fasta as padded.</p>
</dd></dl>

<dl class="method">
<dt id="riboSeed.SeedGenome.purge_old_files">
<code class="descname">purge_old_files</code><span class="sig-paren">(</span><em>all_iters=False</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#SeedGenome.purge_old_files"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.SeedGenome.purge_old_files" title="Permalink to this definition">¶</a></dt>
<dd><p>remove bulky files from two iterations ago, or if
all_iters, remove all big mapping files</p>
</dd></dl>

<dl class="method">
<dt id="riboSeed.SeedGenome.refreshSeqRecGenerator">
<code class="descname">refreshSeqRecGenerator</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#SeedGenome.refreshSeqRecGenerator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.SeedGenome.refreshSeqRecGenerator" title="Permalink to this definition">¶</a></dt>
<dd><p>instead of using stored genbank records, this method restarts
the generator so that each time self.seq_records is accessed
this method can get things going again</p>
</dd></dl>

<dl class="method">
<dt id="riboSeed.SeedGenome.write_fasta_genome">
<code class="descname">write_fasta_genome</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#SeedGenome.write_fasta_genome"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.SeedGenome.write_fasta_genome" title="Permalink to this definition">¶</a></dt>
<dd><p>Given a genbank file, write out as (multi)fasta</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="riboSeed.add_coords_to_clusters">
<code class="descclassname">riboSeed.</code><code class="descname">add_coords_to_clusters</code><span class="sig-paren">(</span><em>seedGenome</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#add_coords_to_clusters"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.add_coords_to_clusters" title="Permalink to this definition">¶</a></dt>
<dd><p>given a genbank file and some locus tags, add the coordinates, etc,
to the entry in the seed Genome</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.check_fastqs_len_equal">
<code class="descclassname">riboSeed.</code><code class="descname">check_fastqs_len_equal</code><span class="sig-paren">(</span><em>file1</em>, <em>file2</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#check_fastqs_len_equal"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.check_fastqs_len_equal" title="Permalink to this definition">¶</a></dt>
<dd><p>using file_len from pyutilsnrw, check that the fastqs contain
the same number of lines, ie tat the pairing looks proper.</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.check_kmer_vs_reads">
<code class="descclassname">riboSeed.</code><code class="descname">check_kmer_vs_reads</code><span class="sig-paren">(</span><em>k</em>, <em>readlen</em>, <em>min_diff=2</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#check_kmer_vs_reads"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.check_kmer_vs_reads" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="riboSeed.convert_bam_to_fastqs_cmd">
<code class="descclassname">riboSeed.</code><code class="descname">convert_bam_to_fastqs_cmd</code><span class="sig-paren">(</span><em>mapping_ob</em>, <em>ref_fasta</em>, <em>samtools_exe</em>, <em>which='mapped'</em>, <em>source_ext='_sam'</em>, <em>single=False</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#convert_bam_to_fastqs_cmd"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.convert_bam_to_fastqs_cmd" title="Permalink to this definition">¶</a></dt>
<dd><p>generate a cmd to convert a bam file to fastq, using samtools</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.copyToHandyDir">
<code class="descclassname">riboSeed.</code><code class="descname">copyToHandyDir</code><span class="sig-paren">(</span><em>outdir</em>, <em>pre</em>, <em>seedGenome</em>, <em>hard=False</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#copyToHandyDir"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.copyToHandyDir" title="Permalink to this definition">¶</a></dt>
<dd><p>copy the resulting contigs</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.decide_proceed_to_target">
<code class="descclassname">riboSeed.</code><code class="descname">decide_proceed_to_target</code><span class="sig-paren">(</span><em>target_len</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#decide_proceed_to_target"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.decide_proceed_to_target" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="riboSeed.estimate_distances_smalt">
<code class="descclassname">riboSeed.</code><code class="descname">estimate_distances_smalt</code><span class="sig-paren">(</span><em>outfile</em>, <em>smalt_exe</em>, <em>ref_genome</em>, <em>fastq1</em>, <em>fastq2</em>, <em>cores=None</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#estimate_distances_smalt"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.estimate_distances_smalt" title="Permalink to this definition">¶</a></dt>
<dd><p>Given fastq pair and a reference, returns path to distance estimations
used by smalt to help later with mapping. if one already exists,
return path to it.</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.evaluate_spades_success">
<code class="descclassname">riboSeed.</code><code class="descname">evaluate_spades_success</code><span class="sig-paren">(</span><em>clu</em>, <em>mapping_ob</em>, <em>proceed_to_target</em>, <em>target_len</em>, <em>include_short_contigs</em>, <em>min_assembly_len</em>, <em>read_len</em>, <em>flank=1000</em>, <em>min_delta=10</em>, <em>keep_best_contig=True</em>, <em>seqname=''</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#evaluate_spades_success"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.evaluate_spades_success" title="Permalink to this definition">¶</a></dt>
<dd><p>return success codes:
0 = include contigs, all good
DEPRECIATED! 1 = include contigs, but dont keep iterating
2 = exclude contigs, and keep from iterating
3 = exclude contigs, error ocurred</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.filter_bam_AS">
<code class="descclassname">riboSeed.</code><code class="descname">filter_bam_AS</code><span class="sig-paren">(</span><em>inbam</em>, <em>outsam</em>, <em>score</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#filter_bam_AS"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.filter_bam_AS" title="Permalink to this definition">¶</a></dt>
<dd><p>This is needed because bwa cannot filter based n alignment score
for paired reads.
<a class="reference external" href="https://sourceforge.net/p/bio-bwa/mailman/message/31968535/">https://sourceforge.net/p/bio-bwa/mailman/message/31968535/</a>
Given a  bam file from bwa (has “AS” tags), write out
reads with AS higher than –score to outsam
read count from <a class="reference external" href="https://www.biostars.org/p/1890/">https://www.biostars.org/p/1890/</a></p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.generate_spades_cmd">
<code class="descclassname">riboSeed.</code><code class="descname">generate_spades_cmd</code><span class="sig-paren">(</span><em>mapping_ob</em>, <em>ngs_ob</em>, <em>ref_as_contig</em>, <em>as_paired=True</em>, <em>addLibs=''</em>, <em>prelim=False</em>, <em>k='21</em>, <em>33</em>, <em>55</em>, <em>77</em>, <em>99'</em>, <em>spades_exe='spades.py'</em>, <em>single_lib=False</em>, <em>logger=None</em>, <em>check_libs=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#generate_spades_cmd"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.generate_spades_cmd" title="Permalink to this definition">¶</a></dt>
<dd><p>return spades command so we can multiprocess the assemblies
wrapper for common spades setting for long illumina reads
ref_as_contig should be either blank, ‘trusted’, or ‘untrusted’
prelim flag is True, only assembly is run, and without coverage corrections
#TODO
the seqname variable is used only for renaming the resulting contigs
during iterative assembly.  It would be nice to inheirit from “ref”,
but that is changed with each iteration. This should probably be addressed
before next major version change
#TODO dynamicllly choose kmers based on read len and whether prelim</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.get_args">
<code class="descclassname">riboSeed.</code><code class="descname">get_args</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#get_args"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.get_args" title="Permalink to this definition">¶</a></dt>
<dd><p>#TODO:     for cli mods:
<a class="reference external" href="http://stackoverflow.com/questions/18025646/">http://stackoverflow.com/questions/18025646/</a></p>
<blockquote>
<div>python-argparse-conditional-requirements</div></blockquote>
<p>make this able to handle different library types such as two unpaired runs</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.get_bam_AS">
<code class="descclassname">riboSeed.</code><code class="descname">get_bam_AS</code><span class="sig-paren">(</span><em>inbam</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#get_bam_AS"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.get_bam_AS" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the mappign scores for downstream QC plotting.</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.get_final_assemblies_cmds">
<code class="descclassname">riboSeed.</code><code class="descname">get_final_assemblies_cmds</code><span class="sig-paren">(</span><em>seedGenome</em>, <em>exes</em>, <em>ref_as_contig</em>, <em>cores</em>, <em>memory</em>, <em>serialize</em>, <em>skip_control=True</em>, <em>kmers='21</em>, <em>33</em>, <em>55</em>, <em>77</em>, <em>99'</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#get_final_assemblies_cmds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.get_final_assemblies_cmds" title="Permalink to this definition">¶</a></dt>
<dd><p>make cmds for runnning of SPAdes and QUAST final assembly and analysis.
if skip_control, just do the de fere novo assembly.  otherwise, do bother
returns list of listed cmds
([[spades_cmd, quast_cmd], [spades_cmd2, quast_cmd2]])</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.get_rec_from_generator">
<code class="descclassname">riboSeed.</code><code class="descname">get_rec_from_generator</code><span class="sig-paren">(</span><em>recordID</em>, <em>gen</em>, <em>method=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#get_rec_from_generator"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.get_rec_from_generator" title="Permalink to this definition">¶</a></dt>
<dd><p>given a record ID and and SeqIO generator return sequence of
genbank record that has the loci, and call method to refresh generator
If on different sequences, return error</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.get_samtools_depths">
<code class="descclassname">riboSeed.</code><code class="descname">get_samtools_depths</code><span class="sig-paren">(</span><em>samtools_exe</em>, <em>bam</em>, <em>chrom</em>, <em>start</em>, <em>end</em>, <em>prep=False</em>, <em>region=None</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#get_samtools_depths"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.get_samtools_depths" title="Permalink to this definition">¶</a></dt>
<dd><p>Use samtools depth and awk to get the average coverage depth of a
particular region</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.get_smalt_full_install_cmds">
<code class="descclassname">riboSeed.</code><code class="descname">get_smalt_full_install_cmds</code><span class="sig-paren">(</span><em>smalt_exe</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#get_smalt_full_install_cmds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.get_smalt_full_install_cmds" title="Permalink to this definition">¶</a></dt>
<dd><p>TODO replace this with swg tests for bambamc installation
In the meantime, this looks for the files included with riboSeed
(a bam file, reference, index, and fastq file), and generates the cmds
to run a little test mapping</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.last_exception">
<code class="descclassname">riboSeed.</code><code class="descname">last_exception</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#last_exception"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.last_exception" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns last exception as a string, or use in logging.
stolen verbatim from pyani</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.make_faux_genome">
<code class="descclassname">riboSeed.</code><code class="descname">make_faux_genome</code><span class="sig-paren">(</span><em>cluster_list</em>, <em>seedGenome</em>, <em>iteration</em>, <em>output_root</em>, <em>nbuff</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#make_faux_genome"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.make_faux_genome" title="Permalink to this definition">¶</a></dt>
<dd><p>stictch together viable assembled contigs.  perhaps more importnatly,
this also re-write thes coords relative to the new “genome”
returns path to new faux_genome</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.make_mapped_partition_cmds">
<code class="descclassname">riboSeed.</code><code class="descname">make_mapped_partition_cmds</code><span class="sig-paren">(</span><em>cluster</em>, <em>mapping_ob</em>, <em>seedGenome</em>, <em>samtools_exe</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#make_mapped_partition_cmds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.make_mapped_partition_cmds" title="Permalink to this definition">¶</a></dt>
<dd><p>returns cmds and region</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.make_modest_spades_cmd">
<code class="descclassname">riboSeed.</code><code class="descname">make_modest_spades_cmd</code><span class="sig-paren">(</span><em>cmd</em>, <em>cores</em>, <em>memory</em>, <em>split=0</em>, <em>serialize=False</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#make_modest_spades_cmd"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.make_modest_spades_cmd" title="Permalink to this definition">¶</a></dt>
<dd><p>adjust spades commands to use set amounts of cores and memory
returns the command, split on “–careful”, cause why would you run
SPAdes with “–reckless”?
if you need to split resouces between, say, two commands, use split 2</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.make_quick_quast_table">
<code class="descclassname">riboSeed.</code><code class="descname">make_quick_quast_table</code><span class="sig-paren">(</span><em>pathlist</em>, <em>write=False</em>, <em>writedir=None</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#make_quick_quast_table"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.make_quick_quast_table" title="Permalink to this definition">¶</a></dt>
<dd><p>given paths to two or more quast reports, this generates dictionary
where the key is the field in the report and the value is a list of
the values for each report.   Hand for passing to the logger function.
This skips any fields not in first report, for better or worse…</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.make_samtools_depth_cmds">
<code class="descclassname">riboSeed.</code><code class="descname">make_samtools_depth_cmds</code><span class="sig-paren">(</span><em>exe</em>, <em>bam</em>, <em>chrom</em>, <em>start</em>, <em>end</em>, <em>region=None</em>, <em>prep=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#make_samtools_depth_cmds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.make_samtools_depth_cmds" title="Permalink to this definition">¶</a></dt>
<dd><p>this just makes the commands to get the depth from samtools.
If prep, the sam file gets sorted and indexed first</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.make_spades_empty_check">
<code class="descclassname">riboSeed.</code><code class="descname">make_spades_empty_check</code><span class="sig-paren">(</span><em>liblist</em>, <em>cmd</em>, <em>logger</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#make_spades_empty_check"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.make_spades_empty_check" title="Permalink to this definition">¶</a></dt>
<dd><p>returns shell/spades cmd as string. All this does is make it a
conditional shell cmd that depends on the presense of the file
needed for assembly.  It is needed so we can bin all
the cmds with multiprocessing.</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.make_unmapped_partition_cmds">
<code class="descclassname">riboSeed.</code><code class="descname">make_unmapped_partition_cmds</code><span class="sig-paren">(</span><em>mapped_regions</em>, <em>samtools_exe</em>, <em>seedGenome</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#make_unmapped_partition_cmds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.make_unmapped_partition_cmds" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="riboSeed.map_to_genome_ref_bwa">
<code class="descclassname">riboSeed.</code><code class="descname">map_to_genome_ref_bwa</code><span class="sig-paren">(</span><em>mapping_ob</em>, <em>ngsLib</em>, <em>cores</em>, <em>samtools_exe</em>, <em>bwa_exe</em>, <em>genome_fasta</em>, <em>score_minimum=None</em>, <em>add_args='-L 0</em>, <em>0 -U 0 -a'</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#map_to_genome_ref_bwa"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.map_to_genome_ref_bwa" title="Permalink to this definition">¶</a></dt>
<dd><p>Map to bam.  maps PE and S reads separately,
then combines them into a X_mapped.bam file
TODO:: break up into execution and comamnd generation</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.map_to_genome_ref_smalt">
<code class="descclassname">riboSeed.</code><code class="descname">map_to_genome_ref_smalt</code><span class="sig-paren">(</span><em>mapping_ob</em>, <em>ngsLib</em>, <em>cores</em>, <em>samtools_exe</em>, <em>smalt_exe</em>, <em>genome_fasta</em>, <em>score_minimum=None</em>, <em>scoring='match=1</em>, <em>subst=-4</em>, <em>gapopen=-4</em>, <em>gapext=-3'</em>, <em>step=3</em>, <em>k=5</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#map_to_genome_ref_smalt"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.map_to_genome_ref_smalt" title="Permalink to this definition">¶</a></dt>
<dd><p>run smalt based on pased args
#TODO rework this to read libtype of ngslib object
requires at least paired end input, but can handle an additional library
of singleton reads. Will not work on just singletons</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.nonify_empty_lib_files">
<code class="descclassname">riboSeed.</code><code class="descname">nonify_empty_lib_files</code><span class="sig-paren">(</span><em>ngsLib</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#nonify_empty_lib_files"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.nonify_empty_lib_files" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="riboSeed.parse_samtools_depth_results">
<code class="descclassname">riboSeed.</code><code class="descname">parse_samtools_depth_results</code><span class="sig-paren">(</span><em>result</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#parse_samtools_depth_results"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.parse_samtools_depth_results" title="Permalink to this definition">¶</a></dt>
<dd><p>parses out the subprocess results from samtools depth</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.parse_subassembly_return_code">
<code class="descclassname">riboSeed.</code><code class="descname">parse_subassembly_return_code</code><span class="sig-paren">(</span><em>cluster</em>, <em>final_contigs_dir</em>, <em>skip_copy=False</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#parse_subassembly_return_code"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.parse_subassembly_return_code" title="Permalink to this definition">¶</a></dt>
<dd><p>given a return code from the above spades success function,
set object attributes as needed
20170531 depreciated return code 1: as we have moved to using the
whole library for mapping, not just the unmapped reads, it is more
important to keep in even unchanging contigs to allow for
competition during mapping</p>
<p>return success codes:
0 = include contigs, all good
1 = include contigs, but dont keep iterating
2 = exclude contigs, and keep from iterating
3 = exclude contigs, error ocurred</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.partition_mapping">
<code class="descclassname">riboSeed.</code><code class="descname">partition_mapping</code><span class="sig-paren">(</span><em>seedGenome</em>, <em>samtools_exe</em>, <em>flank</em>, <em>min_flank_depth</em>, <em>cluster_list=None</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#partition_mapping"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.partition_mapping" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract interesting stuff based on coords, not a binary
mapped/not_mapped condition
Also, if min_flanking_depth, mark reads with low
mapping coverage for exclusion</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.plotAsScores">
<code class="descclassname">riboSeed.</code><code class="descname">plotAsScores</code><span class="sig-paren">(</span><em>score_list</em>, <em>score_min</em>, <em>outdir</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#plotAsScores"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.plotAsScores" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="riboSeed.prepare_next_mapping">
<code class="descclassname">riboSeed.</code><code class="descname">prepare_next_mapping</code><span class="sig-paren">(</span><em>cluster</em>, <em>seedGenome</em>, <em>samtools_exe</em>, <em>flank</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#prepare_next_mapping"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.prepare_next_mapping" title="Permalink to this definition">¶</a></dt>
<dd><p>use within partition mapping funtion;
makes LociMapping, get region coords, write extracted region,</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.printPlot">
<code class="descclassname">riboSeed.</code><code class="descname">printPlot</code><span class="sig-paren">(</span><em>data</em>, <em>line=None</em>, <em>ymax=30</em>, <em>xmax=60</em>, <em>tick=0.2</em>, <em>title='test'</em>, <em>fill=False</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#printPlot"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.printPlot" title="Permalink to this definition">¶</a></dt>
<dd><p>ascii not what your program can do for you…</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.pysam_extract_reads">
<code class="descclassname">riboSeed.</code><code class="descname">pysam_extract_reads</code><span class="sig-paren">(</span><em>sam</em>, <em>textfile</em>, <em>unmapped_sam</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#pysam_extract_reads"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.pysam_extract_reads" title="Permalink to this definition">¶</a></dt>
<dd><p>This replaces the “LC_ALL ” grep call from the above function. On macs,
there is no speedup gained.</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.reportRegionDepths">
<code class="descclassname">riboSeed.</code><code class="descname">reportRegionDepths</code><span class="sig-paren">(</span><em>inp</em>, <em>logger</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#reportRegionDepths"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.reportRegionDepths" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="riboSeed.sam_to_bam">
<code class="descclassname">riboSeed.</code><code class="descname">sam_to_bam</code><span class="sig-paren">(</span><em>samtools_exe</em>, <em>bam</em>, <em>sam</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#sam_to_bam"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.sam_to_bam" title="Permalink to this definition">¶</a></dt>
<dd><p>becasue pysam doesnt like to write bams in an iterator, which makes sense</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.subprocess_run_list">
<code class="descclassname">riboSeed.</code><code class="descname">subprocess_run_list</code><span class="sig-paren">(</span><em>cmdlist</em>, <em>hard=False</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#subprocess_run_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.subprocess_run_list" title="Permalink to this definition">¶</a></dt>
<dd><p>This just allows for sequential cmds with multiprocessing.
It prevents the errors when future commands are looking for and not finding
a needed file.
Logger cant be used with multiprocessing
returns 0 if all is well, otherwise returns 1
if hard == True, quits instead of returning 1</p>
</dd></dl>

<dl class="function">
<dt id="riboSeed.test_smalt_bam_install">
<code class="descclassname">riboSeed.</code><code class="descname">test_smalt_bam_install</code><span class="sig-paren">(</span><em>cmds</em>, <em>logger=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/riboSeed.html#test_smalt_bam_install"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#riboSeed.test_smalt_bam_install" title="Permalink to this definition">¶</a></dt>
<dd><p>using test data tha tcomes with package, ensure that
the bambamc library was properly installed with SMALT instaltation</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">riboSeed package</a><ul>
<li><a class="reference internal" href="#subpackages">Subpackages</a></li>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#riboseed-plotmsa-module">riboSeed.plotMSA module</a></li>
<li><a class="reference internal" href="#riboseed-riboscan-module">riboSeed.riboScan module</a></li>
<li><a class="reference internal" href="#riboseed-riboscore-module">riboSeed.riboScore module</a></li>
<li><a class="reference internal" href="#riboseed-riboseed-module">riboSeed.riboSeed module</a></li>
<li><a class="reference internal" href="#riboseed-riboselect-module">riboSeed.riboSelect module</a></li>
<li><a class="reference internal" href="#riboseed-ribosim-module">riboSeed.riboSim module</a></li>
<li><a class="reference internal" href="#riboseed-ribosketch-module">riboSeed.riboSketch module</a></li>
<li><a class="reference internal" href="#riboseed-ribosnag-module">riboSeed.riboSnag module</a></li>
<li><a class="reference internal" href="#riboseed-ribostack-module">riboSeed.riboStack module</a></li>
<li><a class="reference internal" href="#riboseed-riboswap-module">riboSeed.riboSwap module</a></li>
<li><a class="reference internal" href="#module-riboSeed">Module contents</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/riboSeed.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">riboSeed 0.4.9 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2017, Nicholas Waters, Florence Abram, Fiona Brennan, Ashleigh Holmes, and Leighton Pritchard.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.6.3.
    </div>
  </body>
</html>